{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3054864-6f9a-44d5-b3a3-ff29ccde5e7c",
   "metadata": {},
   "source": [
    "# Submodule 1: Introduction to genome sequencing and assembly\n",
    "--------\n",
    "Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. \n",
    "\n",
    "With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. A critical component of genomics is genome assembly, which involves reconstructing the genome of an organism from short DNA sequences generated by sequencing technologies. Genome assembly is a foundational step that allows researchers to generate a complete picture of the genetic material in a given organism. Assessing the quality and completeness of a genome assembly ensures that it accurately reflects the original genome, providing a reliable foundation for further analysis.\n",
    "\n",
    "Comparative genomics builds on genome assembly by comparing the genomes of different species or individuals. This field focuses on identifying similarities and differences in DNA sequences to understand evolutionary relationships, discover conserved genetic elements, and reveal the genetic underpinnings of adaptation and diversity. Comparative genomics can uncover how species evolve and adapt over time and identify genes associated with specific traits or diseases.\n",
    "\n",
    "In this tutorial, we will explore the key concepts and methodologies in genomics, including how to assemble a genome, assess its quality, and perform comparative analyses. Through hands-on examples, you will learn how to analyze genomic data, interpret results, and apply various computational tools to gain deeper insights into the genetic information that defines life.\n",
    "\n",
    " \n",
    "\n",
    "### Learning Objectives:\n",
    "\n",
    "+ Understand how high throughput sequencing data is generated. \n",
    "\n",
    "+ Develop an understanding of core bioinformatic input/output formats as it relates to comparative genomics. \n",
    "\n",
    "+ Acquire the skills to assemble raw sequencing reads into a draft genome, assess the quality of the genome assembly, and annotate the genome sequence. \n",
    "\n",
    "+ Learn to perform comparative genomic analyses to identify similarities and differences across genomes, run phylogenomic analyses, construct pangenomes to capture genetic diversity, and apply these techniques to address biological questions and hypotheses. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4c2a457-80a1-45f9-9bb8-7cf38ff99281",
   "metadata": {},
   "source": [
    "## 1.1 - Generation of sequencing data\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "<img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/fragmentation3.png?raw=true\" width=\"800\">\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07881f7e-9a58-45d2-9bef-ab24929d8227",
   "metadata": {},
   "source": [
    "## 1.2 - How next-generation sequencing works\n",
    "\n",
    "[![sequencing by synthesis](img/youtube-video-sequencing.PNG)](https://www.youtube.com/watch?v=p4vKJJlNTKA&t=9s \"Sequencing\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d01dbeac-d525-4528-a3c2-ac527faa8108",
   "metadata": {},
   "source": [
    "### 1.2.1 Cost of generating sequencing data\n",
    "\n",
    "CItation: Wetterstrand KA. DNA Sequencing Costs: Data from the NHGRI Genome Sequencing Program (GSP) Available at: www.genome.gov/sequencingcostsdata. Accessed [10/03/2024]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b158a53b-3182-4a63-9827-0d20c375aeee",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 1.2.2 Generate plot showing the cost of sequencing genomes over time\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.dates as mdates\n",
    "\n",
    "# Load the CSV file\n",
    "data_table = \"data/Sequencing_Cost_Data_Table_May2022.csv\"\n",
    "df = pd.read_csv(data_table)\n",
    "\n",
    "# Convert the 'Date' column to datetime format using YYYY-MM\n",
    "df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m')\n",
    "\n",
    "# Clean up 'Cost per Genome' column by removing $ and commas, and converting to numeric\n",
    "df['Cost per Genome'] = df['Cost per Genome'].replace({'\\$': '', ',': ''}, regex=True).astype(float)\n",
    "\n",
    "# Plot the data\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# X-axis: Date, Y-axis: Cost per Genome in log scale\n",
    "plt.plot(df['Date'], df['Cost per Genome'], marker='o', linestyle='-', color='b')\n",
    "\n",
    "# Set log scale for y-axis\n",
    "plt.yscale('log')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Year', fontsize=14, fontweight='bold')\n",
    "plt.ylabel('Cost per Genome (Log scale)', fontsize=12, fontweight='bold')\n",
    "plt.title('Cost per Genome Over Time', fontsize=14, fontweight='bold')\n",
    "\n",
    "# Format the x-axis for better date presentation\n",
    "import matplotlib.dates as mdates\n",
    "\n",
    "# Set major ticks to display years at specific intervals\n",
    "plt.gca().xaxis.set_major_locator(mdates.YearLocator(3))  # Adjust the number here for the interval\n",
    "plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%Y'))\n",
    "\n",
    "# Rotate and format x-ticks\n",
    "plt.xticks(rotation=45, fontsize=8, fontweight='bold')\n",
    "\n",
    "\n",
    "# Set the y-axis limits from $100 to $100,000\n",
    "# plt.ylim(100, 0000)\n",
    "plt.yticks(fontsize=12, fontweight='bold')\n",
    "\n",
    "# Display the plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cost_per_genome_plot.png\", dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7b4226-51c1-4e68-9900-38055832b945",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eca78aeb-2a38-42d8-b163-a4ab47acbbca",
   "metadata": {},
   "source": [
    "## 1.3 - Comparison of sequencing instruments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6bd27d-90a9-4fac-98f5-ce003a5c097c",
   "metadata": {},
   "source": [
    "## 1.4 - Starting data\n",
    "\n",
    "Each of these samples represent the genome of a unique and novel microbe that has not been seen before (except by me). \n",
    "\n",
    "The data utilized in this tutorial was generated by Illumina HiSeq 2500, is paired-end and each read is 250 bps n length. \n",
    "\n",
    "<p align=\"center\">\n",
    "    <img src=\"images/jupyterNotebook_annotated.png\" alt=\"jupyterNotebook\" width=\"50%\"/>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b43a7cc-7801-4f9e-b18c-9d5996a33166",
   "metadata": {},
   "source": [
    "### 1.4.1 - Install sra-tools to access public data available on the NCBI Sequence Read Archive (SRA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9db7442-a02e-46cc-8236-28e5284915a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Install fastq-dump using mamba ( a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 sra-tools -y > /dev/null 2>&1\n",
    "\n",
    "# hide standard error and standard out > /dev/null 2>&1\n",
    "# hide standard error only 2> /dev/null\n",
    "\n",
    "echo \"Installation of sra-tools complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# setup working directory and move into it\n",
    "mkdir wgs-workflow\n",
    "cd wgs-workflow\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "prefetch -v $accession\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1b1056-ae52-4290-a2a0-615d06512281",
   "metadata": {},
   "source": [
    "### 1.4.2 - FASTQ file format\n",
    "These data are in **FASTQ** format (see below).\n",
    "\n",
    "[Link explaining the 'Read Name Format'](http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm): SampleName_Barcode_LaneNumber_001.fastq.gz\n",
    "\n",
    "\n",
    "Important note: In the above command I use the \"\\*\" character to view the Sample directory, I would normally just type out the entire path using tab complete (which is what you should do). This wildcard will match any string of characters. I use this because everyone will have a different Sample name. To make this tutorial as general as possible I need to use these wildcards throughout the tutorial. In addition I may use Sample_X instead of Sample_\\*. In these cases be sure to type out your complete sample name!, the wildcards probably won't work "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7622a4-dd64-497e-96ea-8d9b83d85281",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# If the above code worked you should have two FASTQ files in a directory called raw-reads\n",
    "ls raw-reads/\n",
    "\n",
    "# Pull the raw fastq files you need from the SRA\n",
    "zcat raw-reads/SRR10056829_1.fastq.gz | head -n 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e803f847-b1ba-48a5-8193-a4d3eeba75bb",
   "metadata": {},
   "source": [
    "## 1.5 - Quality assessment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba7784-527e-461b-a3de-a72b79fc52c5",
   "metadata": {},
   "source": [
    "### 1.5.1 Do we have enough coverage? How do we determine our total throughput?\n",
    "\n",
    "INSERT FLASH CARD WITH QUESTIONS ABOUT READ LENGTH AND COVERAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1874fbae-7b77-49de-802f-f620729479b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "zcat raw-reads/*_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab968490-19b8-4bf7-9582-5973da86492f",
   "metadata": {},
   "source": [
    "Whats our total bp of data? This is what we call our sequencing throughput. We multiple the number of reads by the read length (ours is 250) and by 2 because it is paired-end data.\n",
    "(Read length x 2(paired-end) x Number of reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f8ae91a-af8f-42fd-b26a-7a8385e01d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "num_reads=6365120\n",
    "read_length=250\n",
    "genome_size=5000000\n",
    "\n",
    "# calculate our total throughput\n",
    "echo \"Total number of base pairs sequenced\"\n",
    "throughput=$(echo $num_reads \" * \" $read_length \" * 2\" | bc)\n",
    "echo $throughput\n",
    "\n",
    "# calculate our estimated coverage\n",
    "echo \"Total coverage for genome dataset\"\n",
    "est_coverage=$(echo $throughput \" / \" \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414ab82a-3f91-43d3-8c07-025f8a3379b2",
   "metadata": {},
   "source": [
    "### 1.5.2 Quality Assessment with FASTQC\n",
    "\n",
    "Program: FASTQC\n",
    "Manual: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
    "FASTQC explained:\n",
    "Conda:\n",
    "\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and do the hard work for us. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, as with the majority of the assessment tools, it merely reads it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9d122c-320d-458e-a61b-02b6222f65e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Install fastqc using mamba ( a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 fastqc -y > /dev/null 2>&1\n",
    "\n",
    "# > /dev/null 2>&1 : hides standard error and standard out \n",
    "# 2> /dev/null : hides standard error only \n",
    "\n",
    "echo \"Installation of fastqc complete\"\n",
    "fastqc --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "506c05fd-a403-4eb3-bebb-7d0cfcc51246",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir output-fastqc\n",
    "\n",
    "# run the program on the two read files (R1 and R2).\n",
    "fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.\n",
    "ls output-fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfb2d6a-1a60-4b89-9898-1f5079d95962",
   "metadata": {},
   "outputs": [],
   "source": [
    "### view HTML report\n",
    "from IPython.display import HTML\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb6207a-2b47-453f-a4d0-b38bff239dab",
   "metadata": {},
   "source": [
    "## 1.6 - Quality and adapter trimming with FASTP\n",
    "\n",
    "Program: fastp - an ultra-fast all-in-one FASTQ preprocessor\n",
    "Citation: *Chen, S. (2023). Ultrafast oneâ€pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*\n",
    "Manual: https://github.com/OpenGene/fastp\n",
    "Conda: https://anaconda.org/bioconda/fastp\n",
    "\n",
    "Description:\n",
    "Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0df2498-4184-4403-b1dc-9cc93b98f955",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Install fastp using mamba (a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 fastp -y > /dev/null 2>&1\n",
    "\n",
    "# hide standard error and standard out > /dev/null 2>&1\n",
    "# hide standard error only 2> /dev/null\n",
    "\n",
    "fastp --version\n",
    "echo \"Installation Complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18047f7b-4838-4133-bf51-aa1d139caf2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# make a directory to store the output\n",
    "mkdir output-fastp\n",
    "\n",
    "# define input variables\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# run the program\n",
    "fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# the resulting folder should contain new trimmed read files\n",
    "ls output-fastp/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa89305-5b2d-4614-9cd2-c144224187bc",
   "metadata": {},
   "source": [
    "## 1.7 - Genome Assembly\n",
    "\n",
    "Program: SPAdes - St. Petersburg genome assembler\n",
    "Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*\n",
    "Manual: https://ablab.github.io/spades/\n",
    "Conda: https://anaconda.org/bioconda/spades\n",
    "\n",
    "\n",
    "Description: \n",
    "SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. However, it might not be suitable for large genomes projects.\n",
    "\n",
    "SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91bac579-1659-49c4-a60f-478f73ec5d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Install SPAdes\n",
    "mamba install --channel bioconda python=3.9 spades -y > /dev/null 2>&1\n",
    "\n",
    "spades.py --version\n",
    "echo \"Installation Complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6681b8a5-b9a7-4924-bb8b-9961d14bd320",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "## Assemble trimmed reads using the SPAdes assembler.\n",
    "\n",
    "# define input\n",
    "forward=output-fastp/trimmed_1.fastq.gz\n",
    "reverse=output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# assemble reads with SPAdes\n",
    "spades.py -1 $forward -2 $reverse -o output-spades"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
