{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "193c3802-fc11-4ac3-b65b-0acff481d0a2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Submodule 1: Introduction to genome sequencing and assembly\n",
    "--------\n",
    "Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. \n",
    "\n",
    "With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. A critical component of genomics is genome assembly, which involves reconstructing the genome of an organism from short DNA sequences generated by sequencing technologies. Genome assembly is a foundational step that allows researchers to generate a complete picture of the genetic material in a given organism. Assessing the quality and completeness of a genome assembly ensures that it accurately reflects the original genome, providing a reliable foundation for further analysis.\n",
    "\n",
    "Comparative genomics builds on genome assembly by comparing the genomes of different species or individuals. This field focuses on identifying similarities and differences in DNA sequences to understand evolutionary relationships, discover conserved genetic elements, and reveal the genetic underpinnings of adaptation and diversity. Comparative genomics can uncover how species evolve and adapt over time and identify genes associated with specific traits or diseases.\n",
    "\n",
    "In this tutorial, we will explore the key concepts and methodologies in genomics, including how to assemble a genome, assess its quality, and perform comparative analyses. Through hands-on examples, you will learn how to analyze genomic data, interpret results, and apply various computational tools to gain deeper insights into the genetic information that defines life.\n",
    "\n",
    " \n",
    "\n",
    "### Learning Objectives:\n",
    "\n",
    "+ Understand how high throughput sequencing data is generated. \n",
    "\n",
    "+ Develop an understanding of core bioinformatic input/output formats as it relates to comprative genomics. \n",
    "\n",
    "+ Acquire the skills to assemble raw sequencing reads into a draft genome, assess the quality of the genome assembly, and annotate the genome sequence. \n",
    "\n",
    "+ Learn to perform comparative genomic analyses to identify similarities and differences across genomes, run phylogenomic analyses, construct pangenomes to capture genetic diversity, and apply these techniques to address biological questions and hypotheses. \n",
    "\n",
    "\n",
    "## 1.1 - Generation of sequencing data\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "<img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/fragmentation3.png?raw=true\" width=\"800\">\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">\n",
    "\n",
    "\n",
    "## 1.2 - How next-generation sequencing works\n",
    "\n",
    "[![sequencing by synthesis](img/youtube-video-sequencing.PNG)](https://www.youtube.com/watch?v=p4vKJJlNTKA&t=9s \"Sequencing\")\n",
    "\n",
    "\n",
    "\n",
    "## 1.3 - Comparison of sequencing instruments\n",
    "\n",
    "\n",
    "\n",
    "## 1.4 - Starting data\n",
    "\n",
    "Each of these samples represent the genome of a unique and novel microbe that has not been seen before (except by me). \n",
    "\n",
    "The data utilized in this tutorial was generated by Illumina HiSeq 2500, is paired-end and each read is 250 bps n length. \n",
    "\n",
    "<p align=\"center\">\n",
    "    <img src=\"images/jupyterNotebook_annotated.png\" alt=\"jupyterNotebook\" width=\"50%\"/>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b43a7cc-7801-4f9e-b18c-9d5996a33166",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.4.1 - Download sra-tools to access public data available on the NCBI Sequence Read Archive (SRA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9db7442-a02e-46cc-8236-28e5284915a0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Install fastq-dump using mamba ( a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 sra-tools -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5ff37260-1139-4904-b05c-8f91ec9a9e1d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  All requested packages already installed\n",
      "\n",
      "\n",
      "Looking for: ['python=3.9', 'sra-tools']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### 1.4.2 - Download sequnce data from the SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 12: prefetch: command not found\n",
      "bash: line 15: fastq-dump: command not found\n",
      "rm: cannot remove ‘SRR10056829’: No such file or directory\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'\\n# setup working directory and move into it\\nmkdir wgs-workflow\\ncd wgs-workflow\\n\\n# Capture the SRA accession in a variable \\naccession=SRR10056829\\n\\n# Download data from the SRA using the variable from above \\n\\n# prefetching downloads metadata of the SRA records (5.629s)\\nprefetch -v $accession\\n\\n# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\\nfastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\\n\\n# Remove the prefetch directory\\nrm -r $accession\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mget_ipython\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_cell_magic\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mbash\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# setup working directory and move into it\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mmkdir wgs-workflow\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mcd wgs-workflow\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# Capture the SRA accession in a variable \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43maccession=SRR10056829\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# Download data from the SRA using the variable from above \u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# prefetching downloads metadata of the SRA records (5.629s)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mprefetch -v $accession\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mfastq-dump --outdir raw-reads --gzip --split-files \u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m$accession\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m$accession\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m.sra\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m# Remove the prefetch directory\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43mrm -r $accession\u001b[39;49m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/python3/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2541\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2539\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2540\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2541\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2543\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2544\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2545\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2546\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/anaconda3/envs/python3/lib/python3.10/site-packages/IPython/core/magics/script.py:155\u001b[0m, in \u001b[0;36mScriptMagics._make_script_magic.<locals>.named_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    154\u001b[0m     line \u001b[38;5;241m=\u001b[39m script\n\u001b[0;32m--> 155\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshebang\u001b[49m\u001b[43m(\u001b[49m\u001b[43mline\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/python3/lib/python3.10/site-packages/IPython/core/magics/script.py:315\u001b[0m, in \u001b[0;36mScriptMagics.shebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m args\u001b[38;5;241m.\u001b[39mraise_error \u001b[38;5;129;01mand\u001b[39;00m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;66;03m# If we get here and p.returncode is still None, we must have\u001b[39;00m\n\u001b[1;32m    312\u001b[0m     \u001b[38;5;66;03m# killed it but not yet seen its return code. We don't wait for it,\u001b[39;00m\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;66;03m# in case it's stuck in uninterruptible sleep. -9 = SIGKILL\u001b[39;00m\n\u001b[1;32m    314\u001b[0m     rc \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m9\u001b[39m\n\u001b[0;32m--> 315\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(rc, cell)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\n# setup working directory and move into it\\nmkdir wgs-workflow\\ncd wgs-workflow\\n\\n# Capture the SRA accession in a variable \\naccession=SRR10056829\\n\\n# Download data from the SRA using the variable from above \\n\\n# prefetching downloads metadata of the SRA records (5.629s)\\nprefetch -v $accession\\n\\n# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\\nfastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\\n\\n# Remove the prefetch directory\\nrm -r $accession\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# setup working directory and move into it\n",
    "mkdir wgs-workflow\n",
    "cd wgs-workflow\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "prefetch -v $accession\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b0476b1f-8594-4909-bc40-c1cd6b980fd6",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.4.2 - FASTQ file format\n",
    "These data are in **FASTQ** format (see below).\n",
    "\n",
    "[Link explaining the 'Read Name Format'](http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm): SampleName_Barcode_LaneNumber_001.fastq.gz\n",
    "\n",
    "\n",
    "Important note: In the above command I use the \"\\*\" character to view the Sample directory, I would normally just type out the entire path using tab complete (which is what you should do). This wildcard will match any string of characters. I use this because everyone will have a different Sample name. To make this tutorial as general as possible I need to use these wildcards throughout the tutorial. In addition I may use Sample_X instead of Sample_\\*. In these cases be sure to type out your complete sample name!, the wildcards probably won't work "
   ]
  },
  {
   "cell_type": "raw",
   "id": "cb123f74-9d23-4eef-9437-d1c14c933127",
   "metadata": {
    "tags": []
   },
   "source": [
    "%%bash\n",
    "\n",
    "# Pull the raw fastq files you need from the SRA\n",
    "ls\n",
    "\n",
    "echo hello"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a9d24386-bbe1-44a6-a823-71f044f453d4",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.5 - Quality assessment\n",
    "\n",
    "Cells in the code format can be selected with your mouse and executed within the notebook such that the results of the code chunk will print to your screen. The cell below is in the **code format** there are three types of lines in this cell. \n",
    "\n",
    "- The first line `%%bash` indicates the code that follows is in the BASH coding language.\n",
    "- The second line is in green and starts with a `#` this is a comment line. Commented lines are used to describe the action the code line below will take. These lines are important because they help you think logically about what the code is doing. Importantly any line that begins with a `#` will not be executed, so if there is code that you don't want to run in a cell you can add a `#` in front of it and the code will not be executed.\n",
    "- The third line is the code line, in this code we use the `echo` command to print the argument \"Hello World\" to the screen.\n",
    "\n",
    "To run the code in a code cell, click on the cell and use the **Run Code** button in the menu above. The code will be executed and the results will be printed below the code cell in your notebook. You can also run the command by clicking on the code cell and using the `shift + enter` keys. "
   ]
  },
  {
   "cell_type": "raw",
   "id": "a1f8a570-5203-4b09-85d4-02a15f04a516",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Do we have enough coverage? How do we determine our total throughput?\n",
    "\n",
    "INSERT FLASH CARD WITH QUESTIONS ABOUT READ LENGTH AND COVERAGE"
   ]
  },
  {
   "cell_type": "raw",
   "id": "240f0446-5915-4409-b17c-1936495b8e7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "zcat Sample*/*_R1_* | wc -l"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab968490-19b8-4bf7-9582-5973da86492f",
   "metadata": {
    "tags": []
   },
   "source": [
    "Whats our total bp of data? This is what we call our sequencing throughput. We multiple the number of reads by the read length (ours is 250) and by 2 because it is paired-end data.\n",
    "(Read length x 2(paired-end) x Number of reads)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "df397fae-5c2b-4f62-8678-2bb32d5fc31a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "echo \"Number_of_reads * 250 * 2\" | bc"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ecc1ae08-6912-4c40-9ffa-ed9f4fb0d1c2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Examine Read Quality\n",
    "\n",
    "program: FASTQC\n",
    "manual: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
    "\n",
    "FASTQC explained\n",
    "\n",
    "#Run Fastqc\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and do the hard work for us. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, as with the majority of the assessment tools, it merely reads it.\n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "73a3112b-bd3f-424a-a1d6-36517958ae97",
   "metadata": {
    "tags": []
   },
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir fastqc_raw-reads\n",
    "\n",
    "# run the program\n",
    "fastqc Sample_*/*_R1_* Sample_*/*_R2_* -o fastqc_raw-reads\n",
    "ls fastqc_raw-reads\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant."
   ]
  },
  {
   "cell_type": "raw",
   "id": "d3dbe782-1712-481d-bbdb-090b35dc6615",
   "metadata": {
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "raw",
   "id": "d112a645-954f-4074-883c-51477db8d64b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.6 - Quality and adapter trimming"
   ]
  },
  {
   "cell_type": "raw",
   "id": "725a242c-f7ef-456b-9e5b-8fc0cf1f425f",
   "metadata": {
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b89604a-c3cb-45ab-950b-e1917b5ff21f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
