{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3054864-6f9a-44d5-b3a3-ff29ccde5e7c",
   "metadata": {},
   "source": [
    "# Submodule 1: Introduction to genome sequencing and assembly\n",
    "--------\n",
    "## Overview\n",
    "\n",
    "Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. \n",
    "\n",
    "With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. A critical component of genomics is genome assembly, which involves reconstructing the genome of an organism from short DNA sequences generated by sequencing technologies. Genome assembly is a foundational step that allows researchers to generate a complete picture of the genetic material in a given organism. Assessing the quality and completeness of a genome assembly ensures that it accurately reflects the original genome, providing a reliable foundation for further analysis.\n",
    "\n",
    "Comparative genomics builds on genome assembly by comparing the genomes of different species or individuals. This field focuses on identifying similarities and differences in DNA sequences to understand evolutionary relationships, discover conserved genetic elements, and reveal the genetic underpinnings of adaptation and diversity. Comparative genomics can uncover how species evolve and adapt over time and identify genes associated with specific traits or diseases.\n",
    "\n",
    "In this tutorial, we will explore the key concepts and methodologies in genomics, including how to assemble a genome, how to assess its quality, how to annotate the genome, and how to perform comparative analyses.\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/diagram-WGS.png\" width=\"70%\"/>\n",
    "</p>\n",
    "\n",
    "\n",
    "\n",
    "### Learning Objectives:\n",
    "\n",
    "+ Understand how high throughput sequencing data is generated. \n",
    "\n",
    "+ Develop an understanding of core bioinformatic input/output formats as it relates to comparative genomics. \n",
    "\n",
    "+ Acquire the skills to assemble raw sequencing reads into a draft genome, assess the quality of the genome assembly, and annotate the genome sequence. \n",
    "\n",
    "+ Learn to perform comparative genomic analyses to identify similarities and differences across genomes, run phylogenomic analyses, construct pangenomes to capture genetic diversity, and apply these techniques to address biological questions and hypotheses. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15fae599-ea0f-4376-8f97-f1409a2b64d6",
   "metadata": {},
   "source": [
    "## Background: How is sequencing data produced?\n",
    "\n",
    "### What is Next-Generation Sequencing (NGS)?\n",
    "\n",
    "Next-Generation Sequencing (NGS) is a high-throughput method that allows for rapid sequencing of DNA. Illumina remains the most widely used sequenicng platform. It is accurate, inexpensive, and fast, providing up to 540 Gb of data on a single flow cell in about a 1-2 days. Other long-read sequencing technologies (i.e. PacBio and Nanopore) are becoming widely used in the research community, and will be discussed in later modules.\n",
    "\n",
    "Figure 1, outlines the basic workflow for sequencing a microbial genome. It begins with the **isolation of bacterial cultures**, which involves culturing bacteria from a sample on selective media to obtain a pure strain. Once a pure culture is established, **DNA extraction** is performed to isolate the genomic DNA from the bacterial cells.\n",
    "\n",
    "For Illumina-based sequencing the DNA is **fragmented** into smaller pieces and attached to specially coated surfaces, creating a dense cluster of identical fragments. Through a process called **sequencing by synthesis**, fluorescently labeled nucleotides are incorporated one at a time into these clusters. Each incorporated nucleotide emits a distinct signal, allowing the sequencer to determine the DNA sequence in real time.\n",
    "\n",
    "This technology enables millions of fragments to be sequenced simultaneously, significantly increasing throughput and reducing costs.\n",
    "\n",
    "\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "<div style=\"text-align: center;\">\n",
    "<img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/fragmentation3.png?raw=true\" width=\"700\">\n",
    "</div>\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Video Lesson: How next-generation sequencing works.\n",
    "\n",
    "The following video describes the process in detail.\n",
    "\n",
    "[![Sequencing by Synthesis](images/cluster-generation.jpg)](https://www.youtube.com/watch?v=p4vKJJlNTKA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1ee5c7c9-6c44-4da7-8df2-f7c5a40c634a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/p4vKJJlNTKA\" frameborder=\"0\" allowfullscreen></iframe>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/p4vKJJlNTKA\" frameborder=\"0\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa7626f-3af1-4d7c-8601-4dce2708a4d6",
   "metadata": {},
   "source": [
    "## **Install required software**\n",
    "\n",
    "Four main tools are required for for Submodule 1; fastq-dump, fastqc, fastp, and spades.  We will install these tools using __[Conda](https://docs.conda.io/en/latest/)__, an open-source package management system and environment management tool. Conda helps to easily install, update, and manage software dependencies across different platforms.\n",
    "\n",
    "We will install these tools with specific versions, this ensures consistent behavior and reproducibility of results across different environments and systems.\n",
    "\n",
    "### List of software\n",
    "\n",
    "| **Tool**       | **Description**                                                                                                                                                           |\n",
    "|:---------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| **sra-tools**      | fastq-dump is used to retrieve publicly available data from the NCBI Sequence Read Archive (SRA).                                                 |\n",
    "| **fastqc**        | Used to assess the quality of raw sequencing data by generating visual reports on metrics like read length, GC content, and sequence duplication.|\n",
    "| **fastp**   | Used for quality control and pre-processing of FASTQ files, including read trimming, adapter removal, and filtering out low-quality reads.                          |\n",
    "| **spades**      | Used for genome assembly, creating contigs from high-quality, pre-processed sequencing reads.                  |  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f06ed5a-fa8c-464d-b60b-ac9307e4c1ed",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"> <b>Attention:</b> The code below is written in Bash. The <code>%%bash</code> at the beginning of the cell instructs the Jupyter notebook to execute the code as a Bash script when the cell is run. This is the same code you would run from a standard command-line terminal</a>. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc1aadb-f7c1-4e1e-a7e4-77cc6e60141d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: To execute the code you can either press **ctrl-ENTER** when the cell is selected or by pressing the play button at the top of the notebook</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a2fa5230-d4ee-47c1-90be-1cebea67de9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\n",
      "Displaying versions:\u001b[0m\n",
      "\n",
      "fastq-dump : 3.1.1\n",
      "\n",
      "FastQC v0.11.8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "fastp 0.23.4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SPAdes genome assembler v4.0.0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Install tools using mamba (a conda alternative) with specified versions\n",
    "\n",
    "mamba install --channel bioconda \\\n",
    "    python=3.9 \\\n",
    "    sra-tools=3.1.1 \\\n",
    "    fastqc=0.11.8 \\\n",
    "    fastp=0.23.4 \\\n",
    "    spades=4.0.0 \\\n",
    "    -y > /dev/null 2>&1\n",
    "\n",
    "\n",
    "echo -e \"\\033[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\\nDisplaying versions:\\033[0m\"\n",
    "\n",
    "# Confirm installation by checking versions\n",
    "fastq-dump --version\n",
    "fastqc --version\n",
    "fastp --version\n",
    "spades.py --version\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b09b30f-d22a-415f-8dd2-b3338fd07a9f",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>\\</code> to break a long command into multiple lines</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75252e5c-4e8d-4d97-987b-3111db2008d2",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>/dev/null 2>&1</code> to write standard error and standard output to <i>/dev/null</i>, effectively discarding them so your screen doesn't get flooded with text.</div></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6bd27d-90a9-4fac-98f5-ce003a5c097c",
   "metadata": {},
   "source": [
    "## **Download Starting data**\n",
    "The data used for this module is described in a manuscript comparing phenotypic and whole-genome sequencing derived AMR profiles (Painset et al. 2020) and is available under the SRA accession SRR10056829. Genomic DNA was extracted, fragmented and tafed for muktiplexing with **Nextera XT DNA Sample Preperation Kits**. Paired-end data was produced on an **Illumina HiSeq 2500** platform.\n",
    "\n",
    "The manuscript includes sequencing read datasets for 528 isolates of *Campylobacter* spp. (452 *C. jejuni* and 76 *C. coli*) from human (494), food (21) and environmental (2) sources. We will start by walking through the process on just a single dataset. We will later expand this to include a larger set of this data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "prefetch -v $accession\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession\n",
    "\n",
    "echo Process Complete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2be5535-b923-4ab5-91d8-3e8dca3c833b",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: Assigning certain values to variables (like the accession above) allows you to easily adjust input to a set of commands and reduces the amount of typing for long or complicated file names</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1b1056-ae52-4290-a2a0-615d06512281",
   "metadata": {},
   "source": [
    "### FASTQ file format\n",
    "The data we just downloaded are in **FASTQ** file format. Let's examine this data and file format.\n",
    "\n",
    "Let's run three sections of code to learn about the FASTQ file format. Please also see a detailed description of FASTQ files in the NIGMS module [Fundementals of Bionformatics](https://github.com/NIGMS/Fundamentals-of-Bioinformatics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ea85ebdd-d656-468a-95e0-89426db38912",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR10056829_1.fastq.gz\n",
      "SRR10056829_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# If the above code worked you should have two FASTQ files in a directory called raw-reads\n",
    "ls raw-reads/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e89bc5-b10d-4375-b05a-4fe6390528fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class=\"alert alert-block alert-warning\"> <b>Attention:</b> Notice the the files above have the .gz extension. FASTQ data is commonly gzip compressed to save sapce. Most bioinformatic tools accept or require gzipped FASTQs. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7ccb32ad-c885-4644-9d7f-fe9d2f34417d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------- FASTQ READ ---------\n",
      "@SRR10056829.1 1 length=100\n",
      "NCCCCAAGGAATAACATTCATAACCCCCATAGCCGAAGTAATAATCAATAACATTGAAGTCTTTCTTATACCTAAACGCTCATAAATAGGTAATAATGCA\n",
      "+SRR10056829.1 1 length=100\n",
      "#0<FFFFFFFFFFIIFFIIIIIIFIIIIIIIIIFIFFIIIIIIIIFFFIIIIIIIIFIIIIIIIIIIFFFFFFFBFFFFFFFFFFFFFBFBBFFFFFFFF\n",
      "-----------------------------\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# define variable names for reads\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# view the first four lines\n",
    "echo \"-------- FASTQ READ ---------\"\n",
    "zcat $forward | head -n 4\n",
    "echo \"-----------------------------\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2ae8ac-0fb2-4123-afec-0f37f76058fc",
   "metadata": {},
   "source": [
    "The above code should have displayed the first four lines of the FASTQ file. **Each sequencing read in a FASTQ formatted file is made up of four lines**. \n",
    "\n",
    "The length of the sequencing reads can the length depending on the type of sequencing instrument and flow cell used, but they should be consistent across the raw data. In our example datset the reads are 100 bps in length. Illumina sequencers produce reads with length lengths randing between 50 and 300 bps. Run the following code to print the length of the reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5120eb17-44a0-44a5-86ab-68b2b7140825",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "zcat raw-reads/SRR10056829_1.fastq.gz | head -n 2 | tail -n 1 | wc -c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c457802-3e99-4b3a-bb2b-ce19dcf7794c",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: The number above is one higher than the actual read length, this is because of the newline charcater at the end of each line that counted with the wc command</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba7784-527e-461b-a3de-a72b79fc52c5",
   "metadata": {},
   "source": [
    "## Do we have enough data? \n",
    "Most bacterial genomes are composed of a single, circular chromosome, although some bacteria may have linear chromosomes or multiple chromosomes. Plasmids (small circular DNA molecules) often coexist with the main chromosome and carry additional genes. Compared to eukaryotes, a typical bacterial genome is generally small, ranging from about 0.5 to 10 megabases (Mb) in size. For example, *Escherichia coli*, has a genome size of approximately 4.6 Mb. \n",
    "\n",
    "| Organism                        | Genome Size (Mb) |\n",
    "|----------------------------------|------------------|\n",
    "| *Escherichia coli* (Bacteria)    | 4.6              |\n",
    "| *Streptomyces coelicolor* (Bacteria) | 8.7          |\n",
    "| *Mycoplasma genitalium* (Bacteria)   | 0.58         |\n",
    "| *Homo sapiens* (Human)           | 3,200            |\n",
    "| *Drosophila melanogaster* (Fruit Fly) | 139          |\n",
    "| *Arabidopsis thaliana* (Thale Cress) | 135           |\n",
    "| *Saccharomyces cerevisiae* (Yeast)   | 12.1         |\n",
    "| *Caenorhabditis elegans* (Nematode)  | 100           |\n",
    "| *Mus musculus* (Mouse)           | 2,700            |\n",
    "| *Zea mays* (Maize)               | 2,300            |\n",
    "\n",
    "\n",
    "### What is 'coverage'?\n",
    "**Breadth of coverage** refers to the percentage of the genome that has been sequenced at least once. It tells you how much of the genome has any sequencing data aligned to it, but it doesn’t indicate how many times each base has been sequenced. High breadth of coverage means most or all regions of the genome are represented in the sequencing data.\n",
    "\n",
    "**Depth of coverage**, on the other hand, refers to the average number of times each base in the genome is sequenced, often represented as \"X\" (e.g., 30x depth). Depth is important for accuracy, as higher depth ensures that individual bases are sequenced multiple times, improving confidence in variant calling and reducing sequencing errors.\n",
    "\n",
    "When researchers say *coverage* they usually mean *depth of coverage*.\n",
    "\n",
    "\n",
    "The equation for genome depth of coverage is:\n",
    "\n",
    "$$\n",
    "\\text{Coverage} = \\frac{N \\times L}{G}\n",
    "$$\n",
    "\n",
    "Where:\n",
    "- \\( N \\) = Number of reads\n",
    "- \\( L \\) = Read length\n",
    "- \\( G \\) = Genome size\n",
    "\n",
    "\n",
    "The total number of bases sequenced, the numerator in the above equation, is also called the total **throughput** for the sample.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d45fee-1511-49e5-9621-4ac218901c6b",
   "metadata": {},
   "source": [
    "### How do we determine our total throughput and coverage?\n",
    "\n",
    "Lets do the math based on the data we have so far, we know the read length, and we can use a best guess at the genome size as 5 Mb for now. SO lets count our number of reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1874fbae-7b77-49de-802f-f620729479b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6365120\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "zcat raw-reads/*_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7f8ae91a-af8f-42fd-b26a-7a8385e01d70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of base pairs sequenced:\n",
      "318256000\n",
      "The estimated depth of coverage for this genome dataset:\n",
      "63\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "num_reads=6365120/4\n",
    "read_length=100\n",
    "# adjust this number to the organisms genome size.\n",
    "genome_size=5000000\n",
    "\n",
    "# calculate our total throughput\n",
    "echo \"Total number of base pairs sequenced:\"\n",
    "throughput=$(echo $num_reads \" * \" $read_length \" * 2\" | bc)\n",
    "echo $throughput\n",
    "\n",
    "# calculate our estimated coverage\n",
    "echo \"The estimated depth of coverage for this genome dataset:\"\n",
    "echo $throughput \" / \" $genome_size | bc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414ab82a-3f91-43d3-8c07-025f8a3379b2",
   "metadata": {},
   "source": [
    "## Step 1: Quality Assessment with FASTQC\n",
    "\n",
    "Program: FASTQC  \n",
    "Manual: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/  \n",
    "FASTQC explained:  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
    "\n",
    "\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and determine some basic statistics about the data. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, it merely reads it and writes a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "506c05fd-a403-4eb3-bebb-7d0cfcc51246",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output-fastqc’: File exists\n",
      "OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x0000000689000000, 6291456000, 0) failed; error='Not enough space' (errno=12)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#\n",
      "# There is insufficient memory for the Java Runtime Environment to continue.\n",
      "# Native memory allocation (mmap) failed to map 6291456000 bytes. Error detail: committing reserved memory.\n",
      "# An error report file with more information is saved as:\n",
      "# /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/hs_err_pid18182.log\n",
      "SRR10056829_1_fastqc.html\n",
      "SRR10056829_1_fastqc.zip\n",
      "SRR10056829_2_fastqc.html\n",
      "SRR10056829_2_fastqc.zip\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir output-fastqc\n",
    "\n",
    "# run the program on the two read files (R1 and R2).\n",
    "fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.\n",
    "ls output-fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5bfb2d6a-1a60-4b89-9898-1f5079d95962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1000\"\n",
       "            height=\"550\"\n",
       "            src=\"output-fastqc/SRR10056829_1_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fd5782fefe0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## view HTML report\n",
    "from IPython.display import IFrame\n",
    "IFrame('output-fastqc/SRR10056829_1_fastqc.html', width=1000, height=550)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb6207a-2b47-453f-a4d0-b38bff239dab",
   "metadata": {},
   "source": [
    "## Step 2: Quality and adapter trimming with FASTP\n",
    "\n",
    "Program: fastp - an ultra-fast all-in-one FASTQ preprocessor\n",
    "Citation: *Chen, S. (2023). Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*\n",
    "Manual: https://github.com/OpenGene/fastp\n",
    "Conda: https://anaconda.org/bioconda/fastp\n",
    "\n",
    "Description:\n",
    "Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "18047f7b-4838-4133-bf51-aa1d139caf2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 1591280\n",
      "total bases: 159128000\n",
      "Q20 bases: 157594869(99.0365%)\n",
      "Q30 bases: 152891677(96.0809%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 1591280\n",
      "total bases: 159128000\n",
      "Q20 bases: 155820667(97.9216%)\n",
      "Q30 bases: 150621528(94.6543%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1568979\n",
      "total bases: 154244678\n",
      "Q20 bases: 153095395(99.2549%)\n",
      "Q30 bases: 148904585(96.5379%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 1568979\n",
      "total bases: 154244678\n",
      "Q20 bases: 152644379(98.9625%)\n",
      "Q30 bases: 147803821(95.8243%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 3137958\n",
      "reads failed due to low quality: 43808\n",
      "reads failed due to too many N: 794\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 214702\n",
      "bases trimmed due to adapters: 5320114\n",
      "\n",
      "Duplication rate: 2.70292%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 168\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 raw-reads/SRR10056829_1.fastq.gz --in2 raw-reads/SRR10056829_2.fastq.gz --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz \n",
      "fastp v0.23.4, time used: 34 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimmed_1.fastq.gz\n",
      "trimmed_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# make a directory to store the output\n",
    "mkdir output-fastp\n",
    "\n",
    "# define input variables\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# run the program\n",
    "fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# the resulting folder should contain new trimmed read files\n",
    "ls output-fastp/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa89305-5b2d-4614-9cd2-c144224187bc",
   "metadata": {},
   "source": [
    "## Step 3: Genome Assembly\n",
    "\n",
    "- Program: **SPAdes - St. Petersburg genome assembler**\n",
    "- Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*  \n",
    "- Manual: https://ablab.github.io/spades/  \n",
    "\n",
    "SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. However, it might not be suitable for large genomes projects. SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single-end reads."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40624982-3111-483a-8b9d-3a277339e589",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: I recommend running the code below, then watching the following video tutorial or reading more while it runs.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0a21d5-54e6-4cee-97ca-662d2d0c0885",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "== Warning ==  No assembly mode was specified! If you intend to assemble high-coverage multi-cell/isolate data, use '--isolate' option.\n",
      "\n",
      "\n",
      "Command line: /home/ec2-user/anaconda3/envs/python3/bin/spades.py\t-1\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\t-2\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\t-o\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.0.0\n",
      "  Python version: 3.9.20\n",
      "  OS: Linux-5.10.224-212.876.amzn2.x86_64-x86_64-with-glibc2.26\n",
      "\n",
      "Output dir: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Standard mode\n",
      "  For multi-cell/isolate data we recommend to use '--isolate' option; for single-cell MDA data use '--sc'; for metagenomic data use '--meta'; for RNA-Seq use '--rna'.\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz']\n",
      "      right reads: ['/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 3\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/spades.log\n",
      "\n",
      "/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz: max reads length: 100\n",
      "/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz: max reads length: 100\n",
      "\n",
      "Reads length: 100\n",
      "\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/spades-hammer /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.003     1M / 12M   INFO    General                 (main.cpp                  :  76)   Starting BayesHammer, built from N/A, git revision N/A\n",
      "  0:00:00.007     1M / 12M   INFO    General                 (main.cpp                  :  77)   Loading config from \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/configs/config.info\"\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  :  79)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 3 Gb\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  :  87)   Trying to determine PHRED offset\n",
      "  0:00:00.012     1M / 12M   INFO    General                 (main.cpp                  :  93)   Determined value is 33\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (hammer_tools.cpp          :  40)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (main.cpp                  : 114)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.022     1M / 12M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 16 files using 2 threads. This might take a while.\n",
      "  0:00:00.023     1M / 12M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:00.023     1M / 12M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499973 Gb\n",
      "  0:00:00.023     1M / 12M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 4194073\n",
      "  0:00:00.024  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\"\n",
      "  0:00:10.472  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 427146 reads\n",
      "  0:00:20.191  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 854764 reads\n",
      "  0:00:29.969  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1282054 reads\n",
      "  0:00:36.464  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1568979 reads\n",
      "  0:00:36.466  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\"\n",
      "  0:00:46.175  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1997005 reads\n",
      "  0:00:55.799  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2425011 reads\n",
      "  0:01:08.949  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2852691 reads\n",
      "  0:01:15.378  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 113)   Total 3137958 reads processed\n",
      "  0:01:15.457     1M / 582M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:01:16.065     1M / 582M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 10384610 kmers in total.\n",
      "  0:01:16.070     1M / 582M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:01:16.937     8M / 582M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 10384610 kmers, 7512272 bytes occupied (5.78723 bits per kmer).\n",
      "  0:01:16.941     8M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 355)   Arranging kmers in hash map order\n",
      "  0:01:17.517   168M / 582M  INFO    General                 (main.cpp                  : 149)   Clustering Hamming graph.\n",
      "  0:02:10.541   168M / 582M  INFO    General                 (main.cpp                  : 156)   Extracting clusters:\n",
      "  0:02:10.542   168M / 582M  INFO    General                 (concurrent_dsu.cpp        :  19)   Connecting to root\n",
      "  0:02:10.793   168M / 582M  INFO    General                 (concurrent_dsu.cpp        :  35)   Calculating counts\n",
      "  0:02:14.399   304M / 582M  INFO    General                 (concurrent_dsu.cpp        :  64)   Writing down entries\n",
      "  0:02:18.860   168M / 582M  INFO    General                 (main.cpp                  : 168)   Clustering done. Total clusters: 3886912\n",
      "  0:02:18.867    88M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 372)   Collecting K-mer information, this takes a while.\n",
      "  0:02:19.089   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 378)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\"\n",
      "  0:06:52.849   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 378)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\"\n",
      "  0:11:40.117   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 385)   Collection done, postprocessing.\n",
      "  0:11:40.187   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 398)   There are 10384610 kmers in total. Among them 6578842 (63.3518%) are singletons.\n",
      "  0:11:40.188   328M / 582M  INFO    General                 (main.cpp                  : 174)   Subclustering Hamming graph\n",
      "  0:11:48.812   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 637)   Subclustering done. Total 1 non-read kmers were generated.\n",
      "  0:11:48.818   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 638)   Subclustering statistics:\n",
      "  0:11:48.824   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 639)     Total singleton hamming clusters: 1346992. Among them 989104 (73.4306%) are good\n",
      "  0:11:48.855   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 640)     Total singleton subclusters: 16033. Among them 15946 (99.4574%) are good\n",
      "  0:11:48.859   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 641)     Total non-singleton subcluster centers: 2564455. Among them 2132949 (83.1736%) are good\n",
      "  0:11:48.869   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 642)     Average size of non-trivial subcluster: 3.52419 kmers\n",
      "  0:11:48.872   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 643)     Average number of sub-clusters per non-singleton cluster: 1.01597\n",
      "  0:11:48.873   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 644)     Total solid k-mers: 3137999\n",
      "  0:11:48.875   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 645)     Substitution probabilities: (     0.948291    0.0146903    0.0218017    0.0152168 )\n",
      "(    0.0124587     0.959161    0.0113567    0.0170241 )\n",
      "(    0.0170012    0.0113477     0.959175    0.0124756 )\n",
      "(    0.0152029    0.0217921    0.0146812     0.948324 )\n",
      "  0:11:48.915   328M / 582M  INFO    General                 (main.cpp                  : 179)   Finished clustering.\n",
      "  0:11:48.917   328M / 582M  INFO    General                 (main.cpp                  : 198)   Starting solid k-mers expansion in 2 threads.\n",
      "  0:14:31.927   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 0 produced 35808 new k-mers.\n",
      "  0:16:55.498   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 1 produced 1356 new k-mers.\n",
      "  0:19:21.112   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 2 produced 11 new k-mers.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "## Assemble trimmed reads using the SPAdes assembler.\n",
    "\n",
    "# define input\n",
    "forward=output-fastp/trimmed_1.fastq.gz\n",
    "reverse=output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# assemble reads with SPAdes\n",
    "spades.py -1 $forward -2 $reverse -o output-spades --threads 24"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eec824dc-de4e-4456-89ac-955074826565",
   "metadata": {},
   "source": [
    "## Video Lesson: How genome assembly of short reads works.\n",
    "\n",
    "The following video describes the process in detail. Image credit. https://homolog.us/Tutorials/book4/p2.1.html\n",
    "\n",
    "[![Sequencing by Synthesis](images/debruijn.png)](https://www.youtube.com/watch?v=ZmF6QROPlTU)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa1800eb-3f18-4590-aa17-e834e78b2b03",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Video Recap\n",
    "\n",
    "### De Bruijn Graphs in Genome Assembly\n",
    "\n",
    "A **de Bruijn graph** is a structure used in genome assembly to represent sequences using **k-mers**. Here’s how it works:\n",
    "\n",
    "1. **K-mers**: These are short sequences of length *k* derived from a longer DNA sequence. For example, from the sequence \"AGCT\", the 2-mers are \"AG\", \"GC\", and \"CT\".\n",
    "\n",
    "2. **Graph Construction**: Each k-mer becomes a vertex in the graph. An edge is drawn from k-mer **A** to k-mer **B** if the last *k-1* nucleotides of **A** match the first *k-1* nucleotides of **B**. This creates a directed graph where paths through the vertices represent potential sequences in the genome.\n",
    "\n",
    "3. **Contigs and Scaffolds**:\n",
    "   - **Contigs**: These are contiguous sequences formed by traversing paths in the graph.\n",
    "   - **Scaffolds**: Longer constructs that may combine multiple contigs, providing additional structure to the genome assembly.\n",
    "\n",
    "De Bruijn graphs enable efficient handling of sequencing data, especially for complex genomes with repetitive sequences.\n",
    "  \n",
    "  \n",
    "### Key challenges for genome assembly:\n",
    "**<u>Intrinsic Challenges:</u>**\n",
    "1. **Heterozygosity:** The alleles of a gene are not he same, yet we force them into a single consensus sequence.\n",
    "2. **Paralogy vs. Alleleism**: Genes come from other genes by a process of duplication. This results in two or more similar genes in an organism. There are two alleles in a diploid organism that are very similiar. How do you tell a duplicated gene from alleles of a gene?\n",
    "3. **Sequence complexity**: Simple sequence repeats (SSR), large scale repeats like transposable elements (TEs).\n",
    "\n",
    "**<u>Extrinsic Challenges:</u>**\n",
    "1. **Quality of DNA sequences (sequencing errors):** Each sequencing technology has specific patterns of error.\n",
    "2. **Length of DNA sequence reads**: Shorter reads are less likely to be unique or to include unique K-mers (see below).\n",
    "3. **Coverage**: Depth of coverage is a random process at best. Consequently some regions of the genome will have low levels of coverage.\n",
    "4. **Memory intensive**: Inherently requires large amounts of RAM for assebly and storage of input and output.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e738484-5225-4af6-8a36-da56c9b31bd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c03cce-0795-4135-bb0a-45910788588e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a copy of the genome sequence in a new directory. This data will be used for Submodule 2.\n",
    "\n",
    "mkdir assembled-genome\n",
    "cp output-spades/contigs.fasta assembled-genome/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
