{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3054864-6f9a-44d5-b3a3-ff29ccde5e7c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Submodule 1: Introduction to genome sequencing and assembly\n",
    "--------\n",
    "Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. \n",
    "\n",
    "With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. A critical component of genomics is genome assembly, which involves reconstructing the genome of an organism from short DNA sequences generated by sequencing technologies. Genome assembly is a foundational step that allows researchers to generate a complete picture of the genetic material in a given organism. Assessing the quality and completeness of a genome assembly ensures that it accurately reflects the original genome, providing a reliable foundation for further analysis.\n",
    "\n",
    "Comparative genomics builds on genome assembly by comparing the genomes of different species or individuals. This field focuses on identifying similarities and differences in DNA sequences to understand evolutionary relationships, discover conserved genetic elements, and reveal the genetic underpinnings of adaptation and diversity. Comparative genomics can uncover how species evolve and adapt over time and identify genes associated with specific traits or diseases.\n",
    "\n",
    "In this tutorial, we will explore the key concepts and methodologies in genomics, including how to assemble a genome, assess its quality, and perform comparative analyses. Through hands-on examples, you will learn how to analyze genomic data, interpret results, and apply various computational tools to gain deeper insights into the genetic information that defines life.\n",
    "\n",
    " \n",
    "\n",
    "### Learning Objectives:\n",
    "\n",
    "+ Understand how high throughput sequencing data is generated. \n",
    "\n",
    "+ Develop an understanding of core bioinformatic input/output formats as it relates to comparative genomics. \n",
    "\n",
    "+ Acquire the skills to assemble raw sequencing reads into a draft genome, assess the quality of the genome assembly, and annotate the genome sequence. \n",
    "\n",
    "+ Learn to perform comparative genomic analyses to identify similarities and differences across genomes, run phylogenomic analyses, construct pangenomes to capture genetic diversity, and apply these techniques to address biological questions and hypotheses. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4c2a457-80a1-45f9-9bb8-7cf38ff99281",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.1 - Generation of sequencing data\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "<img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/fragmentation3.png?raw=true\" width=\"800\">\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07881f7e-9a58-45d2-9bef-ab24929d8227",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.2 - How next-generation sequencing works\n",
    "\n",
    "[![sequencing by synthesis](img/youtube-video-sequencing.PNG)](https://www.youtube.com/watch?v=p4vKJJlNTKA&t=9s \"Sequencing\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d01dbeac-d525-4528-a3c2-ac527faa8108",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.2.1 Cost of generating sequencing data\n",
    "\n",
    "CItation: Wetterstrand KA. DNA Sequencing Costs: Data from the NHGRI Genome Sequencing Program (GSP) Available at: www.genome.gov/sequencingcostsdata. Accessed [10/03/2024]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b158a53b-3182-4a63-9827-0d20c375aeee",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "### 1.2.2 Generate plot showing the cost of sequencing genomes over time\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.dates as mdates\n",
    "\n",
    "# Load the CSV file\n",
    "data_table = \"data/Sequencing_Cost_Data_Table_May2022.csv\"\n",
    "df = pd.read_csv(data_table)\n",
    "\n",
    "# Convert the 'Date' column to datetime format using YYYY-MM\n",
    "df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m')\n",
    "\n",
    "# Clean up 'Cost per Genome' column by removing $ and commas, and converting to numeric\n",
    "df['Cost per Genome'] = df['Cost per Genome'].replace({'\\$': '', ',': ''}, regex=True).astype(float)\n",
    "\n",
    "# Plot the data\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "# X-axis: Date, Y-axis: Cost per Genome in log scale\n",
    "plt.plot(df['Date'], df['Cost per Genome'], marker='o', linestyle='-', color='b')\n",
    "\n",
    "# Set log scale for y-axis\n",
    "plt.yscale('log')\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Year', fontsize=14, fontweight='bold')\n",
    "plt.ylabel('Cost per Genome (Log scale)', fontsize=12, fontweight='bold')\n",
    "plt.title('Cost per Genome Over Time', fontsize=14, fontweight='bold')\n",
    "\n",
    "# Format the x-axis for better date presentation\n",
    "import matplotlib.dates as mdates\n",
    "\n",
    "# Set major ticks to display years at specific intervals\n",
    "plt.gca().xaxis.set_major_locator(mdates.YearLocator(3))  # Adjust the number here for the interval\n",
    "plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%Y'))\n",
    "\n",
    "# Rotate and format x-ticks\n",
    "plt.xticks(rotation=45, fontsize=8, fontweight='bold')\n",
    "\n",
    "\n",
    "# Set the y-axis limits from $100 to $100,000\n",
    "# plt.ylim(100, 0000)\n",
    "plt.yticks(fontsize=12, fontweight='bold')\n",
    "\n",
    "# Display the plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cost_per_genome_plot.png\", dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7b4226-51c1-4e68-9900-38055832b945",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eca78aeb-2a38-42d8-b163-a4ab47acbbca",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.3 - Comparison of sequencing instruments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6bd27d-90a9-4fac-98f5-ce003a5c097c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 1.4 - Starting data\n",
    "\n",
    "Each of these samples represent the genome of a unique and novel microbe that has not been seen before (except by me). \n",
    "\n",
    "The data utilized in this tutorial was generated by Illumina HiSeq 2500, is paired-end and each read is 250 bps n length. \n",
    "\n",
    "<p align=\"center\">\n",
    "    <img src=\"images/jupyterNotebook_annotated.png\" alt=\"jupyterNotebook\" width=\"50%\"/>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b43a7cc-7801-4f9e-b18c-9d5996a33166",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 1.4.1 - Install sra-tools to access public data available on the NCBI Sequence Read Archive (SRA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f9db7442-a02e-46cc-8236-28e5284915a0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bioconda/linux-64                                           Using cache\n",
      "bioconda/noarch                                             Using cache\n",
      "conda-forge/linux-64                                        Using cache\n",
      "conda-forge/noarch                                          Using cache\n",
      "nvidia/linux-64                                             Using cache\n",
      "nvidia/noarch                                               Using cache\n",
      "pytorch/linux-64                                            Using cache\n",
      "pytorch/noarch                                              Using cache\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  All requested packages already installed\n",
      "\n",
      "\n",
      "Looking for: ['python=3.9', 'sra-tools']\n",
      "\n",
      "Installation of sra-tools complete\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Install fastq-dump using mamba ( a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 sra-tools -y > /dev/null 2>&1\n",
    "\n",
    "# hide standard error and standard out > /dev/null 2>&1\n",
    "# hide standard error only 2> /dev/null\n",
    "\n",
    "echo \"Installation of sra-tools complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘wgs-workflow’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-30T14:49:15 prefetch.3.1.1: 1) Resolving 'SRR10056829'...\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: 'tools/ascp/disabled': not found in configuration\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking 'ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: 'ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking 'ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: 'ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/usr/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/usr/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/usr/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/usr/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/opt/aspera/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/opt/aspera/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/opt/aspera/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/opt/aspera/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/home/ec2-user/.aspera/connect/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/home/ec2-user/.aspera/connect/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Checking '/home/ec2-user/.aspera/connect/bin/ascp'\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: '/home/ec2-user/.aspera/connect/bin/ascp': not found\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2024-09-30T14:49:15 prefetch.3.1.1: 1) Downloading 'SRR10056829'...\n",
      "2024-09-30T14:49:15 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-30T14:49:15 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-30T14:49:20 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-30T14:49:21 prefetch.3.1.1:  'SRR10056829' is valid: 130149382 bytes were streamed from 130139545\n",
      "2024-09-30T14:49:21 prefetch.3.1.1: 1) 'SRR10056829' was downloaded successfully\n",
      "2024-09-30T14:49:21 prefetch.3.1.1: 'SRR10056829' has 0 unresolved dependencies\n",
      "Read 1591280 spots for SRR10056829/SRR10056829.sra\n",
      "Written 1591280 spots for SRR10056829/SRR10056829.sra\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# setup working directory and move into it\n",
    "mkdir wgs-workflow\n",
    "cd wgs-workflow\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "prefetch -v $accession\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1b1056-ae52-4290-a2a0-615d06512281",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 1.4.2 - FASTQ file format\n",
    "These data are in **FASTQ** format (see below).\n",
    "\n",
    "[Link explaining the 'Read Name Format'](http://support.illumina.com/content/dam/illumina-support/help/BaseSpaceHelp_v2/Content/Vault/Informatics/Sequencing_Analysis/BS/swSEQ_mBS_FASTQFiles.htm): SampleName_Barcode_LaneNumber_001.fastq.gz\n",
    "\n",
    "\n",
    "Important note: In the above command I use the \"\\*\" character to view the Sample directory, I would normally just type out the entire path using tab complete (which is what you should do). This wildcard will match any string of characters. I use this because everyone will have a different Sample name. To make this tutorial as general as possible I need to use these wildcards throughout the tutorial. In addition I may use Sample_X instead of Sample_\\*. In these cases be sure to type out your complete sample name!, the wildcards probably won't work "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0e7622a4-dd64-497e-96ea-8d9b83d85281",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR10056829_1.fastq.gz\n",
      "SRR10056829_2.fastq.gz\n",
      "@SRR10056829.1 1 length=100\n",
      "NCCCCAAGGAATAACATTCATAACCCCCATAGCCGAAGTAATAATCAATAACATTGAAGTCTTTCTTATACCTAAACGCTCATAAATAGGTAATAATGCA\n",
      "+SRR10056829.1 1 length=100\n",
      "#0<FFFFFFFFFFIIFFIIIIIIFIIIIIIIIIFIFFIIIIIIIIFFFIIIIIIIIFIIIIIIIIIIFFFFFFFBFFFFFFFFFFFFFBFBBFFFFFFFF\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# If the above code worked you should have two FASTQ files in a directory called raw-reads\n",
    "ls raw-reads/\n",
    "\n",
    "# Pull the raw fastq files you need from the SRA\n",
    "zcat raw-reads/SRR10056829_1.fastq.gz | head -n 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e803f847-b1ba-48a5-8193-a4d3eeba75bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.5 - Quality assessment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba7784-527e-461b-a3de-a72b79fc52c5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 1.5.1 Do we have enough coverage? How do we determine our total throughput?\n",
    "\n",
    "INSERT FLASH CARD WITH QUESTIONS ABOUT READ LENGTH AND COVERAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1874fbae-7b77-49de-802f-f620729479b9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6365120\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "zcat raw-reads/*_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab968490-19b8-4bf7-9582-5973da86492f",
   "metadata": {
    "tags": []
   },
   "source": [
    "Whats our total bp of data? This is what we call our sequencing throughput. We multiple the number of reads by the read length (ours is 250) and by 2 because it is paired-end data.\n",
    "(Read length x 2(paired-end) x Number of reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7f8ae91a-af8f-42fd-b26a-7a8385e01d70",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of base pairs sequenced\n",
      "3182560000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "num_reads=6365120\n",
    "read_length=250\n",
    "genome_size=5000000\n",
    "\n",
    "# calculate our total throughput\n",
    "echo \"Total number of base pairs sequenced\"\n",
    "throughput=$(echo $num_reads \" * \" $read_length \" * 2\" | bc)\n",
    "echo $throughput\n",
    "\n",
    "# calculate our estimated coverage\n",
    "echo \"Total coverage for genome dataset\"\n",
    "est_coverage=$(echo $throughput \" / \" \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414ab82a-3f91-43d3-8c07-025f8a3379b2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 1.5.2 Quality Assessment with FASTQC\n",
    "\n",
    "Program: FASTQC\n",
    "Manual: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
    "FASTQC explained:\n",
    "Conda:\n",
    "\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and do the hard work for us. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, as with the majority of the assessment tools, it merely reads it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3e9d122c-320d-458e-a61b-02b6222f65e8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/anaconda3/envs/python3\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - python=3.9\n",
      "   - fastqc\n",
      "   - ca-certificates\n",
      "   - certifi\n",
      "   - openssl\n",
      "\n",
      "\n",
      "  Package       Version  Build       Channel           Size\n",
      "─────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "─────────────────────────────────────────────────────────────\n",
      "\n",
      "  + xorg-libxt    1.3.0  hd590300_1  conda-forge     Cached\n",
      "  + openjdk      22.0.1  h8651b0f_1  conda-forge      182MB\n",
      "  + fastqc       0.12.1  hdfd78af_0  bioconda          12MB\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 3 packages\n",
      "\n",
      "  Total download: 193MB\n",
      "\n",
      "─────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\n",
      "Looking for: ['python=3.9', 'fastqc']\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "Installation of fastqc complete\n",
      "FastQC v0.12.1\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Install fastqc using mamba ( a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 fastqc -y > /dev/null 2>&1\n",
    "\n",
    "# > /dev/null 2>&1 : hides standard error and standard out \n",
    "# 2> /dev/null : hides standard error only \n",
    "\n",
    "echo \"Installation of fastqc complete\"\n",
    "fastqc --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "506c05fd-a403-4eb3-bebb-7d0cfcc51246",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output-fastqc’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of SRR10056829_1.fastq.gz\n",
      "Started analysis of SRR10056829_2.fastq.gz\n",
      "Approx 5% complete for SRR10056829_2.fastq.gz\n",
      "Approx 5% complete for SRR10056829_1.fastq.gz\n",
      "Approx 10% complete for SRR10056829_2.fastq.gz\n",
      "Approx 10% complete for SRR10056829_1.fastq.gz\n",
      "Approx 15% complete for SRR10056829_2.fastq.gz\n",
      "Approx 15% complete for SRR10056829_1.fastq.gz\n",
      "Approx 20% complete for SRR10056829_2.fastq.gz\n",
      "Approx 20% complete for SRR10056829_1.fastq.gz\n",
      "Approx 25% complete for SRR10056829_1.fastq.gz\n",
      "Approx 25% complete for SRR10056829_2.fastq.gz\n",
      "Approx 30% complete for SRR10056829_1.fastq.gz\n",
      "Approx 30% complete for SRR10056829_2.fastq.gz\n",
      "Approx 35% complete for SRR10056829_1.fastq.gz\n",
      "Approx 35% complete for SRR10056829_2.fastq.gz\n",
      "Approx 40% complete for SRR10056829_1.fastq.gz\n",
      "Approx 40% complete for SRR10056829_2.fastq.gz\n",
      "Approx 45% complete for SRR10056829_1.fastq.gz\n",
      "Approx 45% complete for SRR10056829_2.fastq.gz\n",
      "Approx 50% complete for SRR10056829_1.fastq.gz\n",
      "Approx 50% complete for SRR10056829_2.fastq.gz\n",
      "Approx 55% complete for SRR10056829_1.fastq.gz\n",
      "Approx 55% complete for SRR10056829_2.fastq.gz\n",
      "Approx 60% complete for SRR10056829_1.fastq.gz\n",
      "Approx 60% complete for SRR10056829_2.fastq.gz\n",
      "Approx 65% complete for SRR10056829_1.fastq.gz\n",
      "Approx 65% complete for SRR10056829_2.fastq.gz\n",
      "Approx 70% complete for SRR10056829_1.fastq.gz\n",
      "Approx 70% complete for SRR10056829_2.fastq.gz\n",
      "Approx 75% complete for SRR10056829_1.fastq.gz\n",
      "Approx 75% complete for SRR10056829_2.fastq.gz\n",
      "Approx 80% complete for SRR10056829_1.fastq.gz\n",
      "Approx 80% complete for SRR10056829_2.fastq.gz\n",
      "Approx 85% complete for SRR10056829_1.fastq.gz\n",
      "Approx 85% complete for SRR10056829_2.fastq.gz\n",
      "Approx 90% complete for SRR10056829_1.fastq.gz\n",
      "Approx 90% complete for SRR10056829_2.fastq.gz\n",
      "Approx 95% complete for SRR10056829_1.fastq.gz\n",
      "Approx 95% complete for SRR10056829_2.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for SRR10056829_1.fastq.gz\n",
      "Analysis complete for SRR10056829_2.fastq.gz\n",
      "SRR10056829_1_fastqc.html\n",
      "SRR10056829_1_fastqc.zip\n",
      "SRR10056829_2_fastqc.html\n",
      "SRR10056829_2_fastqc.zip\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir output-fastqc\n",
    "\n",
    "# run the program on the two read files (R1 and R2).\n",
    "fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.\n",
    "ls output-fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfb2d6a-1a60-4b89-9898-1f5079d95962",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### view HTML report\n",
    "from IPython.display import HTML\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb6207a-2b47-453f-a4d0-b38bff239dab",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 1.6 - Quality and adapter trimming with FASTP\n",
    "\n",
    "Program: fastp - an ultra-fast all-in-one FASTQ preprocessor\n",
    "Citation: *Chen, S. (2023). Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*\n",
    "Manual: https://github.com/OpenGene/fastp\n",
    "Conda: https://anaconda.org/bioconda/fastp\n",
    "\n",
    "Description:\n",
    "Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0df2498-4184-4403-b1dc-9cc93b98f955",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Install fastp using mamba (a conda alternative)\n",
    "mamba install --channel bioconda python=3.9 fastp -y > /dev/null 2>&1\n",
    "\n",
    "# hide standard error and standard out > /dev/null 2>&1\n",
    "# hide standard error only 2> /dev/null\n",
    "\n",
    "fastp --version\n",
    "echo \"Installation Complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "18047f7b-4838-4133-bf51-aa1d139caf2b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output-fastp’: File exists\n",
      "Read1 before filtering:\n",
      "total reads: 1591280\n",
      "total bases: 159128000\n",
      "Q20 bases: 157594869(99.0365%)\n",
      "Q30 bases: 152891677(96.0809%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 1591280\n",
      "total bases: 159128000\n",
      "Q20 bases: 155820667(97.9216%)\n",
      "Q30 bases: 150621528(94.6543%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1568979\n",
      "total bases: 154244678\n",
      "Q20 bases: 153095395(99.2549%)\n",
      "Q30 bases: 148904585(96.5379%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 1568979\n",
      "total bases: 154244678\n",
      "Q20 bases: 152644379(98.9625%)\n",
      "Q30 bases: 147803821(95.8243%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 3137958\n",
      "reads failed due to low quality: 43808\n",
      "reads failed due to too many N: 794\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 214702\n",
      "bases trimmed due to adapters: 5320114\n",
      "\n",
      "Duplication rate: 2.70292%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 168\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 raw-reads/SRR10056829_1.fastq.gz --in2 raw-reads/SRR10056829_2.fastq.gz --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz \n",
      "fastp v0.23.4, time used: 35 seconds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trimmed_1.fastq.gz\n",
      "trimmed_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# make a directory to store the output\n",
    "mkdir output-fastp\n",
    "\n",
    "# define input variables\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# run the program\n",
    "fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# the resulting folder should contain new trimmed read files\n",
    "ls output-fastp/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa89305-5b2d-4614-9cd2-c144224187bc",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1.7 - Genome Assembly\n",
    "\n",
    "Program: SPAdes - St. Petersburg genome assembler\n",
    "Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*\n",
    "Manual: https://ablab.github.io/spades/\n",
    "Conda: https://anaconda.org/bioconda/spades\n",
    "\n",
    "\n",
    "Description: \n",
    "SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. However, it might not be suitable for large genomes projects.\n",
    "\n",
    "SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "91bac579-1659-49c4-a60f-478f73ec5d67",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Installation of spades complete\n",
      "SPAdes genome assembler v4.0.0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Install SPAdes\n",
    "mamba install --channel bioconda python=3.9 spades -y > /dev/null 2>&1\n",
    "\n",
    "spades.py --version\n",
    "echo \"Installation Complete\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6681b8a5-b9a7-4924-bb8b-9961d14bd320",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "== Warning ==  No assembly mode was specified! If you intend to assemble high-coverage multi-cell/isolate data, use '--isolate' option.\n",
      "\n",
      "\n",
      "Command line: /home/ec2-user/anaconda3/envs/python3/bin/spades.py\t-1\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\t-2\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\t-o\t/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 4.0.0\n",
      "  Python version: 3.9.20\n",
      "  OS: Linux-5.10.224-212.876.amzn2.x86_64-x86_64-with-glibc2.26\n",
      "\n",
      "Output dir: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Standard mode\n",
      "  For multi-cell/isolate data we recommend to use '--isolate' option; for single-cell MDA data use '--sc'; for metagenomic data use '--meta'; for RNA-Seq use '--rna'.\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz']\n",
      "      right reads: ['/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "  Assembly graph output will use GFA v1.2 format\n",
      "Other parameters:\n",
      "  Dir for temp files: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 3\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/spades.log\n",
      "\n",
      "/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz: max reads length: 100\n",
      "/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz: max reads length: 100\n",
      "\n",
      "Reads length: 100\n",
      "\n",
      "\n",
      "===== Before start started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/spades-hammer /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.001     1M / 12M   INFO    General                 (main.cpp                  :  76)   Starting BayesHammer, built from N/A, git revision N/A\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  :  77)   Loading config from \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/configs/config.info\"\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (main.cpp                  :  79)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
      "  0:00:00.021     1M / 12M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 3 Gb\n",
      "  0:00:00.024     1M / 12M   INFO    General                 (main.cpp                  :  87)   Trying to determine PHRED offset\n",
      "  0:00:00.029     1M / 12M   INFO    General                 (main.cpp                  :  93)   Determined value is 33\n",
      "  0:00:00.032     1M / 12M   INFO    General                 (hammer_tools.cpp          :  40)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.032     1M / 12M   INFO    General                 (main.cpp                  : 114)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.037     1M / 12M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 16 files using 2 threads. This might take a while.\n",
      "  0:00:00.038     1M / 12M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:00.040     1M / 12M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499973 Gb\n",
      "  0:00:00.040     1M / 12M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 4194073\n",
      "  0:00:00.050  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\"\n",
      "  0:00:13.584  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 427146 reads\n",
      "  0:00:26.979  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 854764 reads\n",
      "  0:00:39.971  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1282054 reads\n",
      "  0:00:48.964  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1568979 reads\n",
      "  0:00:48.966  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             :  98)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\"\n",
      "  0:01:06.303  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 1997005 reads\n",
      "  0:01:19.632  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2425011 reads\n",
      "  0:01:34.681  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 108)   Processed 2852693 reads\n",
      "  0:01:43.363  1153M / 1153M INFO   K-mer Splitting          (kmer_data.cpp             : 113)   Total 3137958 reads processed\n",
      "  0:01:43.463     1M / 582M  INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:01:44.282     1M / 582M  INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 10384610 kmers in total.\n",
      "  0:01:44.284     1M / 582M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:01:45.411     8M / 582M  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 10384610 kmers, 7512272 bytes occupied (5.78723 bits per kmer).\n",
      "  0:01:45.412     8M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 355)   Arranging kmers in hash map order\n",
      "  0:01:46.100   168M / 582M  INFO    General                 (main.cpp                  : 149)   Clustering Hamming graph.\n",
      "  0:02:40.949   168M / 582M  INFO    General                 (main.cpp                  : 156)   Extracting clusters:\n",
      "  0:02:40.951   168M / 582M  INFO    General                 (concurrent_dsu.cpp        :  19)   Connecting to root\n",
      "  0:02:41.190   168M / 582M  INFO    General                 (concurrent_dsu.cpp        :  35)   Calculating counts\n",
      "  0:02:43.941   304M / 582M  INFO    General                 (concurrent_dsu.cpp        :  64)   Writing down entries\n",
      "  0:02:48.130   168M / 582M  INFO    General                 (main.cpp                  : 168)   Clustering done. Total clusters: 3886912\n",
      "  0:02:48.138    88M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 372)   Collecting K-mer information, this takes a while.\n",
      "  0:02:48.357   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 378)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\"\n",
      "  0:07:55.856   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 378)   Processing \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\"\n",
      "  0:13:30.331   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 385)   Collection done, postprocessing.\n",
      "  0:13:30.405   328M / 582M  INFO   K-mer Counting           (kmer_data.cpp             : 398)   There are 10384610 kmers in total. Among them 6578842 (63.3518%) are singletons.\n",
      "  0:13:30.405   328M / 582M  INFO    General                 (main.cpp                  : 174)   Subclustering Hamming graph\n",
      "  0:13:41.101   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 637)   Subclustering done. Total 1 non-read kmers were generated.\n",
      "  0:13:41.105   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 638)   Subclustering statistics:\n",
      "  0:13:41.108   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 639)     Total singleton hamming clusters: 1346992. Among them 989104 (73.4306%) are good\n",
      "  0:13:41.112   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 640)     Total singleton subclusters: 16033. Among them 15946 (99.4574%) are good\n",
      "  0:13:41.115   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 641)     Total non-singleton subcluster centers: 2564455. Among them 2132949 (83.1736%) are good\n",
      "  0:13:41.116   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 642)     Average size of non-trivial subcluster: 3.52419 kmers\n",
      "  0:13:41.117   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 643)     Average number of sub-clusters per non-singleton cluster: 1.01597\n",
      "  0:13:41.118   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 644)     Total solid k-mers: 3137999\n",
      "  0:13:41.119   328M / 582M  INFO   Hamming Subclustering    (kmer_cluster.cpp          : 645)     Substitution probabilities: (     0.948291    0.0146903    0.0218017    0.0152168 )\n",
      "(    0.0124587     0.959161    0.0113567    0.0170241 )\n",
      "(    0.0170012    0.0113477     0.959175    0.0124756 )\n",
      "(    0.0152029    0.0217921    0.0146812     0.948324 )\n",
      "  0:13:41.144   328M / 582M  INFO    General                 (main.cpp                  : 179)   Finished clustering.\n",
      "  0:13:41.146   328M / 582M  INFO    General                 (main.cpp                  : 198)   Starting solid k-mers expansion in 2 threads.\n",
      "  0:16:38.581   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 0 produced 35808 new k-mers.\n",
      "  0:19:46.886   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 1 produced 1356 new k-mers.\n",
      "  0:22:46.813   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 2 produced 11 new k-mers.\n",
      "  0:26:07.772   328M / 582M  INFO    General                 (main.cpp                  : 219)   Solid k-mers iteration 3 produced 0 new k-mers.\n",
      "  0:26:07.776   328M / 582M  INFO    General                 (main.cpp                  : 223)   Solid k-mers finalized\n",
      "  0:26:07.782   328M / 582M  INFO    General                 (hammer_tools.cpp          : 226)   Starting read correction in 2 threads.\n",
      "  0:26:07.785   328M / 582M  INFO    General                 (hammer_tools.cpp          : 239)   Correcting pair of reads: \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_1.fastq.gz\" and \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-fastp/trimmed_2.fastq.gz\"\n",
      "  0:26:08.807   469M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 0 of 200000 reads.\n",
      "  0:26:16.984   469M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 0\n",
      "  0:26:17.396   469M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 0\n",
      "  0:26:18.077   474M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 1 of 200000 reads.\n",
      "  0:26:26.652   474M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 1\n",
      "  0:26:27.122   474M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 1\n",
      "  0:26:27.958   474M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 2 of 200000 reads.\n",
      "  0:26:37.038   475M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 2\n",
      "  0:26:37.526   475M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 2\n",
      "  0:26:38.376   475M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 3 of 200000 reads.\n",
      "  0:26:46.341   475M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 3\n",
      "  0:26:46.778   475M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 3\n",
      "  0:26:47.475   475M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 4 of 200000 reads.\n",
      "  0:26:56.472   476M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 4\n",
      "  0:26:56.976   476M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 4\n",
      "  0:26:57.869   476M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 5 of 200000 reads.\n",
      "  0:27:06.349   477M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 5\n",
      "  0:27:06.746   477M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 5\n",
      "  0:27:07.532   477M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 6 of 200000 reads.\n",
      "  0:27:16.941   477M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 6\n",
      "  0:27:17.406   477M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 6\n",
      "  0:27:18.045   477M / 582M  INFO    General                 (hammer_tools.cpp          : 173)   Prepared batch 7 of 168979 reads.\n",
      "  0:27:25.171   478M / 582M  INFO    General                 (hammer_tools.cpp          : 180)   Processed batch 7\n",
      "  0:27:25.718   478M / 582M  INFO    General                 (hammer_tools.cpp          : 190)   Written batch 7\n",
      "  0:27:25.841   328M / 582M  INFO    General                 (hammer_tools.cpp          : 280)   Correction done. Changed 203340 bases in 182154 reads.\n",
      "  0:27:25.849   328M / 582M  INFO    General                 (hammer_tools.cpp          : 281)   Failed to correct 58 bases out of 307924536.\n",
      "  0:27:25.892     1M / 582M  INFO    General                 (main.cpp                  : 256)   Saving corrected dataset description to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/corrected.yaml\n",
      "  0:27:25.897     1M / 582M  INFO    General                 (main.cpp                  : 263)   All done. Exiting.\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== corrected reads compression started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/python3 /home/ec2-user/anaconda3/envs/python3/share/spades/spades_pipeline/scripts/compress_all.py --input_file /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/corrected.yaml --ext_python_modules_home /home/ec2-user/anaconda3/envs/python3/share/spades --max_threads 16 --output_dir /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected --gzip_output\n",
      "\n",
      "== Compressing corrected reads (with pigz)\n",
      "== Files to compress: ['/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/trimmed_1.fastq00.0_0.cor.fastq', '/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/trimmed_2.fastq00.0_0.cor.fastq', '/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/trimmed__unpaired00.0_0.cor.fastq']\n",
      "== Files compression is finished\n",
      "== Dataset yaml file is updated\n",
      "\n",
      "===== corrected reads compression finished. \n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Assembling started. \n",
      "\n",
      "\n",
      "===== K21 started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/spades-core /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K21/configs/config.info\n",
      "\n",
      "  0:00:00.000     1M / 12M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K21/configs/config.info\"\n",
      "  0:00:00.000     1M / 12M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 3 Gb\n",
      "  0:00:00.004     1M / 12M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/dataset.info\") with K=21\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.005     1M / 12M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.008     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.010     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.013     1M / 12M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.017     1M / 12M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (read_converter.cpp        :  78)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (read_converter.cpp        :  99)   Converting paired reads\n",
      "  0:00:00.520    82M / 109M  INFO    General                 (binary_converter.cpp      : 127)   16384 reads processed\n",
      "  0:00:00.591    82M / 115M  INFO    General                 (binary_converter.cpp      : 127)   32768 reads processed\n",
      "  0:00:00.678    82M / 125M  INFO    General                 (binary_converter.cpp      : 127)   65536 reads processed\n",
      "  0:00:01.040    82M / 142M  INFO    General                 (binary_converter.cpp      : 127)   131072 reads processed\n",
      "  0:00:01.775    61M / 151M  INFO    General                 (binary_converter.cpp      : 127)   262144 reads processed\n",
      "  0:00:02.874     0M / 152M  INFO    General                 (binary_converter.cpp      : 127)   524288 reads processed\n",
      "  0:00:04.919     0M / 152M  INFO    General                 (binary_converter.cpp      : 127)   1048576 reads processed\n",
      "  0:00:06.947     0M / 164M  INFO    General                 (binary_converter.cpp      : 143)   1568829 reads written\n",
      "  0:00:06.961     0M / 164M  INFO    General                 (read_converter.cpp        : 113)   Converting single reads\n",
      "  0:00:06.981     0M / 164M  INFO    General                 (binary_converter.cpp      : 143)   130 reads written\n",
      "  0:00:06.983     0M / 164M  INFO    General                 (read_converter.cpp        : 119)   Converting merged reads\n",
      "  0:00:06.987     0M / 164M  INFO    General                 (binary_converter.cpp      : 143)   0 reads written\n",
      "  0:00:06.990     1M / 164M  INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:06.996     1M / 164M  INFO    General                 (construction.cpp          : 150)   Max read length 100\n",
      "  0:00:06.997     1M / 164M  INFO    General                 (construction.cpp          : 156)   Average read length 98.1302\n",
      "  0:00:06.997     1M / 164M  INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:06.997     1M / 164M  INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:06.998     1M / 164M  INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:06.999     1M / 164M  INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499958 Gb\n",
      "  0:00:06.999     1M / 164M  INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355163\n",
      "  0:00:19.513  1161M / 1162M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3454037 reads\n",
      "  0:00:29.462  1161M / 1162M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6275576 reads\n",
      "  0:00:29.466     1M / 1162M INFO    General                 (kmer_splitters.hpp        : 134)   Used 6275576 reads\n",
      "  0:00:29.543     1M / 1162M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:29.591     1M / 1162M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1953020 kmers in total.\n",
      "  0:00:29.594     1M / 1162M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:00:29.598     1M / 1162M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:00:29.601     1M / 1162M INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:29.604     1M / 1162M INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:29.605     1M / 1162M INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499951 Gb\n",
      "  0:00:29.606     1M / 1162M INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 3355114\n",
      "  0:00:30.023  1161M / 1162M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 1953020 kmers\n",
      "  0:00:30.027  1161M / 1162M INFO    General                 (kmer_splitters.hpp        : 202)   Used 1953020 kmers.\n",
      "  0:00:30.030     1M / 1162M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:30.061     1M / 1162M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1948267 kmers in total.\n",
      "  0:00:30.064     1M / 1162M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:30.200     2M / 1162M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1948267 kmers, 1422112 bytes occupied (5.8395 bits per kmer).\n",
      "  0:00:30.205     2M / 1162M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:00:30.252     4M / 1162M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:00:30.686     4M / 1162M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:30.689     4M / 1162M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:00:30.691     4M / 1162M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:30.695     4M / 1162M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:00:31.095     5M / 1162M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 10038\n",
      "  0:00:31.101     5M / 1162M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 10050\n",
      "  0:00:31.103     4M / 1162M INFO   Early tip clipping       (early_simplification.hpp  :  50)   71834 22-mers were removed by early tip clipper\n",
      "  0:00:31.105     4M / 1162M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:00:31.109     4M / 1162M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:00:31.660     5M / 1162M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 18243 sequences extracted\n",
      "  0:00:32.005     5M / 1162M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:00:32.118     5M / 1162M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 2 loops collected\n",
      "  0:00:32.121     5M / 1162M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:00:32.128     5M / 1162M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:00:32.129     5M / 1162M INFO    General                 (debruijn_graph_constructor: 516)   Total 36490 edges to create\n",
      "  0:00:32.132     7M / 1162M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:00:32.147     7M / 1162M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:00:32.153     7M / 1162M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:00:32.155     7M / 1162M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:00:32.161     7M / 1162M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:00:32.162     7M / 1162M INFO    General                 (debruijn_graph_constructor: 542)   Total 13492 vertices to create\n",
      "  0:00:32.166     9M / 1162M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:00:32.174     8M / 1162M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:00:32.175     8M / 1162M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:32.315     9M / 1162M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1953020 kmers, 1425528 bytes occupied (5.83928 bits per kmer).\n",
      "  0:00:32.322    17M / 1162M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:01:04.533    17M / 1162M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:01:04.983    17M / 1162M INFO    General                 (coverage_filling.hpp      :  83)   Processed 36468 edges\n",
      "  0:01:04.997     4M / 1162M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:01:04.997     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 2\n",
      "  0:01:05.002     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 136\n",
      "  0:01:05.002     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 136. Coverage mad: 78.5778\n",
      "  0:01:05.002     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:01:05.156     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:01:05.466     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:01:07.423     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:01:13.912     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:01:24.103     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:01:35.181     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 132.718. Fitted coverage std. dev: 76.3662\n",
      "  0:01:35.189     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.877599\n",
      "  0:01:35.191     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 24\n",
      "  0:01:35.192     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 24\n",
      "  0:01:35.193     4M / 1162M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 1620351\n",
      "  0:01:35.195     4M / 1162M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 132.718\n",
      "  0:01:35.198     4M / 1162M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 24\n",
      "  0:01:35.199     4M / 1162M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:01:35.200     4M / 1162M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:01:35.202     4M / 1162M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:01:35.204     4M / 1162M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:01:35.204     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:01:35.207     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:01:35.209     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial tip clipper\n",
      "  0:01:35.229     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial tip clipper triggered 43 times\n",
      "  0:01:35.229     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial ec remover\n",
      "  0:01:35.460     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial ec remover triggered 5085 times\n",
      "  0:01:35.462     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial isolated edge remover\n",
      "  0:01:35.467     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial isolated edge remover triggered 913 times\n",
      "  0:01:35.467     4M / 1162M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:01:35.468     4M / 1162M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:01:35.470     4M / 1162M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:01:35.471     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:35.472     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.474     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 6 times\n",
      "  0:01:35.485     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.554     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 379 times\n",
      "  0:01:35.555     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.557     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:01:35.558     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:35.559     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.561     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 3 times\n",
      "  0:01:35.562     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.563     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.564     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.566     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 21 times\n",
      "  0:01:35.574     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:35.576     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.577     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.578     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.579     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 7 times\n",
      "  0:01:35.582     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.583     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 11 times\n",
      "  0:01:35.584     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:35.584     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.587     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.587     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.588     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 9 times\n",
      "  0:01:35.589     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.592     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:01:35.600     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:35.601     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.602     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.603     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.605     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:01:35.606     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.612     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 6 times\n",
      "  0:01:35.613     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:35.614     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.615     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:01:35.616     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.624     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:01:35.626     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.627     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:01:35.628     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:35.629     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.631     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.636     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.637     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.638     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.639     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 1 times\n",
      "  0:01:35.648     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:35.650     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.651     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.652     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.653     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.654     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.660     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 10 times\n",
      "  0:01:35.662     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:35.664     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.672     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:01:35.676     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.677     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.682     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.684     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4 times\n",
      "  0:01:35.692     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:35.694     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.696     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.698     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.700     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:01:35.703     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.704     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:01:35.712     4M / 1162M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:35.714     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.716     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.718     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.724     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.725     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.728     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:35.730     4M / 1162M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:01:35.731     4M / 1162M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:01:35.732     4M / 1162M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:35.733     4M / 1162M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:01:35.735     4M / 1162M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:01:35.735     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.736     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.737     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.743     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.743     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.745     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.746     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.749     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:35.750     4M / 1162M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:01:35.752     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:01:35.756     4M / 1162M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 129 times\n",
      "  0:01:35.762     4M / 1162M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:01:35.762     4M / 1162M INFO    General                 (simplification.cpp        : 509)     Average coverage = 142.976\n",
      "  0:01:35.765     4M / 1162M INFO    General                 (simplification.cpp        : 510)     Total length = 1696089\n",
      "  0:01:35.766     4M / 1162M INFO    General                 (simplification.cpp        : 511)     Median edge length: 7120\n",
      "  0:01:35.766     4M / 1162M INFO    General                 (simplification.cpp        : 512)     Edges: 1427\n",
      "  0:01:35.766     4M / 1162M INFO    General                 (simplification.cpp        : 513)     Vertices: 950\n",
      "  0:01:35.766     4M / 1162M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:01:35.767     4M / 1162M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K21/simplified_contigs\"\n",
      "  0:01:35.775     8M / 1162M INFO    General                 (binary_converter.cpp      : 143)   721 reads written\n",
      "  0:01:35.783     4M / 1162M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:01:35.796     1M / 1162M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 1 minutes 35 seconds\n",
      "\n",
      "===== K21 finished. \n",
      "\n",
      "\n",
      "===== K33 started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/spades-core /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K33/configs/config.info\n",
      "\n",
      "  0:00:00.000     1M / 12M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K33/configs/config.info\"\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 3 Gb\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.010     1M / 12M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/dataset.info\") with K=33\n",
      "  0:00:00.011     1M / 12M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
      "  0:00:00.011     1M / 12M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.011     1M / 12M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.011     1M / 12M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.011     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.011     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.012     1M / 12M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.012     1M / 12M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.012     1M / 12M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.033     1M / 12M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.038     1M / 12M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K21/simplified_contigs\n",
      "  0:00:00.039     1M / 12M   INFO    General                 (construction.cpp          : 150)   Max read length 100\n",
      "  0:00:00.039     1M / 12M   INFO    General                 (construction.cpp          : 156)   Average read length 98.1302\n",
      "  0:00:00.039     1M / 12M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.039     1M / 12M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:00.040     1M / 12M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:00.040     1M / 12M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499955 Gb\n",
      "  0:00:00.041     1M / 12M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 1677571\n",
      "  0:00:10.982  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 2044241 reads\n",
      "  0:00:20.498  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 4088542 reads\n",
      "  0:00:29.899  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6134867 reads\n",
      "  0:00:30.782  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6277018 reads\n",
      "  0:00:30.783     1M / 1128M INFO    General                 (kmer_splitters.hpp        : 134)   Used 6277018 reads\n",
      "  0:00:30.876     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:31.015     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1992428 kmers in total.\n",
      "  0:00:31.017     1M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:00:31.021     1M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:00:31.026     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:31.034     1M / 1128M INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:31.043     1M / 1128M INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499948 Gb\n",
      "  0:00:31.045     1M / 1128M INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 1677546\n",
      "  0:00:31.782  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 1992428 kmers\n",
      "  0:00:31.783  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 202)   Used 1992428 kmers.\n",
      "  0:00:31.787     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:31.832     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1990252 kmers in total.\n",
      "  0:00:31.834     1M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:32.050     2M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1990252 kmers, 1452352 bytes occupied (5.83786 bits per kmer).\n",
      "  0:00:32.055     2M / 1128M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:00:32.202     4M / 1128M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:00:33.089     4M / 1128M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:33.093     4M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Early tip clipping (id: construction:early_tip_clipper)\n",
      "  0:00:33.102     4M / 1128M INFO    General                 (construction.cpp          : 298)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:33.105     4M / 1128M INFO   Early tip clipping       (early_simplification.hpp  :  48)   Early tip clipping\n",
      "  0:00:34.008     6M / 1128M INFO   Early tip clipping       (early_simplification.hpp  :  83)   #tipped junctions: 11900\n",
      "  0:00:34.016     6M / 1128M INFO   Early tip clipping       (early_simplification.hpp  :  94)   Clipped tips: 11940\n",
      "  0:00:34.020     4M / 1128M INFO   Early tip clipping       (early_simplification.hpp  :  50)   118513 34-mers were removed by early tip clipper\n",
      "  0:00:34.023     4M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:00:34.030     4M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:00:35.304     5M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 11380 sequences extracted\n",
      "  0:00:35.892     5M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:00:36.082     5M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 3 loops collected\n",
      "  0:00:36.090     5M / 1128M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:00:36.105     5M / 1128M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:00:36.118     5M / 1128M INFO    General                 (debruijn_graph_constructor: 516)   Total 22766 edges to create\n",
      "  0:00:36.131     6M / 1128M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:00:36.145     6M / 1128M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:00:36.151     6M / 1128M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:00:36.153     7M / 1128M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:00:36.160     7M / 1128M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:00:36.161     7M / 1128M INFO    General                 (debruijn_graph_constructor: 542)   Total 9207 vertices to create\n",
      "  0:00:36.165     7M / 1128M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:00:36.172     7M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:00:36.174     7M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:36.355     8M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1992428 kmers, 1453968 bytes occupied (5.83797 bits per kmer).\n",
      "  0:00:36.362    16M / 1128M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:01:16.875    16M / 1128M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:01:17.541    16M / 1128M INFO    General                 (coverage_filling.hpp      :  83)   Processed 22758 edges\n",
      "  0:01:17.573     3M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:01:17.578     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 3\n",
      "  0:01:17.582     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 117\n",
      "  0:01:17.587     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 117. Coverage mad: 68.1996\n",
      "  0:01:17.590     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:01:17.749     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:01:18.003     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:01:18.831     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:01:21.325     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 16\n",
      "  0:01:29.662     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 32\n",
      "  0:01:35.566     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 112.27. Fitted coverage std. dev: 66.8076\n",
      "  0:01:35.573     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 0.789422\n",
      "  0:01:35.575     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 23\n",
      "  0:01:35.576     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 23\n",
      "  0:01:35.578     3M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 1598206\n",
      "  0:01:35.579     3M / 1128M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 112.27\n",
      "  0:01:35.580     3M / 1128M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 23\n",
      "  0:01:35.581     3M / 1128M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:01:35.582     3M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:01:35.584     3M / 1128M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:01:35.584     3M / 1128M INFO    General                 (graph_simplification.hpp  : 674)   Flanking coverage based disconnection disabled\n",
      "  0:01:35.586     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:01:35.588     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 1 times\n",
      "  0:01:35.589     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial tip clipper\n",
      "  0:01:35.592     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial tip clipper triggered 185 times\n",
      "  0:01:35.597     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial ec remover\n",
      "  0:01:35.816     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial ec remover triggered 2967 times\n",
      "  0:01:35.816     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Initial isolated edge remover\n",
      "  0:01:35.830     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Initial isolated edge remover triggered 1021 times\n",
      "  0:01:35.830     3M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:01:35.834     3M / 1128M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:01:35.838     3M / 1128M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:01:35.842     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:35.846     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.848     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 5 times\n",
      "  0:01:35.848     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.929     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 237 times\n",
      "  0:01:35.930     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.942     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 4 times\n",
      "  0:01:35.942     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:35.946     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.952     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 2 times\n",
      "  0:01:35.953     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.958     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 2 times\n",
      "  0:01:35.959     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.960     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 14 times\n",
      "  0:01:35.961     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:35.962     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.962     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.963     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.965     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 5 times\n",
      "  0:01:35.966     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.969     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 7 times\n",
      "  0:01:35.969     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:35.970     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.970     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 1 times\n",
      "  0:01:35.971     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.979     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 8 times\n",
      "  0:01:35.980     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.981     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 3 times\n",
      "  0:01:35.981     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:35.983     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.991     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.992     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:35.995     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:01:35.996     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:35.997     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:01:35.999     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:35.999     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:35.999     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:35.999     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.002     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 1 times\n",
      "  0:01:36.003     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.003     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.004     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.005     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:36.005     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.006     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.007     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.007     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.008     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.008     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.008     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:36.008     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.009     3M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:36.009     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.014     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.014     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.014     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.015     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:36.015     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:36.016     3M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:01:36.016     3M / 1128M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:01:36.016     3M / 1128M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:36.016     3M / 1128M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:01:36.016     3M / 1128M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:01:36.017     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.017     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.017     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.019     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.021     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:36.022     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:36.022     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:36.022     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:36.022     3M / 1128M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:01:36.023     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:01:36.025     3M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 103 times\n",
      "  0:01:36.026     3M / 1128M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:01:36.026     3M / 1128M INFO    General                 (simplification.cpp        : 509)     Average coverage = 120.656\n",
      "  0:01:36.026     3M / 1128M INFO    General                 (simplification.cpp        : 510)     Total length = 1691087\n",
      "  0:01:36.027     3M / 1128M INFO    General                 (simplification.cpp        : 511)     Median edge length: 82314\n",
      "  0:01:36.027     3M / 1128M INFO    General                 (simplification.cpp        : 512)     Edges: 340\n",
      "  0:01:36.027     3M / 1128M INFO    General                 (simplification.cpp        : 513)     Vertices: 250\n",
      "  0:01:36.027     3M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:01:36.027     3M / 1128M INFO    General                 (read_converter.cpp        : 135)   Outputting contigs to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K33/simplified_contigs\"\n",
      "  0:01:36.037     7M / 1128M INFO    General                 (binary_converter.cpp      : 143)   171 reads written\n",
      "  0:01:36.045     3M / 1128M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:01:36.049     1M / 1128M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 1 minutes 36 seconds\n",
      "\n",
      "===== K33 finished. \n",
      "\n",
      "\n",
      "===== K55 started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/spades-core /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/configs/config.info\n",
      "\n",
      "  0:00:00.000     1M / 12M   INFO    General                 (main.cpp                  :  94)   Loaded config from \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/configs/config.info\"\n",
      "  0:00:00.007     1M / 12M   INFO    General                 (memory_limit.cpp          :  55)   Memory limit set to 3 Gb\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (main.cpp                  : 102)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (main.cpp                  : 103)   Maximum k-mer length: 128\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (main.cpp                  : 104)   Assembling dataset (\"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/dataset.info\") with K=55\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (main.cpp                  : 105)   Maximum # of threads to use (adjusted due to OMP capabilities): 2\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (pipeline.cpp              : 212)   SPAdes started\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (pipeline.cpp              : 225)   Starting from stage: read_conversion\n",
      "  0:00:00.015     1M / 12M   INFO    General                 (pipeline.cpp              : 234)   Two-step repeat resolution disabled\n",
      "  0:00:00.016     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 689)   Graph created, vertex min_id: 3, edge min_id: 3\n",
      "  0:00:00.016     1M / 12M   INFO   GraphCore                (graph_core.hpp            : 690)   Vertex size: 48, edge size: 40\n",
      "  0:00:00.016     1M / 12M   INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:00.016     1M / 12M   INFO    General                 (pipeline.cpp              : 245)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.016     1M / 12M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == Binary Read Conversion (id: read_conversion)\n",
      "  0:00:00.016     1M / 12M   INFO    General                 (read_converter.cpp        :  57)   Binary reads detected\n",
      "  0:00:00.019     1M / 12M   INFO   StageManager             (stage.cpp                 : 189)   STAGE == de Bruijn graph construction (id: construction)\n",
      "  0:00:00.019     1M / 12M   INFO    General                 (construction.cpp          : 115)   Contigs from previous K will be used: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K33/simplified_contigs\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (construction.cpp          : 150)   Max read length 100\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (construction.cpp          : 156)   Average read length 98.1302\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (stage.cpp                 : 121)   PROCEDURE == k+1-mer counting (id: construction:kpomer_counting)\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499955 Gb\n",
      "  0:00:00.020     1M / 12M   INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 1677571\n",
      "  0:00:09.943  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 3078873 reads\n",
      "  0:00:19.823  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6160866 reads\n",
      "  0:00:20.460  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 128)   Processed 6275918 reads\n",
      "  0:00:20.460     1M / 1128M INFO    General                 (kmer_splitters.hpp        : 134)   Used 6275918 reads\n",
      "  0:00:20.572     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:20.709     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1995547 kmers in total.\n",
      "  0:00:20.712     1M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Extension index construction (id: construction:extension_index_construction)\n",
      "  0:00:20.718     1M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 453)   Building kmer index\n",
      "  0:00:20.720     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 258)   Splitting kmer instances into 20 files using 2 threads. This might take a while.\n",
      "  0:00:20.724     1M / 1128M INFO    General                 (file_limit.hpp            :  43)   Open file limit set to 8192\n",
      "  0:00:20.728     1M / 1128M INFO    General                 (kmer_splitter.hpp         :  94)   Memory available for splitting buffers: 0.499948 Gb\n",
      "  0:00:20.729     1M / 1128M INFO    General                 (kmer_splitter.hpp         : 102)   Using cell size of 1677546\n",
      "  0:00:21.447  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 197)   Processed 1995547 kmers\n",
      "  0:00:21.449  1161M / 1161M INFO    General                 (kmer_splitters.hpp        : 202)   Used 1995547 kmers.\n",
      "  0:00:21.454     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 264)   Starting k-mer counting.\n",
      "  0:00:21.520     1M / 1128M INFO    General                 (kmer_index_builder.hpp    : 275)   K-mer counting done. There are 1997310 kmers in total.\n",
      "  0:00:21.522     1M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:21.722     2M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1997310 kmers, 1457520 bytes occupied (5.83793 bits per kmer).\n",
      "  0:00:21.725     2M / 1128M INFO    General                 (kmer_index_builder.hpp    : 168)   Merging final buckets.\n",
      "  0:00:21.837     4M / 1128M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 101)   Building k-mer extensions from k+1-mers\n",
      "  0:00:22.577     4M / 1128M INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 106)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:22.595     4M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Condensing graph (id: construction:graph_condensing)\n",
      "  0:00:22.600     4M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 381)   Extracting unbranching paths\n",
      "  0:00:23.528     6M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 400)   Extracting unbranching paths finished. 31468 sequences extracted\n",
      "  0:00:24.224     6M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 336)   Collecting perfect loops\n",
      "  0:00:24.388     6M / 1128M INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 369)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:00:24.395     6M / 1128M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 586)   Sorting edges...\n",
      "  0:00:24.410     6M / 1128M INFO   DeBruijnGraphConstructor (debruijn_graph_constructor: 588)   Edges sorted\n",
      "  0:00:24.417     6M / 1128M INFO    General                 (debruijn_graph_constructor: 516)   Total 62936 edges to create\n",
      "  0:00:24.423     9M / 1128M INFO    General                 (debruijn_graph_constructor: 519)   Collecting link records\n",
      "  0:00:24.461    10M / 1128M INFO    General                 (debruijn_graph_constructor: 521)   Ordering link records\n",
      "  0:00:24.471    10M / 1128M INFO    General                 (debruijn_graph_constructor: 524)   Sorting done\n",
      "  0:00:24.477    10M / 1128M INFO    General                 (debruijn_graph_constructor: 537)   Sorting LinkRecords...\n",
      "  0:00:24.487    10M / 1128M INFO    General                 (debruijn_graph_constructor: 540)   LinkRecords sorted\n",
      "  0:00:24.492    10M / 1128M INFO    General                 (debruijn_graph_constructor: 542)   Total 33231 vertices to create\n",
      "  0:00:24.498    13M / 1128M INFO    General                 (debruijn_graph_constructor: 545)   Connecting the graph\n",
      "  0:00:24.520    11M / 1128M INFO    General                 (stage.cpp                 : 121)   PROCEDURE == Filling coverage indices (PHM) (id: construction:coverage_filling_phm)\n",
      "  0:00:24.524    11M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:24.728    13M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1995547 kmers, 1456232 bytes occupied (5.83793 bits per kmer).\n",
      "  0:00:24.739    21M / 1128M INFO    General                 (coverage_hash_map_builder.:  49)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:53.921    21M / 1128M INFO    General                 (construction.cpp          : 427)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:54.602    21M / 1128M INFO    General                 (coverage_filling.hpp      :  83)   Processed 62932 edges\n",
      "  0:00:54.625     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == EC Threshold Finding (id: ec_threshold_finder)\n",
      "  0:00:54.626     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 182)   Kmer coverage valley at: 0\n",
      "  0:00:54.630     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 202)   K-mer histogram maximum: 1\n",
      "  0:00:54.634     8M / 1128M WARN    General                 (kmer_coverage_model.cpp   : 219)   Too many erroneous kmers, the estimates might be unreliable\n",
      "  0:00:54.638     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 238)   Estimated median coverage: 1. Coverage mad: 3.31519\n",
      "  0:00:54.642     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 260)   Fitting coverage model\n",
      "  0:00:54.648     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 2\n",
      "  0:00:54.655     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 4\n",
      "  0:00:54.670     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 296)   ... iteration 8\n",
      "  0:00:54.704     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 310)   Fitted mean coverage: 1. Fitted coverage std. dev: 1.92739e-16\n",
      "  0:00:54.705     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 335)   Probability of erroneous kmer at valley: 5.37206e-19\n",
      "  0:00:54.709     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 359)   Preliminary threshold calculated as: 1\n",
      "  0:00:54.713     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 363)   Threshold adjusted to: 0\n",
      "  0:00:54.717     8M / 1128M INFO    General                 (kmer_coverage_model.cpp   : 376)   Estimated genome size (ignoring repeats): 0\n",
      "  0:00:54.720     8M / 1128M INFO    General                 (genomic_info_filler.cpp   :  56)   Mean coverage was calculated as 1\n",
      "  0:00:54.725     8M / 1128M INFO    General                 (genomic_info_filler.cpp   :  71)   EC coverage threshold value was calculated as 0\n",
      "  0:00:54.728     8M / 1128M INFO    General                 (genomic_info_filler.cpp   :  72)   Trusted kmer low bound: 0\n",
      "  0:00:54.735     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: early_gapcloser)\n",
      "  0:00:54.748    21M / 1128M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:00:54.761    22M / 1128M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 18719 out of 62932, length: 381876\n",
      "  0:00:54.766    22M / 1128M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 63253)\n",
      "  0:00:54.798    22M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:00:54.923    22M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 294078 kmers, 213136 bytes occupied (5.79808 bits per kmer).\n",
      "  0:00:54.927    24M / 1128M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:00:54.990    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:01:03.504    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:01:03.625    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:01:10.793    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:01:11.021    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:01:19.034    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:01:19.107    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:01:28.986    24M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 1568829 reads processed\n",
      "  0:01:28.996    22M / 1128M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:01:29.000    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:01:29.008    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 104 tips collected, total 180 connection candidates\n",
      "  0:01:29.013    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 2 gaps after checking 178 candidates\n",
      "  0:01:29.017    22M / 1128M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:01:29.022     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Raw Simplification (id: raw_simplification)\n",
      "  0:01:29.025     8M / 1128M INFO    General                 (simplification.cpp        : 129)   PROCEDURE == Initial cleaning\n",
      "  0:01:29.028     8M / 1128M INFO    General                 (simplification.cpp        :  68)   Most init cleaning disabled since detected mean 1 was less than activation coverage 10\n",
      "  0:01:29.031     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Self conjugate edge remover\n",
      "  0:01:29.037     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Self conjugate edge remover triggered 0 times\n",
      "  0:01:29.040     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification (id: simplification)\n",
      "  0:01:29.042     8M / 1128M INFO    General                 (simplification.cpp        : 397)   Graph simplification started\n",
      "  0:01:29.043     8M / 1128M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:01:29.044     8M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:29.045     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.267     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 14321 times\n",
      "  0:01:29.272     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.392     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 225 times\n",
      "  0:01:29.397     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.403     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 2 times\n",
      "  0:01:29.407     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:29.412     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.418     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.421     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.425     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 3 times\n",
      "  0:01:29.427     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.439     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.442     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:29.443     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.444     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.444     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.445     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.446     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.446     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.447     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.447     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.447     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.451     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.451     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.452     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.452     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.453     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.453     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:29.453     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.454     9M / 1128M INFO    General                 (simplification.cpp        : 402)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:29.454     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:01:29.455     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:01:29.456     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:01:29.457     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:01:29.459     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Low coverage edge remover\n",
      "  0:01:29.465     9M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Low coverage edge remover triggered 0 times\n",
      "  0:01:29.467     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Gap Closer (id: late_gapcloser)\n",
      "  0:01:29.475    21M / 1128M INFO    General                 (edge_index.hpp            : 132)   Size of edge index entries: 12/8\n",
      "  0:01:29.483    22M / 1128M INFO    General                 (gap_closer.cpp            : 102)   Total edges in tip neighborhood: 4147 out of 4540, length: 288268\n",
      "  0:01:29.484    22M / 1128M INFO    General                 (edge_index.hpp            : 196)   Using small index (max_id = 63253)\n",
      "  0:01:29.517    22M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:01:29.623    22M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 203628 kmers, 147808 bytes occupied (5.80698 bits per kmer).\n",
      "  0:01:29.625    23M / 1128M INFO    General                 (edge_index_builders.hpp   : 266)   Collecting edge information from graph, this takes a while.\n",
      "  0:01:29.659    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:01:37.811    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:01:37.938    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:01:46.060    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:01:46.180    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:01:53.865    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:01:54.025    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:02:02.530    23M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 1568829 reads processed\n",
      "  0:02:02.537    22M / 1128M INFO    General                 (gap_closer.cpp            : 491)   Initializing gap closer\n",
      "  0:02:02.537    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 406)   Collecting gap candidates\n",
      "  0:02:02.549    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 410)   Total 8 tips collected, total 10 connection candidates\n",
      "  0:02:02.554    22M / 1128M INFO   GapCloser                (gap_closer.cpp            : 431)   Closing short gaps complete: filled 1 gaps after checking 9 candidates\n",
      "  0:02:02.561    22M / 1128M INFO    General                 (gap_closer.cpp            : 495)   Gap closer done\n",
      "  0:02:02.564     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Simplification Cleanup (id: simplification_cleanup)\n",
      "  0:02:02.573     8M / 1128M INFO    General                 (simplification.cpp        : 189)   PROCEDURE == Post simplification\n",
      "  0:02:02.577     8M / 1128M INFO    General                 (graph_simplification.hpp  : 455)   Disconnection of relatively low covered edges disabled\n",
      "  0:02:02.578     8M / 1128M INFO    General                 (graph_simplification.hpp  : 494)   Complex tip clipping disabled\n",
      "  0:02:02.580     8M / 1128M INFO    General                 (graph_simplification.hpp  : 646)   Creating parallel br instance\n",
      "  0:02:02.582     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:02:02.584     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:02:02.585     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:02:02.588     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:02:02.599     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Tip clipper\n",
      "  0:02:02.602     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Tip clipper triggered 0 times\n",
      "  0:02:02.605     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Bulge remover\n",
      "  0:02:02.608     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Bulge remover triggered 0 times\n",
      "  0:02:02.610     8M / 1128M INFO    General                 (simplification.cpp        : 348)   Disrupting self-conjugate edges\n",
      "  0:02:02.614     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 168)   Running Removing isolated edges\n",
      "  0:02:02.619     8M / 1128M INFO   Simplification           (parallel_processing.hpp   : 171)   Removing isolated edges triggered 2048 times\n",
      "  0:02:02.623     8M / 1128M INFO    General                 (simplification.cpp        : 508)   After simplification:\n",
      "  0:02:02.623     8M / 1128M INFO    General                 (simplification.cpp        : 509)     Average coverage = 79.8511\n",
      "  0:02:02.623     8M / 1128M INFO    General                 (simplification.cpp        : 510)     Total length = 1694863\n",
      "  0:02:02.624     8M / 1128M INFO    General                 (simplification.cpp        : 511)     Median edge length: 97686\n",
      "  0:02:02.624     8M / 1128M INFO    General                 (simplification.cpp        : 512)     Edges: 445\n",
      "  0:02:02.624     8M / 1128M INFO    General                 (simplification.cpp        : 513)     Vertices: 336\n",
      "  0:02:02.624     8M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Mismatch Correction (id: mismatch_correction)\n",
      "  0:02:02.624     8M / 1128M INFO    General                 (graph_pack_helpers.cpp    :  44)   Index refill\n",
      "  0:02:02.624     8M / 1128M INFO    General                 (edge_index.hpp            : 175)   Using small index (max_id = 63253)\n",
      "  0:02:02.730     8M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 410)   Building perfect hash indices\n",
      "  0:02:03.563     9M / 1128M INFO   K-mer Index Building     (kmer_index_builder.hpp    : 446)   Index built. Total 1694841 kmers, 1224864 bytes occupied (5.78161 bits per kmer).\n",
      "  0:02:03.570    22M / 1128M INFO    General                 (edge_index_builders.hpp   : 253)   Collecting edge information from graph, this takes a while.\n",
      "  0:02:03.890    22M / 1128M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 24368 kmers to process\n",
      "  0:02:03.899    22M / 1128M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:02:03.900    22M / 1128M INFO    General                 (mismatch_correction.cpp   : 392)   Collect potential mismatches\n",
      "  0:02:03.940    22M / 1128M INFO    General                 (mismatch_correction.cpp   : 193)   Total 116 edges (out of 445) with 8944 potential mismatch positions (77.1034 positions per edge)\n",
      "  0:02:03.941    22M / 1128M INFO    General                 (mismatch_correction.cpp   : 394)   Potential mismatches collected\n",
      "  0:02:03.942    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:02:05.467    23M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:02:05.500    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:02:07.067    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:02:07.116    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:02:08.638    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:02:08.702    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:02:12.961    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 2200000 reads\n",
      "  0:02:20.490    24M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 4200000 reads\n",
      "  0:02:27.945    24M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 6275576 reads processed\n",
      "  0:02:27.989    23M / 1128M INFO    General                 (mismatch_correction.cpp   : 387)   All edges processed\n",
      "  0:02:27.993    22M / 1128M INFO    General                 (mismatch_correction.cpp   : 442)   Corrected 0 nucleotides\n",
      "  0:02:27.998    22M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:02:27.998    22M / 1128M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/before_rr.fasta\"\n",
      "  0:02:28.029    22M / 1128M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph_after_simplification.gfa\"\n",
      "  0:02:28.041    22M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Paired Information Counting (id: late_pair_info_count)\n",
      "  0:02:28.042    22M / 1128M INFO    General                 (graph_pack_helpers.cpp    :  54)   Normalizing k-mer map. Total 24368 kmers to process\n",
      "  0:02:28.045    22M / 1128M INFO    General                 (graph_pack_helpers.cpp    :  56)   Normalizing done\n",
      "  0:02:28.050    22M / 1128M INFO    General                 (pair_info_count.cpp       : 157)   Min edge length for estimation: 97686\n",
      "  0:02:28.053    22M / 1128M INFO    General                 (pair_info_count.cpp       : 168)   Estimating insert size for library #0\n",
      "  0:02:28.058    22M / 1128M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:02:28.060    22M / 1128M INFO    General                 (paired_info_utils.cpp     :  87)   Estimating insert size (takes a while)\n",
      "  0:02:28.100    70M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:02:29.923    70M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:02:29.958    70M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:02:31.775    71M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:02:31.835    71M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:02:33.500    71M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:02:33.566    71M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:02:35.079    70M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 1568829 reads processed\n",
      "  0:02:35.315    70M / 1128M INFO    General                 (paired_info_utils.cpp     : 104)   Edge pairs: 1638\n",
      "  0:02:35.316    70M / 1128M INFO    General                 (paired_info_utils.cpp     : 106)   748405 paired reads (47.7047% of all) aligned to long edges\n",
      "  0:02:35.320    22M / 1128M INFO    General                 (pair_info_count.cpp       : 188)     Insert size = 230.269, deviation = 100.276, left quantile = 118, right quantile = 364, read length = 100\n",
      "  0:02:35.322    22M / 1128M INFO    General                 (pair_info_count.cpp       : 203)   Filtering data for library #0\n",
      "  0:02:35.326    22M / 1128M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:02:35.329    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:02:36.673    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:02:36.691    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:02:37.947    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:02:37.986    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:02:39.251    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:02:39.284    22M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:02:40.467    22M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 1568829 reads processed\n",
      "  0:02:40.468    22M / 1128M INFO    General                 (pair_info_count.cpp       : 207)   Mapping library #0\n",
      "  0:02:40.474    22M / 1128M INFO    General                 (pair_info_count.cpp       : 209)   Mapping paired reads (takes a while)\n",
      "  0:02:40.477    22M / 1128M INFO    General                 (pair_info_count.cpp       :  41)   Selecting usual mapper\n",
      "  0:02:40.480    22M / 1128M INFO    General                 (paired_info_utils.cpp     : 138)   Left insert size quantile 118, right insert size quantile 364, filtering threshold 2, rounding threshold 0\n",
      "  0:02:40.485    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  64)   Starting sequence mapping\n",
      "  0:02:41.882    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 200000 reads\n",
      "  0:02:41.897    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 400000 reads\n",
      "  0:02:43.282    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 600000 reads\n",
      "  0:02:43.305    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 800000 reads\n",
      "  0:02:44.686    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1000000 reads\n",
      "  0:02:44.721    36M / 1128M INFO    General                 (sequence_mapper_notifier.h:  85)   Processed 1200000 reads\n",
      "  0:02:46.008    36M / 1128M INFO    General                 (sequence_mapper_notifier.h: 103)   Total 1568829 reads processed\n",
      "  0:02:46.018    23M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Distance Estimation (id: distance_estimation)\n",
      "  0:02:46.018    23M / 1128M INFO    General                 (distance_estimation.cpp   :  42)   Processing library #0\n",
      "  0:02:46.021    23M / 1128M INFO    General                 (distance_estimation_utils.: 133)   Weight Filter Done\n",
      "  0:02:46.024    23M / 1128M INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SIMPLE distance estimator\n",
      "  0:02:46.032    23M / 1128M INFO    General                 (distance_estimation_utils.:  28)   Filtering info\n",
      "  0:02:46.032    23M / 1128M INFO    General                 (pair_info_filters.hpp     : 243)   Start filtering; library index size: 1477\n",
      "  0:02:46.038    23M / 1128M INFO    General                 (pair_info_filters.hpp     : 264)   Done filtering; library index size: 1140\n",
      "  0:02:46.038    23M / 1128M INFO    General                 (distance_estimation_utils.: 139)   Refining clustered pair information\n",
      "  0:02:46.040    23M / 1128M INFO    General                 (distance_estimation_utils.: 141)   The refining of clustered pair information has been finished\n",
      "  0:02:46.040    23M / 1128M INFO    General                 (distance_estimation_utils.: 143)   Improving paired information\n",
      "  0:02:46.051    23M / 1128M INFO   PairInfoImprover         (pair_info_improver.hpp    :  65)   Paired info stats: missing = 64; contradictional = 86\n",
      "  0:02:46.055    23M / 1128M INFO   PairInfoImprover         (pair_info_improver.hpp    :  65)   Paired info stats: missing = 2; contradictional = 4\n",
      "  0:02:46.055    23M / 1128M INFO    General                 (distance_estimation_utils.:  86)   Filling scaffolding index\n",
      "  0:02:46.058    23M / 1128M INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SMOOTHING distance estimator\n",
      "  0:02:46.079    23M / 1128M INFO    General                 (distance_estimation_utils.:  28)   Filtering info\n",
      "  0:02:46.079    23M / 1128M INFO    General                 (pair_info_filters.hpp     : 243)   Start filtering; library index size: 581\n",
      "  0:02:46.082    23M / 1128M INFO    General                 (pair_info_filters.hpp     : 264)   Done filtering; library index size: 581\n",
      "  0:02:46.085    23M / 1128M INFO    General                 (distance_estimation.cpp   :  51)   Clearing raw paired index\n",
      "  0:02:46.088    22M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Repeat Resolving (id: repeat_resolving)\n",
      "  0:02:46.090    22M / 1128M INFO    General                 (repeat_resolving.cpp      :  88)   Using Path-Extend repeat resolving\n",
      "  0:02:46.094    22M / 1128M INFO    General                 (launcher.cpp              : 603)   ExSPAnder repeat resolving tool started\n",
      "  0:02:46.095    22M / 1128M INFO    General                 (coverage_uniformity_analyz:  25)   genomic coverage is 76.7565 calculated of length 1681418\n",
      "  0:02:46.095    22M / 1128M WARN    General                 (launcher.cpp              : 180)   Your data seems to have high uniform coverage depth. It is strongly recommended to use --isolate option.\n",
      "  0:02:46.096    23M / 1128M INFO    General                 (launcher.cpp              : 420)   Creating main extenders, unique edge length = 2000\n",
      "  0:02:46.097    23M / 1128M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 79.8511\n",
      "  0:02:46.097    23M / 1128M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 1182\n",
      "  0:02:46.097    23M / 1128M INFO    General                 (extenders_logic.cpp       : 341)   Estimated coverage of library #0 is 79.8511\n",
      "  0:02:46.098    23M / 1128M INFO    General                 (extenders_logic.cpp       : 352)   Creating extender; library index size: 1182\n",
      "  0:02:46.099    23M / 1128M INFO    General                 (extenders_logic.cpp       : 550)   Using 1 paired-end library\n",
      "  0:02:46.099    23M / 1128M INFO    General                 (extenders_logic.cpp       : 551)   Using 1 paired-end scaffolding library\n",
      "  0:02:46.099    23M / 1128M INFO    General                 (extenders_logic.cpp       : 552)   Using 0 single read libraries\n",
      "  0:02:46.099    23M / 1128M INFO    General                 (launcher.cpp              : 449)   Total number of extenders is 3\n",
      "  0:02:46.099    23M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 0 paths from 208 (0%)\n",
      "  0:02:46.105    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 21 paths from 208 (10%)\n",
      "  0:02:46.112    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 42 paths from 208 (20%)\n",
      "  0:02:46.117    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 63 paths from 208 (30%)\n",
      "  0:02:46.119    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 84 paths from 208 (40%)\n",
      "  0:02:46.121    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 105 paths from 208 (50%)\n",
      "  0:02:46.124    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 126 paths from 208 (60%)\n",
      "  0:02:46.127    24M / 1128M INFO    General                 (path_extenders.cpp        :  34)   Processed 128 paths from 208 (61%)\n",
      "  0:02:46.127    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 147 paths from 208 (70%)\n",
      "  0:02:46.128    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 168 paths from 208 (80%)\n",
      "  0:02:46.128    24M / 1128M INFO    General                 (path_extenders.cpp        :  36)   Processed 189 paths from 208 (90%)\n",
      "  0:02:46.130    23M / 1128M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:02:46.130    23M / 1128M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:02:46.131    23M / 1128M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:02:46.131    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:02:46.131    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:02:46.131    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:02:46.132    23M / 1128M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:02:46.132    23M / 1128M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:02:46.133    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:02:46.133    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:02:46.134    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:02:46.134    23M / 1128M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:02:46.135    23M / 1128M INFO    General                 (launcher.cpp              : 456)   Closing gaps in paths\n",
      "  0:02:46.135    24M / 1128M INFO    General                 (launcher.cpp              : 486)   Gap closing completed\n",
      "  0:02:46.137    23M / 1128M INFO    General                 (launcher.cpp              : 304)   Traversing tandem repeats\n",
      "  0:02:46.139    23M / 1128M INFO    General                 (launcher.cpp              : 314)   Traversed 0 loops\n",
      "  0:02:46.140    23M / 1128M INFO    General                 (launcher.cpp              : 252)   Finalizing paths\n",
      "  0:02:46.140    23M / 1128M INFO    General                 (launcher.cpp              : 254)   Deduplicating paths\n",
      "  0:02:46.140    23M / 1128M INFO    General                 (launcher.cpp              : 258)   Paths deduplicated\n",
      "  0:02:46.141    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  60)   Removing overlaps\n",
      "  0:02:46.141    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  63)   Sorting paths\n",
      "  0:02:46.141    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  70)   Marking overlaps\n",
      "  0:02:46.141    23M / 1128M INFO   OverlapRemover           (overlap_remover.hpp       : 117)   Marking start/end overlaps\n",
      "  0:02:46.141    23M / 1128M INFO   OverlapRemover           (overlap_remover.hpp       : 120)   Marking remaining overlaps\n",
      "  0:02:46.142    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  73)   Splitting paths\n",
      "  0:02:46.142    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  78)   Deduplicating paths\n",
      "  0:02:46.142    23M / 1128M INFO   PEResolver               (pe_resolver.cpp           :  80)   Overlaps removed\n",
      "  0:02:46.143    23M / 1128M INFO    General                 (launcher.cpp              : 275)   Paths finalized\n",
      "  0:02:46.143    23M / 1128M INFO    General                 (launcher.cpp              : 666)   ExSPAnder repeat resolving tool finished\n",
      "  0:02:46.143    23M / 1128M INFO   StageManager             (stage.cpp                 : 189)   STAGE == Contig Output (id: contig_output)\n",
      "  0:02:46.143    23M / 1128M INFO    General                 (contig_output.hpp         :  20)   Outputting contigs to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/before_rr.fasta\"\n",
      "  0:02:46.170    23M / 1128M INFO    General                 (contig_output_stage.cpp   : 155)   Writing GFA graph to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph_with_scaffolds.gfa\"\n",
      "  0:02:46.196    23M / 1128M INFO    General                 (contig_output_stage.cpp   : 169)   Outputting FastG graph to \"/home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph.fastg\"\n",
      "  0:02:46.267    23M / 1128M INFO    General                 (contig_output_stage.cpp   : 200)   Breaking scaffolds\n",
      "  0:02:46.287    25M / 1128M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/final_contigs.fasta\n",
      "  0:02:46.301    25M / 1128M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/final_contigs.paths\n",
      "  0:02:46.318    25M / 1128M INFO    General                 (contig_output_stage.cpp   : 101)   Outputting contigs to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/scaffolds.fasta\n",
      "  0:02:46.331    25M / 1128M INFO    General                 (contig_output_stage.cpp   : 107)   Outputting FastG paths to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/scaffolds.paths\n",
      "  0:02:46.335    25M / 1128M INFO    General                 (contig_output_stage.cpp   : 114)   Populating GFA with scaffold paths\n",
      "  0:02:46.339    23M / 1128M INFO    General                 (pipeline.cpp              : 292)   SPAdes finished\n",
      "  0:02:46.346     1M / 1128M INFO    General                 (main.cpp                  : 131)   Assembling time: 0 hours 2 minutes 46 seconds\n",
      "\n",
      "===== K55 finished. \n",
      "\n",
      "\n",
      "===== Copy files started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/python3 /home/ec2-user/anaconda3/envs/python3/share/spades/spades_pipeline/scripts/copy_files.py /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/before_rr.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/before_rr.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph_after_simplification.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/assembly_graph_after_simplification.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/final_contigs.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/contigs.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/first_pe_contigs.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/first_pe_contigs.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/strain_graph.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/strain_graph.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/scaffolds.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/scaffolds.fasta /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/scaffolds.paths /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/scaffolds.paths /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph_with_scaffolds.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/assembly_graph_with_scaffolds.gfa /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/assembly_graph.fastg /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/assembly_graph.fastg /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/K55/final_contigs.paths /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/contigs.paths\n",
      "\n",
      "\n",
      "===== Copy files finished. \n",
      "\n",
      "\n",
      "===== Assembling finished. \n",
      "\n",
      "\n",
      "===== Breaking scaffolds started. \n",
      "\n",
      "\n",
      "== Running: /home/ec2-user/anaconda3/envs/python3/bin/python3 /home/ec2-user/anaconda3/envs/python3/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/scaffolds.fasta --misc_dir /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/misc --threshold_for_breaking_scaffolds 3\n",
      "\n",
      "\n",
      "===== Breaking scaffolds finished. \n",
      "\n",
      "\n",
      "===== Terminate started. \n",
      "\n",
      "\n",
      "===== Terminate finished. \n",
      "\n",
      " * Corrected reads are in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/corrected/\n",
      " * Assembled contigs are in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/contigs.fasta\n",
      " * Assembled scaffolds are in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/scaffolds.fasta\n",
      " * Paths in the assembly graph corresponding to the contigs are in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/scaffolds.paths\n",
      " * Assembly graph is in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/assembly_graph_with_scaffolds.gfa\n",
      "\n",
      "======= SPAdes pipeline finished WITH WARNINGS!\n",
      "\n",
      "=== Error correction and assembling warnings:\n",
      " * 0:00:54.634     8M / 1128M WARN    General                 (kmer_coverage_model.cpp   : 219)   Too many erroneous kmers, the estimates might be unreliable\n",
      " * 0:02:46.095    22M / 1128M WARN    General                 (launcher.cpp              : 180)   Your data seems to have high uniform coverage depth. It is strongly recommended to use --isolate option.\n",
      "======= Warnings saved to /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/warnings.log\n",
      "\n",
      "SPAdes log can be found here: /home/ec2-user/SageMaker/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-spades/spades.log\n",
      "\n",
      "Thank you for using SPAdes! If you use it in your research, please cite:\n",
      "\n",
      "  Prjibelski, A., Antipov, D., Meleshko, D., Lapidus, A. and Korobeynikov, A., 2020. Using SPAdes de novo assembler. Current protocols in bioinformatics, 70(1), p.e102.\n",
      "  doi.org/10.1002/cpbi.102\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "## Assemble trimmed reads using the SPAdes assembler.\n",
    "\n",
    "# define input\n",
    "forward=output-fastp/trimmed_1.fastq.gz\n",
    "reverse=output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# assemble reads with SPAdes\n",
    "spades.py -1 $forward -2 $reverse -o output-spades"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
