{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3054864-6f9a-44d5-b3a3-ff29ccde5e7c",
   "metadata": {},
   "source": [
    "# Submodule 1: Introduction to genome sequencing and assembly\n",
    "--------\n",
    "Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. \n",
    "\n",
    "With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. A critical component of genomics is genome assembly, which involves reconstructing the genome of an organism from short DNA sequences generated by sequencing technologies. Genome assembly is a foundational step that allows researchers to generate a complete picture of the genetic material in a given organism. Assessing the quality and completeness of a genome assembly ensures that it accurately reflects the original genome, providing a reliable foundation for further analysis.\n",
    "\n",
    "Comparative genomics builds on genome assembly by comparing the genomes of different species or individuals. This field focuses on identifying similarities and differences in DNA sequences to understand evolutionary relationships, discover conserved genetic elements, and reveal the genetic underpinnings of adaptation and diversity. Comparative genomics can uncover how species evolve and adapt over time and identify genes associated with specific traits or diseases.\n",
    "\n",
    "In this tutorial, we will explore the key concepts and methodologies in genomics, including how to assemble a genome, how to assess its quality, how to annotate the genome, and how to perform comparative analyses.\n",
    " \n",
    "\n",
    "### Learning Objectives:\n",
    "\n",
    "+ Understand how high throughput sequencing data is generated. \n",
    "\n",
    "+ Develop an understanding of core bioinformatic input/output formats as it relates to comparative genomics. \n",
    "\n",
    "+ Acquire the skills to assemble raw sequencing reads into a draft genome, assess the quality of the genome assembly, and annotate the genome sequence. \n",
    "\n",
    "+ Learn to perform comparative genomic analyses to identify similarities and differences across genomes, run phylogenomic analyses, construct pangenomes to capture genetic diversity, and apply these techniques to address biological questions and hypotheses. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46b5570c-8281-4d0f-a7e4-11ae5f858137",
   "metadata": {
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "003ae0cc-a061-44e1-8bd4-251cbfe8274d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Background 01: What is a genome?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15fae599-ea0f-4376-8f97-f1409a2b64d6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Backround 02:How is sequencing data produced?\n",
    "\n",
    "### What is Next-Generation Sequencing (NGS)?\n",
    "\n",
    "Next-Generation Sequencing (NGS) is a high-throughput method that allows for rapid sequencing of DNA. Illumina remains the most widely used sequenicng platform. It is accurate, inexpensive, and fast, providing up to 540 Gb of data on a single flow cell in about a 1-2 days. Other long-read sequencing technologies (i.e. PacBio and Nanopore) are becoming widely used in the research community, \n",
    "\n",
    "Figure 1, outlines the basic workflow for sequencing a microbial genome. It begins with the **isolation of bacterial cultures**, which involves culturing bacteria from a sample on selective media to obtain a pure strain. Once a pure culture is established, **DNA extraction** is performed to isolate the genomic DNA from the bacterial cells.\n",
    "\n",
    "For Illumina-based sequencing the DNA is **fragmented** into smaller pieces and attached to specially coated surfaces, creating a dense cluster of identical fragments. Through a process called **sequencing by synthesis**, fluorescently labeled nucleotides are incorporated one at a time into these clusters. Each incorporated nucleotide emits a distinct signal, allowing the sequencer to determine the DNA sequence in real time.\n",
    "\n",
    "This technology enables millions of fragments to be sequenced simultaneously, significantly increasing throughput and reducing costs.\n",
    "\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "<div style=\"text-align: center;\">\n",
    "<img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/fragmentation3.png?raw=true\" width=\"700\">\n",
    "</div>\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">\n",
    "\n",
    "Figure 1. Sequencing a bacterial isolate.\n",
    "\n",
    "## How next-generation sequencing works\n",
    "\n",
    "The following video describes the process in detail.\n",
    "\n",
    "[![Sequencing by Synthesis](images/cluster-generation.jpg)](https://www.youtube.com/watch?v=p4vKJJlNTKA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd90d797-1c42-4a44-a3b8-d4fad1ab98b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0a1459e1-27cd-4760-a822-d1f6fe9512a9",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1ee5c7c9-6c44-4da7-8df2-f7c5a40c634a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/p4vKJJlNTKA\" frameborder=\"0\" allowfullscreen></iframe>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/p4vKJJlNTKA\" frameborder=\"0\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b80936d-baa8-48db-93e9-ca122ed8558a",
   "metadata": {},
   "source": [
    "## **Starting Dataset**\n",
    "\n",
    "The dataset we will use in this tutorial \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa7626f-3af1-4d7c-8601-4dce2708a4d6",
   "metadata": {},
   "source": [
    "## **Install required software**\n",
    "\n",
    "Four main tools are required for for Submodule 1; fastq-dump, fastqc, fastp, and spades.  We will install these tools using __[Conda](https://docs.conda.io/en/latest/)__, an open-source package management system and environment management tool. Conda helps to easily install, update, and manage software dependencies across different platforms.\n",
    "\n",
    "### List of software\n",
    "- **sra-tools** - The tool `fastq-dump` will be used to retrieve publicly available data from the NCBI Sequence Read Archive (SRA).\n",
    "\n",
    "- **fastqc** - Used to assess the quality of raw sequencing data by generating visual reports on metrics like read length, GC content, and sequence duplication.\n",
    "\n",
    "- **fastp** - Used for quality control and pre-processing of FASTQ files, including read trimming, adapter removal, and filtering out low-quality reads.\n",
    "\n",
    "- **spades** - Used for genome assembly, creating contigs from high-quality, pre-processed sequencing reads.\n",
    "\n",
    "\n",
    "We will install these tools with specific versions, this ensures consistent behavior and reproducibility of results across different environments and systems."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f06ed5a-fa8c-464d-b60b-ac9307e4c1ed",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: The code below is written in Bash. The <code>%%bash</code> at the beginning of the cell instructs the Jupyter notebook to execute the code as a Bash script when the cell is run. This is the same code you would run from a standard command-line terminal</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc1aadb-f7c1-4e1e-a7e4-77cc6e60141d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: To execute the code you can either press **ctrl-ENTER** when the cell is selected or by pressing the play button at the top of the notebook</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a2fa5230-d4ee-47c1-90be-1cebea67de9c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\n",
      "Displaying versions:\u001b[0m\n",
      "\n",
      "fastq-dump : 3.1.1\n",
      "\n",
      "FastQC v0.11.8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "fastp 0.23.4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SPAdes genome assembler v4.0.0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Install all tools using mamba (a conda alternative) with specified versions\n",
    "\n",
    "mamba install --channel bioconda \\\n",
    "    python=3.9 \\\n",
    "    sra-tools=3.1.1 \\\n",
    "    fastqc=0.11.8 \\\n",
    "    fastp=0.23.4 \\\n",
    "    spades=4.0.0 \\\n",
    "    -y > /dev/null 2>&1\n",
    "\n",
    "\n",
    "echo -e \"\\033[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\\nDisplaying versions:\\033[0m\"\n",
    "\n",
    "# Confirm installation by checking versions\n",
    "fastq-dump --version\n",
    "fastqc --version\n",
    "fastp --version\n",
    "spades.py --version\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b09b30f-d22a-415f-8dd2-b3338fd07a9f",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>\\</code> to break a long command into multiple lines</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75252e5c-4e8d-4d97-987b-3111db2008d2",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>/dev/null 2>&1</code> to write standard error and standard output to <i>/dev/null</i>, effectively discarding them.</div></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6bd27d-90a9-4fac-98f5-ce003a5c097c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## ** Download Starting data **\n",
    "\n",
    "In this tutorial we utilize data obtained from \n",
    "\n",
    "The data utilized in this tutorial was generated by Illumina HiSeq 2500, is paired-end and each read is 250 bps n length. \n",
    "\n",
    "<p align=\"center\">\n",
    "    <img src=\"images/jupyterNotebook_annotated.png\" alt=\"jupyterNotebook\" width=\"50%\"/>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b43a7cc-7801-4f9e-b18c-9d5996a33166",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-10-07T21:27:16 prefetch.3.1.1: 1) Resolving 'SRR10056829'...\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: 'tools/ascp/disabled': not found in configuration\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking 'ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: 'ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking 'ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: 'ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/usr/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/usr/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/usr/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/usr/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/opt/aspera/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/opt/aspera/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/opt/aspera/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/opt/aspera/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/home/ec2-user/.aspera/connect/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/home/ec2-user/.aspera/connect/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Checking '/home/ec2-user/.aspera/connect/bin/ascp'\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: '/home/ec2-user/.aspera/connect/bin/ascp': not found\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2024-10-07T21:27:16 prefetch.3.1.1: 1) Downloading 'SRR10056829'...\n",
      "2024-10-07T21:27:16 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-10-07T21:27:16 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-10-07T21:27:22 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-10-07T21:27:22 prefetch.3.1.1:  'SRR10056829' is valid: 130149382 bytes were streamed from 130139525\n",
      "2024-10-07T21:27:22 prefetch.3.1.1: 1) 'SRR10056829' was downloaded successfully\n",
      "2024-10-07T21:27:22 prefetch.3.1.1: 'SRR10056829' has 0 unresolved dependencies\n",
      "Read 1591280 spots for SRR10056829/SRR10056829.sra\n",
      "Written 1591280 spots for SRR10056829/SRR10056829.sra\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "prefetch -v $accession\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession\n",
    "\n",
    "echo Process Complete"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1b1056-ae52-4290-a2a0-615d06512281",
   "metadata": {},
   "source": [
    "### FASTQ file format\n",
    "The data we just downloaded are in **FASTQ** file format. Let's examine this data and file format.\n",
    "\n",
    "Let's run three sections of code. 1. Ensures the data we downloaded exists. 2. Defines the forward and reverse variable names for later use and prints the first entry of the forward reads. 3.)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "ea85ebdd-d656-468a-95e0-89426db38912",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR10056829_1.fastq.gz\n",
      "SRR10056829_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# If the above code worked you should have two FASTQ files in a directory called raw-reads\n",
    "ls raw-reads/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7ccb32ad-c885-4644-9d7f-fe9d2f34417d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------- FASTQ READ ---------\n",
      "@SRR10056829.1 1 length=100\n",
      "NCCCCAAGGAATAACATTCATAACCCCCATAGCCGAAGTAATAATCAATAACATTGAAGTCTTTCTTATACCTAAACGCTCATAAATAGGTAATAATGCA\n",
      "+SRR10056829.1 1 length=100\n",
      "#0<FFFFFFFFFFIIFFIIIIIIFIIIIIIIIIFIFFIIIIIIIIFFFIIIIIIIIFIIIIIIIIIIFFFFFFFBFFFFFFFFFFFFFBFBBFFFFFFFF\n",
      "-----------------------------\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# define variable names for reads\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# view the first four lines\n",
    "echo \"-------- FASTQ READ ---------\"\n",
    "zcat $forward | head -n 4\n",
    "echo \"-----------------------------\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2ae8ac-0fb2-4123-afec-0f37f76058fc",
   "metadata": {},
   "source": [
    "The above code should have displayed the first four lines of the FASTQ file. Each sequencing read in a FASTQ formatted file is made up of four lines. \n",
    "\n",
    "The length of the sequencing reads can the length depending on the type of sequencing instrument and flow cell used, but they should be consistent across the raw data. In our example datset the reads are 100 bps in length. Illumina sequencers produce reads with length lengths randing between 50 and 300 bps. Run the following code to print the length of the reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5120eb17-44a0-44a5-86ab-68b2b7140825",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "zcat raw-reads/SRR10056829_1.fastq.gz | head -n 2 | tail -n 1 | wc -c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c457802-3e99-4b3a-bb2b-ce19dcf7794c",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: The number above is one higher than the actual read length, this is because of the newline charcater at the end of each line that counted with the wc command</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba7784-527e-461b-a3de-a72b79fc52c5",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Do we have enough data? \n",
    "\n",
    "A typical bacterial genome is around 5 million basepairs in length. However, this can range from XXX to XXX depending on the species in question.\n",
    "\n",
    "### What is depth of 'coverage'?\n",
    "\n",
    "### How do we determine our total throughput?\n",
    "\n",
    "INSERT FLASH CARD WITH QUESTIONS ABOUT READ LENGTH AND COVERAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1874fbae-7b77-49de-802f-f620729479b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6365120\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "zcat raw-reads/*_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab968490-19b8-4bf7-9582-5973da86492f",
   "metadata": {},
   "source": [
    "Whats our total bp of data? This is what we call our sequencing throughput. We multiple the number of reads by the read length (ours is 250) and by 2 because it is paired-end data.\n",
    "(Read length x 2(paired-end) x Number of reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "7f8ae91a-af8f-42fd-b26a-7a8385e01d70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of base pairs sequenced:\n",
      "1273024000\n",
      "Total coverage for genome dataset:\n",
      "254\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "num_reads=6365120\n",
    "read_length=100\n",
    "genome_size=5000000\n",
    "\n",
    "# calculate our total throughput\n",
    "echo \"Total number of base pairs sequenced:\"\n",
    "throughput=$(echo $num_reads \" * \" $read_length \" * 2\" | bc)\n",
    "echo $throughput\n",
    "\n",
    "# calculate our estimated coverage\n",
    "echo \"Total coverage for genome dataset:\"\n",
    "echo $throughput \" / \" $genome_size | bc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414ab82a-3f91-43d3-8c07-025f8a3379b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.5.2 Quality Assessment with FASTQC\n",
    "\n",
    "Program: FASTQC  \n",
    "Manual: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/  \n",
    "FASTQC explained:  \n",
    "Conda:  \n",
    "\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and determine some basic statistics about the data. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, it merely reads it and writes a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "506c05fd-a403-4eb3-bebb-7d0cfcc51246",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir output-fastqc\n",
    "\n",
    "# run the program on the two read files (R1 and R2).\n",
    "fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.\n",
    "ls output-fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfb2d6a-1a60-4b89-9898-1f5079d95962",
   "metadata": {},
   "outputs": [],
   "source": [
    "## view HTML report\n",
    "from IPython.display import IFrame\n",
    "IFrame('output-fastqc/SRR10056829_1_fastqc.html', width=1000, height=550)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb6207a-2b47-453f-a4d0-b38bff239dab",
   "metadata": {},
   "source": [
    "## 1.6 - Quality and adapter trimming with FASTP\n",
    "\n",
    "Program: fastp - an ultra-fast all-in-one FASTQ preprocessor\n",
    "Citation: *Chen, S. (2023). Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*\n",
    "Manual: https://github.com/OpenGene/fastp\n",
    "Conda: https://anaconda.org/bioconda/fastp\n",
    "\n",
    "Description:\n",
    "Ultrafast one-pass FASTQ data preprocessing, quality control, and deduplication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0df2498-4184-4403-b1dc-9cc93b98f955",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18047f7b-4838-4133-bf51-aa1d139caf2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# make a directory to store the output\n",
    "mkdir output-fastp\n",
    "\n",
    "# define input variables\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# run the program\n",
    "fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# the resulting folder should contain new trimmed read files\n",
    "ls output-fastp/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa89305-5b2d-4614-9cd2-c144224187bc",
   "metadata": {},
   "source": [
    "## 1.7 - Genome Assembly\n",
    "\n",
    "Program: SPAdes - St. Petersburg genome assembler\n",
    "Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*\n",
    "Manual: https://ablab.github.io/spades/\n",
    "Conda: https://anaconda.org/bioconda/spades\n",
    "\n",
    "\n",
    "Description: \n",
    "SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. However, it might not be suitable for large genomes projects.\n",
    "\n",
    "SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91bac579-1659-49c4-a60f-478f73ec5d67",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6681b8a5-b9a7-4924-bb8b-9961d14bd320",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "## Assemble trimmed reads using the SPAdes assembler.\n",
    "\n",
    "# define input\n",
    "forward=output-fastp/trimmed_1.fastq.gz\n",
    "reverse=output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# assemble reads with SPAdes\n",
    "spades.py -1 $forward -2 $reverse -o output-spades"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_pytorch_p310",
   "language": "python",
   "name": "conda_pytorch_p310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
