{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a251c4b-9e9c-4ce1-b715-a8f1ea7dc7cd",
   "metadata": {},
   "source": [
    "# Submodule 2: Assessment of genome assembly and genome annotation\n",
    "--------\n",
    "## Overview\n",
    "In this submodule, you will begin with the genome that you assembled in Submodule 1. The primary goal of this submodule is to assess the quality of the assembled genome through the lens of what we call the \"5 Cs\": Contiguity, Completeness, Contamination, Coverage, and Content. By utilizing a combination of bioinformatics tools, participants will evaluate the assembled genome and generate outputs that include visualizations, a cleaned genome sequence and functional annotations. These outputs wil will be used in submodule 4.\n",
    "\n",
    "\n",
    "### Learning Objectives\n",
    "Through this submodule, users will gain hands-on experience in quality assessment, resulting in a deeper understanding of genomic data integrity and the significance of accurate genome sequences.\n",
    "\n",
    "- **Understand and Apply the 5 Cs of Genome Quality**:  \n",
    "  Understand how to assess the overall quality of a genome sequence by examing Contiguity (QUAST), Completeness (BUSCO), Contamination (BLAST/BlobTools), Coverage (BWA/Samtools), and Content (Prokka gene annotations).\n",
    "\n",
    "- **Generate and Interpret Visualizations**:  \n",
    "  Gain proficiency in using bioinformatics tools and foster skills in data analysis and interpretation.\n",
    "\n",
    "- **Relate the Central Dogma of Molecular Biology to Genome Annotation**:  \n",
    "  Connect the principles of the central dogma (DNA → RNA → Protein) to the process of genome annotation, understanding how gene annotations contribute to functional genomics and biological interpretations.\n",
    "\n",
    "- **Produce a Clean and Annotated Genome**:  \n",
    "  Participants will refine the genome based on their assessments, ensuring a high-quality, annotated genome that can be used for further analysis or research applications."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "048beb3d-6133-4361-96c3-eb05e9c28135",
   "metadata": {},
   "source": [
    "## **Install required software**\n",
    "\n",
    "Several additional tools are required for Submodule 2; quast, busco, bwa, samtools, blast, blobtools, and prokka.  As with submodule 1, tools are preinstalled into a docker image, but we will demonstrate how to install these tools using __[Conda](https://docs.conda.io/en/latest/)__.\n",
    "\n",
    "Each piece of software, along with links to publications and documentation, will be described in turn. Below is a brief summary of these tools.\n",
    "\n",
    "### List of software\n",
    "| **Tool**       | **Description**                                                                                                                                                           |\n",
    "|:---------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| **QUAST**      | Used for evaluating and reporting the quality of genome assemblies by comparing them against reference genomes or generating statistical summaries.                          |\n",
    "| **BUSCO**      | Utilized for assessing genome completeness by searching for conserved single-copy orthologs from specific lineage datasets.                                                 |\n",
    "| **BWA**        | A fast and memory-efficient tool for aligning sequence reads to large reference genomes, commonly used in variant calling pipelines.                                         |\n",
    "| **Samtools**   | Used for manipulating and processing sequence alignments stored in SAM/BAM format. Essential for sorting, indexing, and viewing alignment files.                            |\n",
    "| **BLAST**      | A widely used tool for comparing an input sequence to a database of sequences, identifying regions of local similarity and aiding in functional annotation.                  |\n",
    "| **BlobTools**  | A versatile tool for visualizing and analyzing genome assemblies, helping to identify contamination or misassembled regions by correlating sequence features with taxonomy.   |\n",
    "| **Prokka**     | Used for rapid annotation of prokaryotic genomes, identifying genes, coding sequences, rRNAs, tRNAs, and other genomic features.                                             |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1719c82c-e638-42e4-a816-846fa52c646c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mamba install --channel bioconda     quast=5.2.0     busco=5.4.6     bwa=0.7.18     samtools=1.18     blast=2.15.0     blobtools=1.0.1     prokka=1.14.6     -y > /dev/null 2>&1\n",
      "Installation of quast, busco, bwa, samtools, blast, blobtools, and prokka complete.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Install all tools using mamba (a conda alternative) with specific versions\n",
    "\n",
    "echo \"mamba install --channel bioconda \\\n",
    "    quast=5.2.0 \\\n",
    "    busco=5.4.6 \\\n",
    "    bwa=0.7.18 \\\n",
    "    samtools=1.18 \\\n",
    "    blast=2.15.0 \\\n",
    "    blobtools=1.0.1 \\\n",
    "    prokka=1.14.6 \\\n",
    "    -y > /dev/null 2>&1\"\n",
    "\n",
    "echo \"Installation of quast, busco, bwa, samtools, blast, blobtools, and prokka complete.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d496898-5b34-480f-8702-e9f2a58139fb",
   "metadata": {},
   "source": [
    "## Starting Data\n",
    "\n",
    "This submodule starts with the **genome FASTA** file, this will be the primary input for all programs. We will define this as the variable *genome* now, and use that for the remainder of the workflow. This enables the starting data to be easily changed if a user wants to run this tutorial with their own data.\n",
    "\n",
    "We will also need to original reads from Submodule 1 for the read mapping step. We will use BWA to calculate sequencing coverage and it requires **paired-end sequencing reads in FASTQ format**. We will define these here as *forward* for the R1 sequencing reads and *reverse* for the R2 sequencing reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a93fc2f4-b579-4d18-9e3b-1a4a47026de6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# starting genome from submodule 1\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "# raw reads from submodule 1\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "638c8c9b-bd31-4b61-b2b4-c84822767ca4",
   "metadata": {},
   "source": [
    "## Process 1: **Contiguity** assessment using QUAST\n",
    "- Program: **QUAST (Quality Assessment Tool for Genome Assemblies)**\n",
    "- Citation: *Gurevich, A., Saveliev, V., Vyahhi, N., & Tesler, G. (2013). QUAST: quality assessment tool for genome assemblies. Bioinformatics, 29(8), 1072-1075.*\n",
    "- Manual: https://github.com/ablab/quast\n",
    "\n",
    "QUAST is a tool used to evaluate and compare the quality of genome assemblies by providing metrics such as N50, number of contigs, genome length, and misassemblies. Did you get one contig representing your entire genome? Or did you get thousands of contigs representing a highly fragmented genome?\n",
    "\n",
    "QUAST has many functionalities which we will not explore in this tutorial, I encourage you to explore these, for now we are going to use it in its simplest form. The program parses the genome FASTA file and records statsitics about each contig, the length, GC content etc. This type of information is something you would typically provide in a publication or as a way to assess different assemblers/options you may use. \n",
    "\n",
    "The **input** to the program is the genome assembly **FASTA** and the output are various tables and an html/pdf you can export and view."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "829a5245-9a06-4a98-8a00-cd1e0e3db39a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Python locale settings can't be changed\n",
      "/opt/conda/bin/quast.py assembled-genome/contigs.fasta\n",
      "\n",
      "Version: 5.2.0\n",
      "\n",
      "System information:\n",
      "  OS: Linux-5.10.230-223.885.amzn2.x86_64-x86_64-with-glibc2.35 (linux_64)\n",
      "  Python version: 3.10.14\n",
      "  CPUs number: 2\n",
      "\n",
      "Started: 2025-01-03 16:06:45\n",
      "\n",
      "Logging to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/quast.log\n",
      "NOTICE: Maximum number of threads is set to 1 (use --threads option to set it manually)\n",
      "\n",
      "CWD: /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis\n",
      "Main parameters: \n",
      "  MODE: default, threads: 1, min contig length: 500, min alignment length: 65, min alignment IDY: 95.0, \\\n",
      "  ambiguity: one, min local misassembly length: 200, min extensive misassembly length: 1000\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  assembled-genome/contigs.fasta ==> contigs\n",
      "\n",
      "2025-01-03 16:06:49\n",
      "Running Basic statistics processor...\n",
      "  Contig files: \n",
      "    contigs\n",
      "  Calculating N50 and L50...\n",
      "    contigs, N50 = 180742, L50 = 3, auN = 228778.9, Total length = 1693722, GC % = 30.29, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/Nx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/GC_content_plot.pdf\n",
      "  Drawing contigs GC content plot...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/contigs_GC_content_plot.pdf\n",
      "  Drawing Coverage histogram (bin size: 10x)...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/coverage_histogram.pdf\n",
      "  Drawing contigs coverage histogram (bin size: 10x)...\n",
      "    saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/basic_stats/contigs_coverage_histogram.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2025-01-03 16:06:52\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating PDF with all tables and plots...\n",
      "  2 of 2: Creating Icarus viewers...\n",
      "Done\n",
      "\n",
      "2025-01-03 16:06:53\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/report.html\n",
      "  PDF version (tables and plots) is saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/report.pdf\n",
      "  Icarus (contig browser) is saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/icarus.html\n",
      "  Log is saved to /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/quast_results/results_2025_01_03_16_06_45/quast.log\n",
      "\n",
      "Finished: 2025-01-03 16:06:53\n",
      "Elapsed time: 0:00:07.446101\n",
      "NOTICEs: 2; WARNINGs: 1; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "# run quast on the genome assembly\n",
    "quast.py $genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975d4b87-b3d8-4b91-acd5-2ee5a7c324aa",
   "metadata": {},
   "source": [
    "## Process 2: **Completeness** assessment using BUSCO\n",
    "\n",
    "- Program: **BUSCO - Benchmarking Universal Single-Copy Orthologs**\n",
    "- Citation: *Seppey, M., Manni, M., & Zdobnov, E. M. (2019). BUSCO: assessing genome assembly and annotation completeness. Gene prediction: methods and protocols, 227-245.*  \n",
    "- Manual: https://busco.ezlab.org/\n",
    "\n",
    "BUSCO is a program utilized to assess the completeness of a genome assembly in terms of the number of found and universal genes. This program makes use of the OrthoDB set of single-copy orthologous that are found in at least 90% of all the organisms in question. There are different data sets for various taxonomic groups (Eukaryotes, Metazoa, Bacteria, Gammaproteobacteria, etc. etc.). The idea is that a newly sequenced genome *should* contain most of these highly conserved genes. If your genome doesn't contain a large portion of these single-copy orthologs it may indicate that your genome is not complete.\n",
    "\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/busco_sampling.png\" width=\"40%\"/>\n",
    "</p>\n",
    "\n",
    "\n",
    "The input to the program is your genome assembly (contigs) as well as a selection of which database to use. The output is a directory with a short summary of the results, a full table with coordinates for each orthologous gene is located in your assembly, and a directory with the nucleotide and amino acid sequences of all the identified sequences.\n",
    "\n",
    "We will focus on the main summary output as a way to provide a simple QC assessment of our assembly, the outputs provided by BUSCO however have many uses, such as phylogenomics and gene prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "221ac192-7110-4b92-81f6-1ef79a029784",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-03 16:07:28 INFO:\tDownloading information on latest versions of BUSCO data...\n",
      "2025-01-03 16:07:30 INFO:\tDownloading file 'https://busco-data.ezlab.org/v5/data/information/lineages_list.2024-12-05.txt.tar.gz'\n",
      "2025-01-03 16:07:31 INFO:\tDecompressing file '/home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/busco_downloads/information/lineages_list.2024-12-05.txt.tar.gz'\n",
      "################################################\n",
      "\n",
      "Datasets available to be used with BUSCO v5.8.1 and above:\n",
      "\n",
      "- archaea_odb12\n",
      "    - euryarchaeota_odb12\n",
      "        - methanomicrobia_odb12\n",
      "            - methanosarcinaceae_odb12\n",
      "            - methanosarcina_odb12\n",
      "            - methanomicrobiales_odb12\n",
      "               - methanomicrobiaceae_odb12\n",
      "        - halobacteria_odb12\n",
      "            - halobacteriales_odb12\n",
      "                - halobacteriaceae_odb12\n",
      "                - haloarculaceae_odb12\n",
      "                    - haloarcula_odb12\n",
      "            - natrialbaceae_odb12\n",
      "                - natrinema_odb12\n",
      "                - haloterrigena_odb12\n",
      "            - haloferacales_odb12\n",
      "                - haloferacaceae_odb12\n",
      "                    - halorubrum_odb12\n",
      "                    - haloferax_odb12\n",
      "        - methanococcales_odb12\n",
      "            - methanococcaceae_odb12\n",
      "        - methanobacteria_odb12\n",
      "            - methanobrevibacter_odb12\n",
      "            - methanobacterium_odb12\n",
      "        - thermococcaceae_odb12\n",
      "            - thermococcus_odb12\n",
      "    - thermoplasmata_odb12\n",
      "    - nitrososphaerota_odb12\n",
      "        - nitrosopumilales_odb12\n",
      "            - nitrosopumilus_odb12\n",
      "    - thermoprotei_odb12\n",
      "        - sulfolobales_odb12\n",
      "        - thermoproteales_odb12\n",
      "            - thermoproteaceae_odb12\n",
      "        - desulfurococcales_odb12\n",
      "            - desulfurococcaceae_odb12\n",
      "- bacteria_odb12\n",
      "    - delta-epsilon_subdivisions_odb12\n",
      "        - deltaproteobacteria_odb12\n",
      "    - acidobacteriota_odb12\n",
      "        - acidobacteriaceae_odb12\n",
      "    - synergistota_odb12\n",
      "    - fusobacteriota_odb12\n",
      "        - fusobacteriaceae_odb12\n",
      "            - fusobacterium_odb12\n",
      "        - leptotrichia_odb12\n",
      "    - campylobacterota_odb12\n",
      "        - arcobacter_odb12\n",
      "    - spirochaetota_odb12\n",
      "        - leptospiraceae_odb12\n",
      "        - brachyspira_odb12\n",
      "        - spirochaetales_odb12\n",
      "            - treponema_odb12\n",
      "            - borreliaceae_odb12\n",
      "            - spirochaetaceae_odb12\n",
      "    - desulfuromonadales_odb12\n",
      "    - geobacteraceae_odb12\n",
      "        - geobacter_odb12\n",
      "    - desulfovibrionales_odb12\n",
      "    - desulfobacterales_odb12\n",
      "        - desulfobacteraceae_odb12\n",
      "    - thermotogota_odb12\n",
      "    - aquificae_odb12\n",
      "    - mycoplasmatota_odb12\n",
      "        - mycoplasmatales_odb12\n",
      "            - mycoplasma_odb12\n",
      "        - acholeplasmataceae_odb12\n",
      "            - phytoplasma_odb12\n",
      "        - entomoplasmatales_odb12\n",
      "            - spiroplasma_odb12\n",
      "    - actinomycetota_odb12\n",
      "        - coriobacteriia_odb12\n",
      "            - collinsella_odb12\n",
      "            - atopobiaceae_odb12\n",
      "                - olsenella_odb12\n",
      "            - eggerthellaceae_odb12\n",
      "        - actinomycetes_odb12\n",
      "            - frankia_odb12\n",
      "            - streptosporangiales_odb12\n",
      "                - nocardiopsaceae_odb12\n",
      "                    - nocardiopsis_odb12\n",
      "                - actinomadura_odb12\n",
      "            - kitasatosporales_odb12\n",
      "                - streptacidiphilus_odb12\n",
      "                - kitasatospora_odb12\n",
      "                - streptomyces_odb12\n",
      "            - pseudonocardiaceae_odb12\n",
      "                - pseudonocardia_odb12\n",
      "                - amycolatopsis_odb12\n",
      "            - propionibacteriales_odb12\n",
      "                - aeromicrobium_odb12\n",
      "                - nocardioides_odb12\n",
      "                - propionibacteriaceae_odb12\n",
      "            - micromonosporaceae_odb12\n",
      "                - actinoplanes_odb12\n",
      "            - mycobacteriales_odb12\n",
      "                - gordonia_odb12\n",
      "                - nocardiaceae_odb12\n",
      "                    - rhodococcus_odb12\n",
      "                    - nocardia_odb12\n",
      "                - mycobacteriaceae_odb12\n",
      "                    - mycolicibacterium_odb12\n",
      "                    - mycobacterium_odb12\n",
      "                - corynebacterium_odb12\n",
      "            - micrococcales_odb12\n",
      "                - microbacteriaceae_odb12\n",
      "                    - leucobacter_odb12\n",
      "                    - microbacterium_odb12\n",
      "                    - agromyces_odb12\n",
      "                    - curtobacterium_odb12\n",
      "                    - leifsonia_odb12\n",
      "                - intrasporangiaceae_odb12\n",
      "                - brachybacterium_odb12\n",
      "                - brevibacterium_odb12\n",
      "                - cellulomonadaceae_odb12\n",
      "                - micrococcaceae_odb12\n",
      "                    - kocuria_odb12\n",
      "                    - arthrobacter_odb12\n",
      "            - bifidobacteriaceae_odb12\n",
      "                - bifidobacterium_odb12\n",
      "            - actinomycetaceae_odb12\n",
      "                - actinomyces_odb12\n",
      "    - chloroflexota_odb12\n",
      "    - cyanobacteriota_odb12\n",
      "        - leptolyngbya_odb12\n",
      "        - synechococcales_odb12\n",
      "            - prochlorococcus_odb12\n",
      "            - synechococcaceae_odb12\n",
      "        - oscillatoriales_odb12\n",
      "        - chroococcales_odb12\n",
      "        - nostocales_odb12\n",
      "            - nostoc_odb12\n",
      "    - thermaceae_odb12\n",
      "        - meiothermus_odb12\n",
      "        - thermus_odb12\n",
      "    - deinococcus_odb12\n",
      "    - bacillota_odb12\n",
      "        - bacilli_odb12\n",
      "            - lactobacillales_odb12\n",
      "                - enterococcaceae_odb12\n",
      "                    - enterococcus_odb12\n",
      "                - lactobacillaceae_odb12\n",
      "                    - weissella_odb12\n",
      "                    - lactobacillus_odb12\n",
      "                    - pediococcus_odb12\n",
      "                - carnobacteriaceae_odb12\n",
      "                    - carnobacterium_odb12\n",
      "                - aerococcaceae_odb12\n",
      "                - streptococcaceae_odb12\n",
      "            - bacillales_odb12\n",
      "                - staphylococcaceae_odb12\n",
      "                    - salinicoccus_odb12\n",
      "                    - staphylococcus_odb12\n",
      "                - exiguobacterium_odb12\n",
      "                - alicyclobacillaceae_odb12\n",
      "                    - alicyclobacillus_odb12\n",
      "                - paenibacillaceae_odb12\n",
      "                    - brevibacillus_odb12\n",
      "                    - paenibacillus_odb12\n",
      "                - listeriaceae_odb12\n",
      "                    - listeria_odb12\n",
      "                - planococcaceae_odb12\n",
      "                    - sporosarcina_odb12\n",
      "                    - planococcus_odb12\n",
      "                - bacillaceae_odb12\n",
      "                    - virgibacillus_odb12\n",
      "                    - lysinibacillus_odb12\n",
      "                    - oceanobacillus_odb12\n",
      "                    - bacillus_odb12\n",
      "        - selenomonadales_odb12\n",
      "            - selenomonadaceae_odb12\n",
      "                - selenomonas_odb12\n",
      "        - veillonellaceae_odb12\n",
      "            - megasphaera_odb12\n",
      "            - veillonella_odb12\n",
      "        - erysipelotrichaceae_odb12\n",
      "        - clostridia_odb12\n",
      "            - thermoanaerobacterales_odb12\n",
      "            - eubacteriales_odb12\n",
      "                - clostridiaceae_odb12\n",
      "                - desulfosporosinus_odb12\n",
      "                - ruminococcus_odb12\n",
      "                - peptococcaceae_odb12\n",
      "                - eubacteriaceae_odb12\n",
      "                - peptostreptococcaceae_odb12\n",
      "                - lachnospiraceae_odb12\n",
      "                    - butyrivibrio_odb12\n",
      "                    - blautia_odb12\n",
      "                    - lachnoclostridium_odb12\n",
      "        - tissierellia_odb12\n",
      "            - anaerococcus_odb12\n",
      "            - peptoniphilus_odb12\n",
      "    - bacteroidota-chlorobiota_group_odb12\n",
      "        - cytophagia_odb12\n",
      "            - cytophagaceae_odb12\n",
      "            - cyclobacteriaceae_odb12\n",
      "                - algoriphagus_odb12\n",
      "            - hymenobacteraceae_odb12\n",
      "                - hymenobacter_odb12\n",
      "                - pontibacter_odb12\n",
      "        - bacteroidia_odb12\n",
      "            - bacteroidales_odb12\n",
      "                - bacteroides_odb12\n",
      "                - dysgonomonas_odb12\n",
      "                - prevotellaceae_odb12\n",
      "                - porphyromonadaceae_odb12\n",
      "                - rikenellaceae_odb12\n",
      "                    - alistipes_odb12\n",
      "        - chitinophagaceae_odb12\n",
      "        - sphingobacteriia_odb12\n",
      "            - pedobacter_odb12\n",
      "            - sphingobacterium_odb12\n",
      "        - flavobacteriia_odb12\n",
      "            - polaribacter_odb12\n",
      "            - nonlabens_odb12\n",
      "            - aquimarina_odb12\n",
      "            - maribacter_odb12\n",
      "            - flavobacterium_odb12\n",
      "            - tenacibaculum_odb12\n",
      "            - capnocytophaga_odb12\n",
      "            - chryseobacterium_odb12\n",
      "        - chlorobiota_odb12\n",
      "    - verrucomicrobiota_odb12\n",
      "    - chlamydiota_odb12\n",
      "        - chlamydia_odb12\n",
      "    - planctomycetota_odb12\n",
      "        - planctomycetaceae_odb12\n",
      "    - pseudomonadota_odb12\n",
      "        - betaproteobacteria_odb12\n",
      "            - burkholderiales_odb12\n",
      "                - comamonadaceae_odb12\n",
      "                    - variovorax_odb12\n",
      "                    - comamonas_odb12\n",
      "                    - acidovorax_odb12\n",
      "                - oxalobacteraceae_odb12\n",
      "                    - herbaspirillum_odb12\n",
      "                    - janthinobacterium_odb12\n",
      "                    - massilia_odb12\n",
      "                - alcaligenaceae_odb12\n",
      "                    - bordetella_odb12\n",
      "                    - achromobacter_odb12\n",
      "                - burkholderiaceae_odb12\n",
      "                    - pandoraea_odb12\n",
      "                    - ralstonia_odb12\n",
      "                    - burkholderia_odb12\n",
      "                    - paraburkholderia_odb12\n",
      "                    - cupriavidus_odb12\n",
      "            - nitrosomonadales_odb12\n",
      "                - methylophilaceae_odb12\n",
      "                - nitrosomonas_odb12\n",
      "            - thauera_odb12\n",
      "            - neisseriales_odb12\n",
      "                - neisseria_odb12\n",
      "                - chromobacteriaceae_odb12\n",
      "        - alphaproteobacteria_odb12\n",
      "            - rickettsiales_odb12\n",
      "                - anaplasmataceae_odb12\n",
      "                - rickettsiaceae_odb12\n",
      "            - pelagibacter_odb12\n",
      "            - hyphomicrobiales_odb12\n",
      "                - rhizobiaceae_odb12\n",
      "                    - sinorhizobium_odb12\n",
      "                    - rhizobium-agrobacterium_group_odb12\n",
      "                        - rhizobium_odb12\n",
      "                        - agrobacterium_odb12\n",
      "                - bartonella_odb12\n",
      "                - phyllobacteriaceae_odb12\n",
      "                    - mesorhizobium_odb12\n",
      "                - hyphomicrobiaceae_odb12\n",
      "                    - hyphomicrobium_odb12\n",
      "                - nitrobacteraceae_odb12\n",
      "                    - bradyrhizobium_odb12\n",
      "                    - rhodopseudomonas_odb12\n",
      "                    - afipia_odb12\n",
      "                - devosia_odb12\n",
      "                - bosea_odb12\n",
      "                - labrenzia_odb12\n",
      "                - aurantimonadaceae_odb12\n",
      "                    - aureimonas_odb12\n",
      "                - methylobacteriaceae_odb12\n",
      "                    - methylobacterium_odb12\n",
      "            - hyphomonadaceae_odb12\n",
      "                - hyphomonas_odb12\n",
      "            - caulobacteraceae_odb12\n",
      "                - asticcacaulis_odb12\n",
      "                - caulobacter_odb12\n",
      "                - brevundimonas_odb12\n",
      "            - sphingomonadales_odb12\n",
      "                - sphingopyxis_odb12\n",
      "                - novosphingobium_odb12\n",
      "                - sphingobium_odb12\n",
      "                - sphingomonas_odb12\n",
      "                - erythrobacteraceae_odb12\n",
      "                    - porphyrobacter_odb12\n",
      "                    - erythrobacter_odb12\n",
      "            - rhodobacterales_odb12\n",
      "                - paracoccaceae_odb12\n",
      "                    - paracoccus_odb12\n",
      "                    - rhodobacter_odb12\n",
      "                - ruegeria_odb12\n",
      "                - roseovarius_odb12\n",
      "                - sulfitobacter_odb12\n",
      "                - phaeobacter_odb12\n",
      "                - loktanella_odb12\n",
      "                - jannaschia_odb12\n",
      "            - rhodospirillales_odb12\n",
      "                - acetobacteraceae_odb12\n",
      "                    - acetobacter_odb12\n",
      "                - rhodospirillaceae_odb12\n",
      "                - azospirillum_odb12\n",
      "        - gammaproteobacteria_odb12\n",
      "            - enterobacterales_odb12\n",
      "                - enterobacteriaceae_odb12\n",
      "                    - enterobacter_odb12\n",
      "                    - citrobacter_odb12\n",
      "                - morganellaceae_odb12\n",
      "                - yersiniaceae_odb12\n",
      "                    - yersinia_odb12\n",
      "                - pectobacteriaceae_odb12\n",
      "                    - pectobacterium_odb12\n",
      "                - erwiniaceae_odb12\n",
      "                    - erwinia_odb12\n",
      "                    - pantoea_odb12\n",
      "                    - buchnera_odb12\n",
      "            - pseudomonadales_odb12\n",
      "                - marinobacter_odb12\n",
      "                - pseudomonas_odb12\n",
      "            - thiotrichales_odb12\n",
      "                - francisellaceae_odb12\n",
      "                - piscirickettsiaceae_odb12\n",
      "            - moraxellaceae_odb12\n",
      "                - psychrobacter_odb12\n",
      "                - moraxella_odb12\n",
      "                - acinetobacter_odb12\n",
      "            - cellvibrionales_odb12\n",
      "                - cellvibrionaceae_odb12\n",
      "            - pasteurellales_odb12\n",
      "            - aeromonadaceae_odb12\n",
      "                - aeromonas_odb12\n",
      "            - vibrionales_odb12\n",
      "                - vibrio_odb12\n",
      "                - photobacterium_odb12\n",
      "            - alteromonadales_odb12\n",
      "                - alteromonas_odb12\n",
      "                - psychromonas_odb12\n",
      "                - idiomarina_odb12\n",
      "                - shewanella_odb12\n",
      "                - colwellia_odb12\n",
      "                - pseudoalteromonas_odb12\n",
      "            - oceanospirillales_odb12\n",
      "                - halomonadaceae_odb12\n",
      "                    - halomonas_odb12\n",
      "                - alcanivorax_odb12\n",
      "                - oceanospirillaceae_odb12\n",
      "                    - marinobacterium_odb12\n",
      "                    - marinomonas_odb12\n",
      "            - methylococcaceae_odb12\n",
      "            - xanthomonadales_odb12\n",
      "                - pseudoxanthomonas_odb12\n",
      "                - luteimonas_odb12\n",
      "                - lysobacter_odb12\n",
      "                - stenotrophomonas_odb12\n",
      "                - xanthomonas_odb12\n",
      "                - rhodanobacteraceae_odb12\n",
      "            - chromatiales_odb12\n",
      "                - ectothiorhodospiraceae_odb12\n",
      "                    - thioalkalivibrio_odb12\n",
      "                - chromatiaceae_odb12\n",
      "            - legionellales_odb12\n",
      "                - legionellaceae_odb12\n",
      "- eukaryota_odb10\n",
      "     - alveolata_odb10\n",
      "         - apicomplexa_odb10\n",
      "             - aconoidasida_odb10\n",
      "                 - plasmodium_odb10\n",
      "             - coccidia_odb10\n",
      "     - euglenozoa_odb10\n",
      "     - fungi_odb10\n",
      "         - ascomycota_odb10\n",
      "             - dothideomycetes_odb10\n",
      "                 - capnodiales_odb10\n",
      "                 - pleosporales_odb10\n",
      "             - eurotiomycetes_odb10\n",
      "                 - chaetothyriales_odb10\n",
      "                 - eurotiales_odb10\n",
      "                 - onygenales_odb10\n",
      "             - leotiomycetes_odb10\n",
      "                 - helotiales_odb10\n",
      "             - saccharomycetes_odb10\n",
      "             - sordariomycetes_odb10\n",
      "                 - glomerellales_odb10\n",
      "                 - hypocreales_odb10\n",
      "         - basidiomycota_odb10\n",
      "             - agaricomycetes_odb10\n",
      "                 - agaricales_odb10\n",
      "                 - boletales_odb10\n",
      "                 - polyporales_odb10\n",
      "             - tremellomycetes_odb10\n",
      "         - microsporidia_odb10\n",
      "         - mucoromycota_odb10\n",
      "             - mucorales_odb10\n",
      "     - metazoa_odb10\n",
      "         - arthropoda_odb10\n",
      "             - arachnida_odb10\n",
      "             - insecta_odb10\n",
      "                 - endopterygota_odb10\n",
      "                     - diptera_odb10\n",
      "                     - hymenoptera_odb10\n",
      "                     - lepidoptera_odb10\n",
      "                 - hemiptera_odb10\n",
      "         - mollusca_odb10\n",
      "         - nematoda_odb10\n",
      "         - vertebrata_odb10\n",
      "             - actinopterygii_odb10\n",
      "                 - cyprinodontiformes_odb10\n",
      "             - tetrapoda_odb10\n",
      "                 - mammalia_odb10\n",
      "                     - eutheria_odb10\n",
      "                         - euarchontoglires_odb10\n",
      "                             - glires_odb10\n",
      "                             - primates_odb10\n",
      "                         - laurasiatheria_odb10\n",
      "                             - carnivora_odb10\n",
      "                             - cetartiodactyla_odb10\n",
      "                 - sauropsida_odb10\n",
      "                     - aves_odb10\n",
      "                         - passeriformes_odb10\n",
      "     - stramenopiles_odb10\n",
      "     - viridiplantae_odb10\n",
      "         - chlorophyta_odb10\n",
      "         - embryophyta_odb10\n",
      "             - liliopsida_odb10\n",
      "                 - poales_odb10\n",
      "             - eudicots_odb10\n",
      "                 - brassicales_odb10\n",
      "                 - fabales_odb10\n",
      "                 - solanales_odb10\n",
      "- viruses (no root dataset)\n",
      "     - alphaherpesvirinae_odb10\n",
      "     - baculoviridae_odb10\n",
      "     - rudiviridae_odb10\n",
      "     - betaherpesvirinae_odb10\n",
      "     - herpesviridae_odb10\n",
      "     - poxviridae_odb10\n",
      "     - tevenvirinae_odb10\n",
      "     - aviadenovirus_odb10\n",
      "     - enquatrovirus_odb10\n",
      "     - teseptimavirus_odb10\n",
      "     - bclasvirinae_odb10\n",
      "     - fromanvirus_odb10\n",
      "     - skunavirus_odb10\n",
      "     - betabaculovirus_odb10\n",
      "     - pahexavirus_odb10\n",
      "     - alphabaculovirus_odb10\n",
      "     - tunavirinae_odb10\n",
      "     - simplexvirus_odb10\n",
      "     - gammaherpesvirinae_odb10\n",
      "     - varicellovirus_odb10\n",
      "     - cheoctovirus_odb10\n",
      "     - guernseyvirinae_odb10\n",
      "     - tequatrovirus_odb10\n",
      "     - chordopoxvirinae_odb10\n",
      "     - peduovirus_odb10\n",
      "     - iridoviridae_odb10\n",
      "     - spounavirinae_odb10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# View available sets\n",
    "busco --list-datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2cbfe6ed-5542-456f-9abf-9c081ef5eefd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-03 16:09:27 INFO:\t***** Start a BUSCO v5.4.6 analysis, current time: 01/03/2025 16:09:27 *****\n",
      "2025-01-03 16:09:27 INFO:\tConfiguring BUSCO with local environment\n",
      "2025-01-03 16:09:27 INFO:\tMode is genome\n",
      "2025-01-03 16:09:27 INFO:\tDownloading information on latest versions of BUSCO data...\n",
      "2025-01-03 16:09:29 INFO:\tInput file is /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/assembled-genome/contigs.fasta\n",
      "2025-01-03 16:09:29 INFO:\tDownloading file 'https://busco-data.ezlab.org/v5/data/lineages/bacteria_odb10.2024-01-08.tar.gz'\n",
      "2025-01-03 16:09:31 INFO:\tDecompressing file '/home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/busco_downloads/lineages/bacteria_odb10.tar.gz'\n",
      "2025-01-03 16:09:33 INFO:\tRunning BUSCO using lineage dataset bacteria_odb10 (prokaryota, 2024-01-08)\n",
      "2025-01-03 16:09:33 INFO:\tRunning 1 job(s) on bbtools, starting at 01/03/2025 16:09:33\n",
      "2025-01-03 16:09:35 INFO:\t[bbtools]\t1 of 1 task(s) completed\n",
      "2025-01-03 16:09:35 INFO:\t***** Run Prodigal on input to predict and extract genes *****\n",
      "2025-01-03 16:09:35 INFO:\tRunning Prodigal with genetic code 11 in single mode\n",
      "2025-01-03 16:09:35 INFO:\tRunning 1 job(s) on prodigal, starting at 01/03/2025 16:09:35\n",
      "2025-01-03 16:09:41 INFO:\t[prodigal]\t1 of 1 task(s) completed\n",
      "2025-01-03 16:09:41 INFO:\tGenetic code 11 selected as optimal\n",
      "2025-01-03 16:09:41 INFO:\t***** Run HMMER on gene sequences *****\n",
      "2025-01-03 16:09:41 INFO:\tRunning 124 job(s) on hmmsearch, starting at 01/03/2025 16:09:41\n",
      "2025-01-03 16:09:43 INFO:\t[hmmsearch]\t13 of 124 task(s) completed\n",
      "2025-01-03 16:09:44 INFO:\t[hmmsearch]\t25 of 124 task(s) completed\n",
      "2025-01-03 16:09:45 INFO:\t[hmmsearch]\t38 of 124 task(s) completed\n",
      "2025-01-03 16:09:45 INFO:\t[hmmsearch]\t50 of 124 task(s) completed\n",
      "2025-01-03 16:09:46 INFO:\t[hmmsearch]\t63 of 124 task(s) completed\n",
      "2025-01-03 16:09:47 INFO:\t[hmmsearch]\t75 of 124 task(s) completed\n",
      "2025-01-03 16:09:47 INFO:\t[hmmsearch]\t87 of 124 task(s) completed\n",
      "2025-01-03 16:09:48 INFO:\t[hmmsearch]\t100 of 124 task(s) completed\n",
      "2025-01-03 16:09:50 INFO:\t[hmmsearch]\t112 of 124 task(s) completed\n",
      "2025-01-03 16:09:51 INFO:\t[hmmsearch]\t124 of 124 task(s) completed\n",
      "2025-01-03 16:09:52 INFO:\tResults:\tC:89.5%[S:89.5%,D:0.0%],F:4.8%,M:5.7%,n:124\t   \n",
      "\n",
      "2025-01-03 16:09:52 INFO:\t\n",
      "\n",
      "\t--------------------------------------------------\n",
      "\t|Results from dataset bacteria_odb10              |\n",
      "\t--------------------------------------------------\n",
      "\t|C:89.5%[S:89.5%,D:0.0%],F:4.8%,M:5.7%,n:124      |\n",
      "\t|111\tComplete BUSCOs (C)                       |\n",
      "\t|111\tComplete and single-copy BUSCOs (S)       |\n",
      "\t|0\tComplete and duplicated BUSCOs (D)        |\n",
      "\t|6\tFragmented BUSCOs (F)                     |\n",
      "\t|7\tMissing BUSCOs (M)                        |\n",
      "\t|124\tTotal BUSCO groups searched               |\n",
      "\t--------------------------------------------------\n",
      "2025-01-03 16:09:52 INFO:\tBUSCO analysis done. Total running time: 23 seconds\n",
      "2025-01-03 16:09:52 INFO:\tResults written in /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/output-busco\n",
      "2025-01-03 16:09:52 INFO:\tFor assistance with interpreting the results, please consult the userguide: https://busco.ezlab.org/busco_userguide.html\n",
      "\n",
      "2025-01-03 16:09:52 INFO:\tVisit this page https://gitlab.com/ezlab/busco#how-to-cite-busco to see how to cite BUSCO\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "# run BUSCO\n",
    "busco -i $genome -m genome -o output-busco -l bacteria --cpu 24"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe2dd37b-99b4-4e8e-b08c-43c99e9b809e",
   "metadata": {},
   "source": [
    "### Examine the BUSCO output.\n",
    "The first file we will look at is the 'short_summary_busco_output.txt'. This is a file which summarizes the main findings, how many of the expected genes did we find? This summary breaks the report into four main categories: **complete single-copy genes, complete duplicated genes, fragmented genes, and missing genes**. \n",
    "\n",
    "We are hopeful that the majority of our genes will be found as 'complete single-copy'. Duplicated genes could indicate that that particular gene underwent a gene duplication event or that we had a miss assembly and essentially have two copies of a region of our genome. Fragmented genes are an artifact of the fact that our genome did not assemble perfectly. Some of our genome is fragmented into multiple contigs, and with that some of our genes are going to be fragmented as well. This is why it is important to inspect the N50 of the genome with QUAST. We want the majority of our contigs to be at least as big as a gene, if it's not than we will have many fragmented genes as a result.\n",
    "\n",
    "Next we will view the 'full_table_busco_output.tsv' file. This is a file which shows the coordinates for all the associated single copy genes in our genome. It also provides information about the status of that ortholog (missing, complete, fragmented). This tsv file can be exported and viewed in excel.\n",
    "\n",
    "The final files we will examine are in a directory called 'single_copy_busco_sequences/'. This houses all the amino acid and protein sequences. This is a rich source for comparative genomics and other sorts of analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "880ef695-b48a-4f92-9248-4b929599e997",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Header:\n",
      "# Busco id\tStatus\tSequence\tGene Start\tGene End\tStrand\tScore\tLength\tOrthoDB url\tDescription\n",
      "#############################################\n",
      "Fragmented genes:\n",
      "841869at2\tFragmented\tNODE_2_length_184940_cov_88.881445\t72417\t73193\t-\t82.2\t144\thttps://v10-1.orthodb.org/?query=841869at2\tUDP-N-acetylenolpyruvoylglucosamine reductase\n",
      "1822215at2\tFragmented\tNODE_10_length_47326_cov_75.539443\t38096\t38521\t-\t130.2\t108\thttps://v10-1.orthodb.org/?query=1822215at2\tRibosomal protein L13\n",
      "1842956at2\tFragmented\tNODE_4_length_104950_cov_85.510082\t63946\t64599\t+\t24.7\t43\thttps://v10-1.orthodb.org/?query=1842956at2\tBiotin--acetyl-CoA-carboxylase ligase\n",
      "1990141at2\tFragmented\tNODE_1_length_512707_cov_68.489338\t117764\t118036\t+\t95.8\t63\thttps://v10-1.orthodb.org/?query=1990141at2\tRibosomal protein S15\n",
      "2040741at2\tFragmented\tNODE_2_length_184940_cov_88.881445\t59294\t59527\t+\t38.1\t42\thttps://v10-1.orthodb.org/?query=2040741at2\tRibosomal protein L24\n",
      "2063644at2\tFragmented\tNODE_2_length_184940_cov_88.881445\t58480\t58665\t+\t20.6\t16\thttps://v10-1.orthodb.org/?query=2063644at2\tRibosomal protein L29/L35\n",
      "#############################################\n",
      "Missing genes:\n",
      "95696at2\tMissing\n",
      "182107at2\tMissing\n",
      "874197at2\tMissing\n",
      "1036075at2\tMissing\n",
      "1491686at2\tMissing\n",
      "1758685at2\tMissing\n",
      "1937493at2\tMissing\n",
      "#############################################\n",
      "Duplicate genes:\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# examine the table (first ten lines only)\n",
    "echo \"Header:\"\n",
    "grep '# Busco id' output-busco/run_bacteria_odb10/full_table.tsv\n",
    "echo '#############################################'\n",
    "\n",
    "# see the categories of genes\n",
    "echo \"Fragmented genes:\"\n",
    "awk -F'\\t' '$2 == \"Fragmented\"' output-busco/run_bacteria_odb10/full_table.tsv\n",
    "\n",
    "echo '#############################################'\n",
    "echo \"Missing genes:\"\n",
    "awk -F'\\t' '$2 == \"Missing\"' output-busco/run_bacteria_odb10/full_table.tsv\n",
    "\n",
    "\n",
    "echo '#############################################'\n",
    "echo \"Duplicate genes:\"\n",
    "awk -F'\\t' '$2 == \"Duplicate\"' output-busco/run_bacteria_odb10/full_table.tsv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b999c3-d1c8-409b-a590-6a533947d91f",
   "metadata": {},
   "source": [
    "## Process 3: **Coverage** assessment using BWA\n",
    "\n",
    "- Program: **BWA** and **Samtools**\n",
    "- Citations: *Li, H., & Durbin, R. (2009). Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25(14), 1754–1760. https://doi.org/10.1093/bioinformatics/btp324*, *Li, H., et al. (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics, 25(16), 2078–2079. https://doi.org/10.1093/bioinformatics/btp352*\n",
    "- BWA manual: https://bio-bwa.sourceforge.net/bwa.shtml\n",
    "- SAMtools manual: http://www.htslib.org/doc/samtools-1.2.html\n",
    "- SAM format specifications: https://samtools.github.io/hts-specs/SAMv1.pdf\n",
    "\n",
    "Read Mapping refers to the process of aligning short reads to a reference sequence. This reference can be a complete genome, a transcriptome, or in our case de novo assembly. Read mapping is fundamental to many commonly used pipelines like differential expression or SNP analysis. We will be using it to calculate the average coverage of each of our contigs and to calculate the overall coverage of our genome (a requirement for genbank submission).The main output of read mapping is a **Sequence Alignment Map format (SAM)**. The file provides information about where our sequencing reads match to our assembly and information about how it maps. There are hundreds of programs that use SAM files as a primary input. A BAM file is the binary version of a SAM, and can be converted very easily using samtools.\n",
    "\n",
    "Many programs perform read mapping. The recommended program depends on what you are trying to do. My favorite is 'BWA mem' which balances performance and accuracy well. The input to the program is a referece assembly and reads to map (forward and reverse). The output is a SAM file. By default BWA writes the SAM file to standard output, I therefore save it directly to a file. There are lots of options, please see the manual to understand what I am using."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "37c7007c-b87e-4334-8de6-738c4182c138",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.01 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.40 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.01 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.01 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.24 sec\n",
      "[main] Version: 0.7.18-r1243-dirty\n",
      "[main] CMD: bwa index assembled-genome/contigs.fasta\n",
      "[main] Real time: 0.927 sec; CPU: 0.683 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 2400000 sequences (240000000 bp)...\n",
      "[M::process] read 782560 sequences (78256000 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (35, 1187092, 2, 39)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (98, 120, 203)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 413)\n",
      "[M::mem_pestat] mean and std.dev: (148.31, 75.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 518)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (156, 210, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 525)\n",
      "[M::mem_pestat] mean and std.dev: (221.48, 93.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 648)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (67, 110, 184)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 418)\n",
      "[M::mem_pestat] mean and std.dev: (123.62, 67.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 535)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2400000 reads in 120.844 CPU sec, 74.596 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 387048, 2, 14)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (155, 209, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 527)\n",
      "[M::mem_pestat] mean and std.dev: (221.45, 93.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 651)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (76, 238, 384)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1000)\n",
      "[M::mem_pestat] mean and std.dev: (214.00, 138.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1308)\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 782560 reads in 41.831 CPU sec, 26.657 real sec\n",
      "[main] Version: 0.7.18-r1243-dirty\n",
      "[main] CMD: bwa mem -t 24 assembled-genome/contigs.fasta raw-reads/SRR10056829_1.fastq.gz raw-reads/SRR10056829_2.fastq.gz\n",
      "[main] Real time: 109.565 sec; CPU: 168.411 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SQ\tSN:NODE_1_length_512707_cov_68.489338\tLN:512707\n",
      "@SQ\tSN:NODE_2_length_184940_cov_88.881445\tLN:184940\n",
      "@SQ\tSN:NODE_3_length_180742_cov_76.756479\tLN:180742\n",
      "@SQ\tSN:NODE_4_length_104950_cov_85.510082\tLN:104950\n",
      "@SQ\tSN:NODE_5_length_103244_cov_79.882449\tLN:103244\n",
      "@SQ\tSN:NODE_6_length_97797_cov_68.419922\tLN:97797\n",
      "@SQ\tSN:NODE_7_length_89725_cov_76.544173\tLN:89725\n",
      "@SQ\tSN:NODE_8_length_82897_cov_77.232382\tLN:82897\n",
      "@SQ\tSN:NODE_9_length_60299_cov_72.795349\tLN:60299\n",
      "@SQ\tSN:NODE_10_length_47326_cov_75.539443\tLN:47326\n",
      "@SQ\tSN:NODE_11_length_45856_cov_136.897164\tLN:45856\n",
      "@SQ\tSN:NODE_12_length_29079_cov_92.851640\tLN:29079\n",
      "@SQ\tSN:NODE_13_length_25629_cov_151.390162\tLN:25629\n",
      "@SQ\tSN:NODE_14_length_24523_cov_77.475764\tLN:24523\n",
      "@SQ\tSN:NODE_15_length_23101_cov_86.272846\tLN:23101\n",
      "@SQ\tSN:NODE_16_length_21883_cov_60.060808\tLN:21883\n",
      "@SQ\tSN:NODE_17_length_13932_cov_88.718743\tLN:13932\n",
      "@SQ\tSN:NODE_18_length_12703_cov_68.384567\tLN:12703\n",
      "@SQ\tSN:NODE_19_length_7107_cov_55.165060\tLN:7107\n",
      "@SQ\tSN:NODE_20_length_5811_cov_98.398193\tLN:5811\n",
      "@SQ\tSN:NODE_21_length_4145_cov_68.936430\tLN:4145\n",
      "@SQ\tSN:NODE_22_length_3290_cov_223.922720\tLN:3290\n",
      "@SQ\tSN:NODE_23_length_3166_cov_367.779492\tLN:3166\n",
      "@SQ\tSN:NODE_24_length_3108_cov_294.149034\tLN:3108\n",
      "@SQ\tSN:NODE_25_length_1510_cov_77.180069\tLN:1510\n",
      "@SQ\tSN:NODE_26_length_1195_cov_46.663158\tLN:1195\n",
      "@SQ\tSN:NODE_27_length_1019_cov_247.715768\tLN:1019\n",
      "@SQ\tSN:NODE_28_length_747_cov_61.440751\tLN:747\n",
      "@SQ\tSN:NODE_29_length_678_cov_163.922953\tLN:678\n",
      "@SQ\tSN:NODE_30_length_613_cov_388.458781\tLN:613\n",
      "@SQ\tSN:NODE_31_length_322_cov_82.385768\tLN:322\n",
      "@SQ\tSN:NODE_32_length_266_cov_1.066351\tLN:266\n",
      "@SQ\tSN:NODE_33_length_243_cov_0.957447\tLN:243\n",
      "@SQ\tSN:NODE_34_length_238_cov_0.983607\tLN:238\n",
      "@SQ\tSN:NODE_35_length_230_cov_0.851429\tLN:230\n",
      "@SQ\tSN:NODE_36_length_229_cov_1.034483\tLN:229\n",
      "@SQ\tSN:NODE_37_length_229_cov_0.655172\tLN:229\n",
      "@SQ\tSN:NODE_38_length_208_cov_124.045752\tLN:208\n",
      "@SQ\tSN:NODE_39_length_111_cov_8.017857\tLN:111\n",
      "@SQ\tSN:NODE_40_length_108_cov_127.698113\tLN:108\n",
      "@SQ\tSN:NODE_41_length_108_cov_6.226415\tLN:108\n",
      "@SQ\tSN:NODE_42_length_108_cov_5.264151\tLN:108\n",
      "@SQ\tSN:NODE_43_length_108_cov_3.886792\tLN:108\n",
      "@SQ\tSN:NODE_44_length_107_cov_1.096154\tLN:107\n",
      "@SQ\tSN:NODE_45_length_104_cov_4.040816\tLN:104\n",
      "@SQ\tSN:NODE_46_length_102_cov_107.106383\tLN:102\n",
      "@SQ\tSN:NODE_47_length_102_cov_1.808511\tLN:102\n",
      "@SQ\tSN:NODE_48_length_101_cov_4.000000\tLN:101\n",
      "@SQ\tSN:NODE_49_length_100_cov_2.000000\tLN:100\n",
      "@SQ\tSN:NODE_50_length_100_cov_1.377778\tLN:100\n",
      "@SQ\tSN:NODE_51_length_99_cov_6.568182\tLN:99\n",
      "@SQ\tSN:NODE_52_length_99_cov_4.977273\tLN:99\n",
      "@SQ\tSN:NODE_53_length_99_cov_2.340909\tLN:99\n",
      "@SQ\tSN:NODE_54_length_97_cov_8.738095\tLN:97\n",
      "@SQ\tSN:NODE_55_length_97_cov_4.476190\tLN:97\n",
      "@SQ\tSN:NODE_56_length_97_cov_4.119048\tLN:97\n",
      "@SQ\tSN:NODE_57_length_97_cov_2.500000\tLN:97\n",
      "@SQ\tSN:NODE_58_length_94_cov_3.128205\tLN:94\n",
      "@SQ\tSN:NODE_59_length_93_cov_5.342105\tLN:93\n",
      "@SQ\tSN:NODE_60_length_93_cov_2.763158\tLN:93\n",
      "@SQ\tSN:NODE_61_length_93_cov_2.000000\tLN:93\n",
      "@SQ\tSN:NODE_62_length_92_cov_32.810811\tLN:92\n",
      "@SQ\tSN:NODE_63_length_92_cov_27.351351\tLN:92\n",
      "@SQ\tSN:NODE_64_length_92_cov_2.216216\tLN:92\n",
      "@SQ\tSN:NODE_65_length_92_cov_1.243243\tLN:92\n",
      "@SQ\tSN:NODE_66_length_89_cov_8.705882\tLN:89\n",
      "@SQ\tSN:NODE_67_length_89_cov_5.911765\tLN:89\n",
      "@SQ\tSN:NODE_68_length_89_cov_4.852941\tLN:89\n",
      "@SQ\tSN:NODE_69_length_84_cov_102.241379\tLN:84\n",
      "@SQ\tSN:NODE_70_length_84_cov_5.551724\tLN:84\n",
      "@SQ\tSN:NODE_71_length_84_cov_1.655172\tLN:84\n",
      "@SQ\tSN:NODE_72_length_83_cov_1.000000\tLN:83\n",
      "@SQ\tSN:NODE_73_length_82_cov_8.888889\tLN:82\n",
      "@SQ\tSN:NODE_74_length_82_cov_1.518519\tLN:82\n",
      "@SQ\tSN:NODE_75_length_81_cov_1.230769\tLN:81\n",
      "@SQ\tSN:NODE_76_length_80_cov_6.400000\tLN:80\n",
      "@SQ\tSN:NODE_77_length_80_cov_4.800000\tLN:80\n",
      "@SQ\tSN:NODE_78_length_80_cov_4.560000\tLN:80\n",
      "@SQ\tSN:NODE_79_length_80_cov_4.200000\tLN:80\n",
      "@SQ\tSN:NODE_80_length_80_cov_4.000000\tLN:80\n",
      "@SQ\tSN:NODE_81_length_80_cov_3.800000\tLN:80\n",
      "@SQ\tSN:NODE_82_length_80_cov_1.800000\tLN:80\n",
      "@SQ\tSN:NODE_83_length_80_cov_1.800000\tLN:80\n",
      "@SQ\tSN:NODE_84_length_77_cov_3.772727\tLN:77\n",
      "@SQ\tSN:NODE_85_length_77_cov_2.954545\tLN:77\n",
      "@SQ\tSN:NODE_86_length_77_cov_2.318182\tLN:77\n",
      "@SQ\tSN:NODE_87_length_77_cov_1.909091\tLN:77\n",
      "@SQ\tSN:NODE_88_length_77_cov_1.727273\tLN:77\n",
      "@SQ\tSN:NODE_89_length_77_cov_1.000000\tLN:77\n",
      "@SQ\tSN:NODE_90_length_73_cov_9.722222\tLN:73\n",
      "@SQ\tSN:NODE_91_length_73_cov_1.666667\tLN:73\n",
      "@SQ\tSN:NODE_92_length_72_cov_25.000000\tLN:72\n",
      "@SQ\tSN:NODE_93_length_72_cov_9.352941\tLN:72\n",
      "@SQ\tSN:NODE_94_length_72_cov_4.117647\tLN:72\n",
      "@SQ\tSN:NODE_95_length_72_cov_1.352941\tLN:72\n",
      "@SQ\tSN:NODE_96_length_69_cov_6.714286\tLN:69\n",
      "@SQ\tSN:NODE_97_length_69_cov_1.928571\tLN:69\n",
      "@SQ\tSN:NODE_98_length_67_cov_9.916667\tLN:67\n",
      "@SQ\tSN:NODE_99_length_67_cov_1.500000\tLN:67\n",
      "@SQ\tSN:NODE_100_length_66_cov_11.000000\tLN:66\n",
      "@SQ\tSN:NODE_101_length_65_cov_14.100000\tLN:65\n",
      "@SQ\tSN:NODE_102_length_65_cov_3.100000\tLN:65\n",
      "@SQ\tSN:NODE_103_length_65_cov_1.200000\tLN:65\n",
      "@SQ\tSN:NODE_104_length_64_cov_61.888889\tLN:64\n",
      "@SQ\tSN:NODE_105_length_64_cov_14.000000\tLN:64\n",
      "@SQ\tSN:NODE_106_length_64_cov_9.555556\tLN:64\n",
      "@SQ\tSN:NODE_107_length_64_cov_5.888889\tLN:64\n",
      "@SQ\tSN:NODE_108_length_64_cov_5.000000\tLN:64\n",
      "@SQ\tSN:NODE_109_length_64_cov_5.000000\tLN:64\n",
      "@SQ\tSN:NODE_110_length_62_cov_20.714286\tLN:62\n",
      "@SQ\tSN:NODE_111_length_62_cov_3.000000\tLN:62\n",
      "@SQ\tSN:NODE_112_length_61_cov_2.000000\tLN:61\n",
      "@SQ\tSN:NODE_113_length_60_cov_69.000000\tLN:60\n",
      "@SQ\tSN:NODE_114_length_60_cov_54.200000\tLN:60\n",
      "@SQ\tSN:NODE_115_length_60_cov_23.600000\tLN:60\n",
      "@SQ\tSN:NODE_116_length_60_cov_17.000000\tLN:60\n",
      "@SQ\tSN:NODE_117_length_60_cov_15.800000\tLN:60\n",
      "@SQ\tSN:NODE_118_length_60_cov_6.400000\tLN:60\n",
      "@SQ\tSN:NODE_119_length_60_cov_3.600000\tLN:60\n",
      "@SQ\tSN:NODE_120_length_59_cov_3.000000\tLN:59\n",
      "@SQ\tSN:NODE_121_length_58_cov_6.000000\tLN:58\n",
      "@SQ\tSN:NODE_122_length_58_cov_4.333333\tLN:58\n",
      "@SQ\tSN:NODE_123_length_58_cov_2.000000\tLN:58\n",
      "@SQ\tSN:NODE_124_length_58_cov_1.666667\tLN:58\n",
      "@SQ\tSN:NODE_125_length_58_cov_1.000000\tLN:58\n",
      "@SQ\tSN:NODE_126_length_56_cov_77.000000\tLN:56\n",
      "@SQ\tSN:NODE_127_length_56_cov_20.000000\tLN:56\n",
      "@HD\tVN:1.5\tSO:unsorted\tGO:query\n",
      "@PG\tID:bwa\tPN:bwa\tVN:0.7.18-r1243-dirty\tCL:bwa mem -t 24 assembled-genome/contigs.fasta raw-reads/SRR10056829_1.fastq.gz raw-reads/SRR10056829_2.fastq.gz\n",
      "SRR10056829.1\t83\tNODE_7_length_89725_cov_76.544173\t49556\t60\t100M\t=\t49460\t-196\tTGCATTATTACCTATTTATGAGCGTTTAGGTATAAGAAAGACTTCAATGTTATTGATTATTACTTCGGCTATGGGGGTTATGAATGTTATTCCTTGGGGN\tFFFFFFFFBBFBFFFFFFFFFFFFFBFFFFFFFIIIIIIIIIIFIIIIIIIIFFFIIIIIIIIFFIFIIIIIIIIIFIIIIIIFFIIFFFFFFFFFF<0#\tNM:i:1\tMD:Z:99T0\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.1\t163\tNODE_7_length_89725_cov_76.544173\t49460\t60\t100M\t=\t49556\t196\tAGCTAATGTATATCAAGTTTGCATTCTTACTGTTTTTATTGCTGCTGCTGTGCATTTAGATGGTTCAGGTGCTTCTACTTTTCTTATTGTAATACCTGCA\tBBBFFFFFFFFFFIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIFFFIIIIIFFIFFIIIFFIIIIIIIIIIFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.2\t99\tNODE_17_length_13932_cov_88.718743\t6243\t60\t100M\t=\t6331\t188\tNTCATATGCCCTTAGAAGATATGACCAAACAAGATAGAATAAGAGCTACATATCAACATTGTTGTTTAAAATGGGCTTCAAGAGAAGCCATGACAAATGC\t#0<BBBFBFFFFBFFIFBFFFFIIFBFFFFIIIIFFBFFFFF0BFF<FFBBFFFIIIFIFF7B<77<F<BF<B7<<BFFFFFB<<<7<BBBBBBBBBBBB\tNM:i:1\tMD:Z:0C99\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.2\t147\tNODE_17_length_13932_cov_88.718743\t6331\t60\t100M\t=\t6243\t-188\tCATGACAAATGCTAGCTTAAGAGCTAGATTTAAAATAGGTGATAAGCAACATTCAAAAGCTTCAAAAATAATTAGCGATACTTTAGAGGCAGGTTTGATT\tBFFBBFBBB<<BBB<<BBBBBBFFFBFBFFBFFBFFB<<<IFFBIFFFBFFIFFFFFBFFIIFFBBFFFBFFFFIFFFFFFIFFFIFFFFFFFFBFF<BB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.3\t83\tNODE_6_length_97797_cov_68.419922\t31917\t60\t100M\t=\t31702\t-315\tTTAAATGTAGTTAATATCACGAAAAGAAGAGCTATTTCTGAGGATATGTGTAAAAGTGGTGTAGTGGGGATTAATGGCGTTATTGTTAAAGCGAGTAAAN\tFFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFB0#\tNM:i:1\tMD:Z:99G0\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.3\t163\tNODE_6_length_97797_cov_68.419922\t31702\t60\t100M\t=\t31917\t315\tCCCCAGAATACGCTAAAACTACTTTTTTTACTTCATTTTTCATAAGATATCCTTTGTTAAAATAGATTTGAAAATTGTAAAATTGTATCCAAATTTCATT\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.4\t83\tNODE_1_length_512707_cov_68.489338\t143612\t60\t100M\t=\t143504\t-208\tCTTGATTAAGATAAATTTTAACAAGATCTGAGGCATGCAACATCAATGAAGAAAGTGCAATTAATAAAGCAAGTCGTTTTAACAATGTTTTTCCTTAGGT\tBFBBBBFFFFFFBFFBFFFFFFFFFFIIIIIFIIIIIIIFFIIIIIFIFIIFIFFFFIIIIIIIIIIIIIIIIIIIIIIIIIFFFFIFFFFFFFBFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.4\t163\tNODE_1_length_512707_cov_68.489338\t143504\t60\t100M\t=\t143612\t208\tCATTATCATCATAGTATCCAAGTGAAATATTTTTATCTCCTATTTCACTTAACCAAAAATCCTTTTGAGTTAATTCTTTTTCAATCGCTACACCAACAGC\tBBBFFFFFFFBFFFFFFIIFIIIIIIIFIIIIIIFFFBFFFFIIIIIIIIIIIIIIIIIIFFIIFIFFFFIIFFIFFIIIIIFFFFFFFFFFFBFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.5\t99\tNODE_2_length_184940_cov_88.881445\t17062\t60\t100M\t=\t17107\t145\tCAACTATCGCACTCCCCACAAGCTTCATCTTCACTCTCATAACAAGACCAAGTAAGTTCTAAAGGTACATTTTCTTTTAAGGCTAGTTCTACAATTTTAG\tBBBFFFFFFFFFFIIFIFIIFFFFFFIIBFFFFIIIIFFFIIIIIIFFIIIIBFFIIFIIIIBFFFFIIFFFFFFFFFFFFFFFBBBB<BBBF<<BBFFB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.5\t147\tNODE_2_length_184940_cov_88.881445\t17107\t60\t100M\t=\t17062\t-145\tGACCAAGTAAGTTCTAAAGGTACATTTTCTTTTAAGGCTAGTTCTACAATTTTAGCTTTGTTTAGTCTAACAAGCGGAGTTTTTATACAAACTTTGAAAT\tBBFFFFBFBBBFFBBFFFFFB<B<FFFFBB<FFIIIIFFIFIIIFIFFBBFBFFBIIIIIFIIIIIFFIIIIIIFFFFBIIFIIIIFFFFFFFFFFFBB<\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.6\t83\tNODE_2_length_184940_cov_88.881445\t37824\t60\t100M\t=\t37761\t-163\tTTTTTAAAGCAATATTTTATAAAGGATTTAAGATGATGAGATCACTTTGGTCTGGCGTAAGCGGACTACAAGCACATCAAGTTGCGATGGATGTTGAAGG\tFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFIFBIIIIIIIIIIIIIIIIIIIIFIIFIIIIIIIIIIIIIIIIFIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.6\t163\tNODE_2_length_184940_cov_88.881445\t37761\t60\t100M\t=\t37824\t163\tGAACAGAACTTGCTTTTAAACTTCACATAAACGCAAAAGTTTTTTAAAGCCAAAGCGTTAAATTTTTTAAAGCAATATTTTATAAAGGATTTAAGATGAT\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIFFIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.7\t99\tNODE_1_length_512707_cov_68.489338\t102339\t60\t100M\t=\t102525\t287\tCTACCATAAATTTCAAGTTCATCTGTACCTATGCCTTTGATTTCAAGTCCTGCGTCTTTTAAAACTTCGCAAAGTTGAACTACTTCGGGTTCTTTTGCAA\tBBBFFFFFFFFFFIFIFFFFFIFFIIIIIIFFFIIIIFFFFIIFIIIFFFFIIIIFIIIIIFFIIIFIIIIIFFFFFFFFFFFFFBBFB7<BBFBBBBBB\tNM:i:0\tMD:Z:100\tMC:Z:67M1D33M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.7\t147\tNODE_1_length_512707_cov_68.489338\t102525\t60\t67M1D33M\t=\t102339\t-287\tATTTATTTGCCTTTTAATTTTCCACTAGCTCCGAATTAGCCTTGTTTGTTTTTTTTTTTTGCCCCCATTTTTCTAAGGCTAAAAGATGTAAATCAATAGG\t#########################################################FB'0FFFB'BFIIIIIIIIIIFIFIIIIFIFFFFFFFFFFBBB\tNM:i:11\tMD:Z:4C0A11A12A3C0A12A3G1A1A10^T33\tMC:Z:100M\tMQ:i:60\tAS:i:47\tXS:i:0\n",
      "SRR10056829.8\t99\tNODE_5_length_103244_cov_79.882449\t5513\t60\t100M\t=\t5645\t232\tGTATAATTTTTACTAAGATTACTTATTTTTAGTGCTTTAATATTATTCTTTATTGATATAATAAATTTGAGTTAATTATATTTTACTTATTGTTTTTTAG\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIBFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII<FFIIIIIIFFIIIIIIIIIIIFBFFFFFFB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.8\t147\tNODE_5_length_103244_cov_79.882449\t5645\t60\t100M\t=\t5513\t-232\tTGTTTTGAGCTCAGATTAATACTTCTGAGCCCATTTCGATCTCGTTGAGTTTTAAGGTATAAGCTTTTTATAGCTTTATTTTCCCTCCCATTTTTAACTC\tBBFFFFFFBBBFFFFFFFFB<BBB<BBBFFFBFFFFIFFBBBIIIIFFFIIIIIIIIFIIIIFIFIFIIIIFFIIIIIIIIFIIFFIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.9\t83\tNODE_23_length_3166_cov_367.779492\t571\t60\t2S98M\t=\t571\t-98\tAGCTCCGAAATATATTTAGGTATAGCGTTGTGTCGTAATATAAGGGGGTAGAGCACTGAATGGGCTAGGGCATACACCAATGTACCAAACCCTATCAAAC\tBB<<BFFFFFFFFFFBBFFFFFFFFFFFFFFFFIIIIIFIIIIIIIIIIIIIIIFFFIIFFFIIFFFIIFIFFFFFIIIFFFFFFBFFBFFFFFFFFBBB\tNM:i:0\tMD:Z:98\tMC:Z:98M2S\tMQ:i:60\tAS:i:98\tXS:i:0\n",
      "SRR10056829.9\t163\tNODE_23_length_3166_cov_367.779492\t571\t60\t98M2S\t=\t571\t98\tCTCCGAAATATATTTAGGTATAGCGTTGTGTCGTAATATAAGGGGGTAGAGCACTGAATGGGCTAGGGCATACACCAATGTACCAAACCCTATCAAACCT\tBBBFFFFFFFFFFIIIIIFFIIIIIFFIFFFFFFIIIIIIIIIIIIIIIIIIFFFFFFFFFFFBFFFFBFFFFFFFFFFF<BFFFFFFFFFFBFFFFFFF\tNM:i:0\tMD:Z:98\tMC:Z:2S98M\tMQ:i:60\tAS:i:98\tXS:i:0\n",
      "SRR10056829.10\t99\tNODE_7_length_89725_cov_76.544173\t36939\t60\t100M\t=\t37115\t276\tGTTCTAAATCCGCTATACTTCTTGAAACTTTTAAAATTTTATTAAGCGATCGCAAAGAAAGATTATAACGTGAAATTGCTAAATCTAGCGTATCTTTGGC\tBBBFFFFFFFFFFIIIFFFIFFIIIIIIIIIIIIIIIIIIIIIIIFFFFFIIIIIIIIIIIIIIIFFFFFFFFFFBBFFFFFFFFFFFFFBFFFFFBFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.10\t147\tNODE_7_length_89725_cov_76.544173\t37115\t60\t100M\t=\t36939\t-276\tTTACTAAAAATAAAAGCTTGTAAAATTTTTTCACTCATTTCTTTAGAACTTATACTCGTTTTATCATCCTTGCTGATCTCATCCATAGCTACATAAAGAT\tFFBFFFFFFFFFFFBFFFFFFFFFFFIIIIFFBFBFIIFFBIFIIIIIFIFFFBFFFIFIFFFFFFBFFIIFIIFFFIIFBFIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.11\t99\tNODE_2_length_184940_cov_88.881445\t14251\t60\t100M\t=\t14385\t234\tTTTAAATATAGATCCAAAAAGTGAGCTTGGACAAAGTGCTCAATATATTAATTCATTTATTTCCACCGTTCAAAACTCAATCATAGAAGCCATAGAAAAT\tBBBFFFFFFFFFFIIIIIIIIFFFIIIIIIIFIIIIFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFBFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.11\t147\tNODE_2_length_184940_cov_88.881445\t14385\t60\t100M\t=\t14251\t-234\tAAAACAACTCTATGATGCTTGAAAACAGTTCTAATTCGCAACATGAACAAATTCAAGGAGTTAAAGAAATCACTCATGTTTTAGATGATCATATCAATCT\tFFFFFFFFFFFFFFFFFFFFFFFFIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.12\t83\tNODE_2_length_184940_cov_88.881445\t100069\t60\t100M\t=\t99663\t-506\tGATATGCATCTTGGTTCGGTATTTTCGCCTGTAAGCCTTGGAATTATTCTAGGATTATTTTTAGGAAAGCAACTAGGTGTTTTTACTTTTTGTTTTATTN\tFFFFFFFFFFFFFFBBFFFFBBFFFFFFIFIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIFFFIFFFFFFFFFF<0#\tNM:i:1\tMD:Z:99G0\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.12\t163\tNODE_2_length_184940_cov_88.881445\t99663\t60\t100M\t=\t100069\t506\tGCTTTTGGGAAATAAAATTCCTACAAGTTTAAAACTTTTTTTACTCTCTTTAGCTATTTTTGATGATTTGGGTGCTATTGTTATTATTGCTTTGTTTTAT\tBBBFFFFFFFFFBFIFFFIFFIIIIFIFFIFFFIFIIIIIIIFFFI<FFFIFIIFFFIIIIIFFBFFFFFFF7BBBBFFFBBFFFBFFFBFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.13\t83\tNODE_1_length_512707_cov_68.489338\t91656\t60\t9S91M\t=\t91656\t-91\tAAGAGACAGAGTAAAGCTAGTATTTTACTCACACATCAACCTAAAACAGCTTTGCTTTATGATTTAAGTGATTTTGATCTTGTTTTAAGTGGGCATACAC\tFFFFFBFFFFFFFFFFFFFFFFFFFFIIIFFIFIIIIFFIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:91\tMC:Z:91M9S\tMQ:i:60\tAS:i:91\tXS:i:0\n",
      "SRR10056829.13\t163\tNODE_1_length_512707_cov_68.489338\t91656\t60\t91M9S\t=\t91656\t91\tAGTAAAGCTAGTATTTTACTCACACATCAACCTAAAACAGCTTTGCTTTATGATTTAAGTGATTTTGATCTTGTTTTAAGTGGGCATACACCTGTCTCTT\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIFFFIIIIIIFFIIIFIIIIIFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:91\tMC:Z:9S91M\tMQ:i:60\tAS:i:91\tXS:i:0\n",
      "SRR10056829.14\t99\tNODE_10_length_47326_cov_75.539443\t46708\t60\t78M22S\t=\t46708\t78\tATACCATAACGATCTGTTTCTGATTTAAATGGACGAATAAGCTGTGAATGACAAGCATTACAACTATCTTTAATATATCTGTCTCTTATACACATCTCCG\tBBBFFFFFFFFFFIIIFIIIIIIIIIIIIIIIIIIFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:78\tMC:Z:22S78M\tMQ:i:60\tAS:i:78\tXS:i:0\n",
      "SRR10056829.14\t147\tNODE_10_length_47326_cov_75.539443\t46708\t60\t22S78M\t=\t46708\t-78\tGTCAGATGTGTATAAGAGACAGATACCATAACGATCTGTTTCTGATTTAAATGGACGAATAAGCTGTGAATGACAAGCATTACAACTATCTTTAATATAT\tFFFFFFFFFFFFFFFFFFFFFIIIFIIIFIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:78\tMC:Z:78M22S\tMQ:i:60\tAS:i:78\tXS:i:0\n",
      "SRR10056829.15\t99\tNODE_15_length_23101_cov_86.272846\t12979\t60\t100M\t=\t13175\t296\tCACTTATTGCTTTAATTTTAAAAAAACTTAAGCAAAGGAAGGCAAAATGTTTAATGGAAAAAATATCTTAATCACGGGCGGAACAGGCTCGTTTGGAAAA\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBBF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.15\t147\tNODE_15_length_23101_cov_86.272846\t13175\t60\t100M\t=\t12979\t-296\tAATGCTCCTTGTATGCGTTATTTCATAGGAGATGTAAGAGATAAAGAGCGTTTAAGTGCTGCTATGCGTGATGTGGATTTTGTTATACATGCTGCTGCAA\tFFBFFBFFFFFFFBBBFFFBFFFFF<FFFFFFIFFIIFFBFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIFFFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.16\t99\tNODE_3_length_180742_cov_76.756479\t145518\t60\t100M\t=\t145764\t346\tTAATAGGTGTATCATCGCCTGGGAAATCATAAGAGCTTAATAATTCTCTAATTTCCATTTCAACTAACTCTAAAAGTTCAGCATCATCAACCATATCTGC\tBBBFFFBBFFFFFFFFFIIIIFIIIIIIIIIIFFFIIIIIIIIIIIFIIIIIIIIIIFFIIIFIIFBIIFIIFFFFFFFIFFFFFFFFBFFFFFFFBFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.16\t147\tNODE_3_length_180742_cov_76.756479\t145764\t60\t100M\t=\t145518\t-346\tCATAATCTGCGTGACCTGGGCAGTCAACGTGTGCATAGTGACGATTGTCTGTTTCATATTCAATATGAGAAGTAGCAATAGTAATACCACGCTCTTTTTC\tFFFFFFFFFBFFFFFFFBBFBFFFFFFFFIIFFFIIIFFFIIFFIIIFIIIIIFIFFFBFFFIBIIIIFIIIFIIIIIIIIFIIFFIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.17\t99\tNODE_4_length_104950_cov_85.510082\t78709\t60\t43M57S\t=\t78762\t153\tTATTAAAGATGGAGATAGGGTTCTTTTAGGGCTTAGTGGCGGGGAAAAATTTTTTGCTTTTGCCCCTCTTTTTAATAGTATGCAAACTCCCCCTCCCTTT\tBBBFFFFFFFFFFIIIIIIIFIIIIIIIIIIIIIIIFIIIIIB#########################################################\tNM:i:0\tMD:Z:43\tMC:Z:100M\tMQ:i:60\tAS:i:43\tXS:i:0\n",
      "SRR10056829.17\t147\tNODE_4_length_104950_cov_85.510082\t78762\t60\t100M\t=\t78709\t-153\tTGGCTTTAGCCCATCTTTTAAATCGTATGCAAGCTCACGCTCCTTTTAAATTTGAACTTGAAGCGGTAACTTTGAGTTATGGTATGGGCGAGGATTATTC\tFFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFFBIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:43M57S\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.18\t83\tNODE_1_length_512707_cov_68.489338\t375493\t60\t100M\t=\t375411\t-182\tTCGCTCACACTATATCGGACATAGTAAAAATACAGATATAAAAGCTTTAGAATTTCTAGAAAAAACTTACAAAATAAAACTATATTGCTTATGTCCTAAN\t<BBB<70<0'<BBB<B<B7BBBBFFFFFBBBBFBIIFFBFIFBFFFFBFFFFFFFBFFFBIFFFFBBFBFBBBBFBFFFFFF<BFFFBFFFBF<BFB<0#\tNM:i:1\tMD:Z:99C0\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.18\t163\tNODE_1_length_512707_cov_68.489338\t375411\t60\t100M\t=\t375493\t182\tATTTTTATCAAAATGGGTCATCTTATGCTTTTGATACCAAACAAGAAAATCTTTTAAAACTGGCTCCTGATTTTAAAAATGATCGCTCACACTATATCGG\tBBBFFFFFFFFFFIIIIFFFFFBBFFFFFBFFFFBFFIIIIIIFBFFFFFIIIFBFFFIBBBFF7BFFIBBFFFFFFFFFFBFFFBBBBBBBBBFBBBB<\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.19\t99\tNODE_1_length_512707_cov_68.489338\t448770\t60\t100M\t=\t448890\t220\tAGATCGAAATTAGAAAAATCCAAATCCGCATTTAAAAGTTTAGCATTTAAATTTCCAGAAACTAAAGCTCCATATCCTGCAAGAGCTAATAATTTTTCCA\tB<<FFFBFFFFFFFIBFFFFIFFIIIIIIFFFIFFFFFFIIFFFIIIIIIIIII<FFBFFFFIFFFFFFFFFFFFBBFFFFFFFFBBBBBBBFFFBBBBF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.19\t147\tNODE_1_length_512707_cov_68.489338\t448890\t60\t100M\t=\t448770\t-220\tGCAAAAATTTTATTATTTTTCAAACTTGCCTTTACCTCTCCACCAAGTCCGATAACTTGAGCATTTAAACTTGATAAGGCATTATTTACAACACTTGTTT\t<<BBBBBBBBBBBBBBBB<0<BBBB7F<BBB<70F<0FBF<0BBFFBBIFFBBBF0BBBFFIFFFFIFF<7BFBIFFFBFFBFFIFFBB<<<0FFBF<0B\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.20\t83\tNODE_3_length_180742_cov_76.756479\t91355\t60\t100M\t=\t91323\t-132\tTACCTAAATGAGTTTGTCCACCTTTATGAGGGGTGACACCTCCAACTATATTTGTACCATAAGCCATGCATTGCTCAGCGTGAAAAGTTGCTTCTTTACC\tFFFFFFFFFFFFBFFFFFFFFFFFFFFFFFIIIIIIIFIIIIIIIIFIFFIIIIFBIIIIIIIIIIIIIIIIIFFIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.20\t163\tNODE_3_length_180742_cov_76.756479\t91323\t60\t100M\t=\t91355\t132\tAACCGCATCTGCAACGGTATCAAAAACTGGTTTACCTAAATGAGTTTGTCCACCTTTATGAGGGGTGACACCTCCAACTATATTTGTACCATAAGCCATG\tBBBFFFFFFFFFFIIIIFFIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIFBFFBFFFFFFFFFFFFFFFFFFFFBFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.21\t83\tNODE_4_length_104950_cov_85.510082\t62805\t60\t100M\t=\t62533\t-372\tAATGGCTACAAGCAACCATAAAAACAGGTAGTGAGCACGGAACAGGTGGTATAGTAACCAAACTTAAGGCAGCTAAATTTTTACTTGAACACAATAAAAA\tFFFFFFFFFFFFFFFFFFFFFFFFFIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.21\t163\tNODE_4_length_104950_cov_85.510082\t62533\t60\t100M\t=\t62805\t372\tGGATTTTGACTCTAGAAAAGCTACTAAACACGCAAAAAATGCTATCGATATGATGATAAATTTAGGAATTTTGCCTATTATCAATGAAAATGACGCTACA\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIFIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.22\t99\tNODE_1_length_512707_cov_68.489338\t167687\t60\t100M\t=\t167736\t149\tACCTATTCCCAAGCATAAGGTAAAAAATGCAAGCCCTAAAGCTTCGGCAATAGAGCTGAATTTAAAATGTGAAAAATCGGGATAAAATAAATAAGAAAAA\tBBBFFFFFFFFFFIIIIIIIFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIFFFFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.22\t147\tNODE_1_length_512707_cov_68.489338\t167736\t60\t100M\t=\t167687\t-149\tATAGAGCTGAATTTAAAATGTGAAAAATCGGGATAAAATAAATAAGAAAAAGCATTTGCAAAACCTTGCTTAAATCCCATGCAATAAAACAACATAAAAA\tFFFFFFFFFFFFFFFFFFFFFFFFFIIIIIFFFBIIIIFIIIFIIFIIIIIIIIIIIIIIIFFIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.23\t99\tNODE_6_length_97797_cov_68.419922\t35305\t60\t100M\t=\t35585\t380\tAAGATCCTGAATTTAAAAAATATGTTTATGGGGATTCTAAAACACTTTTAGGTGTTAAACTCCCAAAAAGTAATTGGGTTGCAATGTGGATTTTCTTAGG\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIFFFFIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.23\t147\tNODE_6_length_97797_cov_68.419922\t35585\t60\t100M\t=\t35305\t-380\tTACTAAAACAGATGCAAAGAAAATCGGCTCTAATGAAATTTTTAAATCAGGTATGATAGCACTTGTTGCGGTATTTGGAATTTCTTGGATGGCTGATACT\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFIIIFIIIIIIIIIIIFFIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.24\t99\tNODE_12_length_29079_cov_92.851640\t10791\t60\t100M\t=\t10883\t192\tNATATCAGAAGCCCCTGCGCTAAGTGCAGATATCACAGGCCCCATAGTACGCCCACCTAGAAAATAATCCTCTGAATTTTTATTTTGCTTATAAAAATAA\t#0<FFFFFFFFFFIIIIIIIIIIIFFIIIIIIIIIIIIIIIIIFIIIFIIIIIIIIFFFFFFFFFFFFFFFFFFBFFFFFFBFFFFFFFFFFFFFFFFFF\tNM:i:1\tMD:Z:0C99\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n",
      "SRR10056829.24\t147\tNODE_12_length_29079_cov_92.851640\t10883\t60\t100M\t=\t10791\t-192\tAAAAATAAAAACCTATATAAAGCATTAATAAGGCATACACTACAAACATCACTGCTATTGGGGTATTAATTTGAACCACTTCCATCAATCTCTCCTTGAT\tFFFFFFFFFFBFFFFFFFFFFFFFFFIIFIIIIIFFFFFFBFIIFIIIFFFFFFIIIIIIIIIIIIIIIIIIIFFFFFFFIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.25\t83\tNODE_2_length_184940_cov_88.881445\t135899\t60\t100M\t=\t135807\t-192\tAAGCTACCAAAAGCCCTATAATCACTAAAATTTTAGAATGCGTATTTAAATTCAACATAGTAGTTTCTGCCATTCCTAACTAGGTATTGATCATTTATAC\tFFFFFBFFFFFFFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIFFFF0IIIIIIIIIIFFFFFFFFFFBBB\tNM:i:1\tMD:Z:76C23\tMC:Z:100M\tMQ:i:60\tAS:i:95\tXS:i:0\n",
      "SRR10056829.25\t163\tNODE_2_length_184940_cov_88.881445\t135807\t60\t100M\t=\t135899\t192\tAATTTCATCTTGGCTAAAGATATATGAAAAAAGTTCTTTTGGGTTTAAATTCTCATCTCCTAAACAAAGTGCTGTAAAAGAAGCAATTATATAAGCTACC\tBBBFFFFFFFFFFIIIIIIFIIIIIIIIIIIIIFFFFIIIIIIFIIIIIIIFIIIIFIIIIIIIIIIIIBFFFFFFFFFFFFFF0BBFFFFFFFFFFFFF\tNM:i:1\tMD:Z:84G15\tMC:Z:100M\tMQ:i:60\tAS:i:95\tXS:i:0\n",
      "SRR10056829.26\t83\tNODE_10_length_47326_cov_75.539443\t24899\t60\t100M\t=\t24813\t-186\tAAGAATATTACCGCAAACTTACAGGTAAAATAGCCTCAGATGGTGAAAATAACAATGTTGTAAATTCAAGCAATGAAACTCTTTATAATTCTGTTTATAG\t<FFFB<BB00'7'7<000'00B7'<BFBBFFBFF<IFF0FFBF<FBF<<<FBF<<7F7<FB0<BIIFB<07F<BBBBFBBFIIFB<B<BFFB0B<FFBB7\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.26\t163\tNODE_10_length_47326_cov_75.539443\t24813\t60\t100M\t=\t24899\t186\tTGACATGGCAAATAAAATCATCCAACTTCAATATGAAAAAGTTAATTTTTATAATGAAGATGGCACTATCGATAATCTTACTATGGAAGAATATTACCGC\t<B<BF<FF<<FBFFFIBBB<BFBBFBBBFBB<7BFFI0<<B<<BBFBBBB77BB7'BBB'BB'77<0BFFB'B<0BF<F<7B<<7B<7<B<<<B<<<B##\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.27\t99\tNODE_8_length_82897_cov_77.232382\t28528\t60\t100M\t=\t28615\t187\tAATACTCCCTTGTCCTTGCTAAGCCTCATCATATCATTAGCAACCTTATCTGCTTTCATAACATATTCCAAAGGAGTATTACAGATATTATAGCTTGTCC\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.27\t147\tNODE_8_length_82897_cov_77.232382\t28615\t60\t100M\t=\t28528\t-187\tTTATAGCTTGTCCCTTGATTAGTTTGAGTACAATCTACTAATCCAGAATTTGAAGCCTTAAGATTAAAATAAGGATAAAAATTAATACCCCAGCTATTAT\tFFFFFFFFFFFFFFFFFFFFIIIIIFIIIFIIIIIIFIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.28\t83\tNODE_1_length_512707_cov_68.489338\t459256\t60\t100M\t=\t459082\t-274\tGAAATGGAGCTAGAGGCGATTTAATCATTATTGTAAAAATAGAAGAAAGTCCTATTTACACAAGAGAAGATGATGATTTATATCAAAAGGTTGATATTTC\tFFFFFFFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.28\t163\tNODE_1_length_512707_cov_68.489338\t459082\t60\t100M\t=\t459256\t274\tGTTTTGAAGAAGAAATTTTAGACAGCAATCTAGAGCTTCACATTCCTTTCGAAAAAGCTGTTAAAGGCGGGGAACATAGTTTTAATTTTCAAGGCGAAAC\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIIIIFFFF<B<BBBB<BFFBFBFFF<<B<BBBFBB<\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.29\t99\tNODE_6_length_97797_cov_68.419922\t33460\t60\t78M22S\t=\t33460\t78\tATTTTATAAACATAATGTAAATTCGGCTTTTGTGGATAGTAAAGGACTTGCCAAAAAAAGTGTTTATGAGCTTTTAAACTGTCTCTTATACACATCTCCG\tBB<FFFFBBFFBF<FBF<BBBBFBB0BBFBFB0BFBBB0<BFBFBFBFF<F0FFIIFFBFBFBFFB<<07B7<B##########################\tNM:i:0\tMD:Z:78\tMC:Z:22S78M\tMQ:i:60\tAS:i:78\tXS:i:0\n",
      "SRR10056829.29\t147\tNODE_6_length_97797_cov_68.419922\t33460\t60\t22S78M\t=\t33460\t-78\tGTCAGATGTGTATAAGAGACAGATTTTATAAACATAATGTAAATTCGGCTTTTGTGGATAGTAAAGGACTTGCCAAAAAAAGTGTTTATGAGCTTTTAAA\tBBBFBFBFBFBFBBB<BBB7B<BBFFFBBBB<BBBFFFFBFB<BBFIFFBBBFBFFBBBIIIFFFFFFFFBB<'IFFBBBBFFFIFFFFFFFFFFFFBB<\tNM:i:0\tMD:Z:78\tMC:Z:78M22S\tMQ:i:60\tAS:i:78\tXS:i:0\n",
      "SRR10056829.30\t99\tNODE_1_length_512707_cov_68.489338\t255509\t60\t100M\t=\t255657\t248\tGCACATCAAATTTGCTTCTAATTTCTTTTAATTCATCGCGAATTAAATTTTCAAGCAAGGTTTCACTTTTTAAAATTTCTTCAAGTCTTGCAATTTCTTT\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIBFFIIIIIIFFFFFFFFFFFFFFFFFFFFFFBFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.30\t147\tNODE_1_length_512707_cov_68.489338\t255657\t60\t100M\t=\t255509\t-248\tTAAACGACCTAGTTTCATATCTAAAATAGCATTGGCTTGAAGCTCACTAAGACCAAATTTAGCTACCAAAGAATCTCTTGCGGTATTATTATCAGAACTA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFIIIIIIIFIIIIIIIFFFIIIFFBIIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.31\t99\tNODE_5_length_103244_cov_79.882449\t80634\t60\t100M\t=\t80681\t147\tTATTAGAAGAAAATAAAAAATGAATTCTGACCACCATAAAAATAAAAGCAAGAACTATGTAAATACAAGAAAAGAAGCATAAAAATATACCGCCAATATA\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFIIIIFIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.31\t147\tNODE_5_length_103244_cov_79.882449\t80681\t60\t100M\t=\t80634\t-147\tGCAAGAACTATGTAAATACAAAAAAAAAAGCAAAAAAATATACCGCCAATATAAAGAATTTCGTTCCCAGCTTCATATGAACTCGTCACCCCTTCATAAG\t#############BB<00'<7'FBB<'000'0'FFB<7'70'FFFIIIIFFFIIIFIIIIFIIFFIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFBBB\tNM:i:3\tMD:Z:21G4G5T67\tMC:Z:100M\tMQ:i:60\tAS:i:85\tXS:i:0\n",
      "SRR10056829.32\t99\tNODE_1_length_512707_cov_68.489338\t191281\t60\t100M\t=\t191435\t254\tGTTGAAAGAATCTCAAAACCTAAATTTTCAAGCTCTTTGCCAAATTCTACTATGCCTTCTTTATCACTCACACTAAGTAATGCTCTCATTTTTACCTTTC\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFIIFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.32\t147\tNODE_1_length_512707_cov_68.489338\t191435\t60\t100M\t=\t191281\t-254\tGCCGATTTTAAGCTCTTCTTCGCTAACATTGTTTGCATTTAAAATATCAGGATGGTATTTTTTTGCCAAATTGCGGTATTGTTTTTTTACCGCATTTAAA\tFFFBFFFFFFBFFBBFBBFB<BBBFBBBBB<FBBBFBFFBFFB<<BBFFFFFFFBFFFIIIIIFFIFIIFIFFFIFFFIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.33\t99\tNODE_9_length_60299_cov_72.795349\t33019\t60\t100M\t=\t33023\t104\tCTCTTTATATTTAGCACATAAAATGAACTTAGGTATGGATAAGTCTTGTTGGCCTGTTGCAGGAAGTTTTTATTTAACCAAACAAAAACTTTTAAATGGC\tBBBFFFFFFFFFFFFFFIFIIIFFIIIIFIFFIFFFFFFFFFFBBFFFFFIIIFIIFFIIIIIIIFFFFIIIIFFFFIFFFFFBFFFFFBFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.33\t147\tNODE_9_length_60299_cov_72.795349\t33023\t60\t100M\t=\t33019\t-104\tTTATATTTAGCACATAAAATGAACTTAGGTATGGATAAGTCTTGTTGGCCTGTTGCAGGAAGTTTTTATTTAACCAAACAAAAACTTTTAAATGGCAAGG\tFFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFIIIIIIIIFFIIFIIIIIIIIIIIIIIIIIIIIIIIIIIFFFIIIIFIIIIIFIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.34\t83\tNODE_1_length_512707_cov_68.489338\t395717\t60\t100M\t=\t395517\t-300\tAAAATCTTTTTATCCATACTCCTTGCATCAGCATGGGAATGATGATGTTCATGAGAATGTTCTTCGTGATTATGATGACAAGATTTGTTAGCTAAAGGCT\tFFFFFBFFFFFFFFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:55M2D45M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.34\t163\tNODE_1_length_512707_cov_68.489338\t395517\t60\t55M2D45M\t=\t395717\t300\tCTTCAAGACGAAAATACCCAAAGGTCTTTTGATGATCTTGCCATTTTTCTACAGCTAATGGCTAAAAAACTAAGAGCTAAGGCAAAAACATCAGAAAACA\tBBBFFFFFFFFFFIIIIIIIIIIIFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:2\tMD:Z:55^TA45\tMC:Z:100M\tMQ:i:60\tAS:i:92\tXS:i:0\n",
      "SRR10056829.35\t99\tNODE_1_length_512707_cov_68.489338\t301771\t60\t100M\t=\t301861\t190\tATACTTTACAGTGCTAATGCTTTAAATATAGCAAGCTGCACTATAGGTGGTTTTGATAAAGAAAAGCTTGATTCTTATTTATCACTTGATATTCAAAAAG\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFIFIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFFFFFFFFFFFFFF\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.35\t147\tNODE_1_length_512707_cov_68.489338\t301861\t60\t100M\t=\t301771\t-190\tATTCAAAAAGAAAGATCAAGTTTGGTAGTGGCTTTAGGATATTGCAACGATGAAAAAAATCCTCAAAAAAATCGTTTTAGTTTTGATGAAGTTGTAAAAT\tFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFBFFFFFFBFBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFFBBB\tNM:i:0\tMD:Z:100\tMC:Z:100M\tMQ:i:60\tAS:i:100\tXS:i:0\n",
      "SRR10056829.36\t83\tNODE_2_length_184940_cov_88.881445\t151997\t60\t100M\t=\t151876\t-221\tGCTTAGCACAAAAAAGCACAGCATCATTGATTTTTGCAGGACTTGCTAAAACGATTTTTTCACAACCTGCAATTTTTGCAGGAATAGCTAGCATTAAAAN\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIFFFFFFFFFF<0#\tNM:i:1\tMD:Z:99C0\tMC:Z:100M\tMQ:i:60\tAS:i:99\tXS:i:0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# starting genome from submodule 1\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "# raw reads from submodule 1\n",
    "forward=raw-reads/SRR10056829_1.fastq.gz\n",
    "reverse=raw-reads/SRR10056829_2.fastq.gz\n",
    "\n",
    "# Step 1: Index your reference genome. This is a requirement before read mapping.\n",
    "bwa index $genome\n",
    "# Step 2: Map the reads and construct a SAM file.\n",
    "bwa mem -t 24 $genome $forward $reverse > raw_mapped.sam\n",
    "# view the file with less, note that to see the data you have to scroll down past all the headers (@SQ).\n",
    "head -n 200 raw_mapped.sam | less -S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6757d9d4-53e5-43a9-b76c-a91b2b19e59f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 24 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3182761 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "3182560 + 0 primary\n",
      "0 + 0 secondary\n",
      "201 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "3177280 + 0 mapped (99.83% : N/A)\n",
      "3177079 + 0 primary mapped (99.83% : N/A)\n",
      "3182560 + 0 paired in sequencing\n",
      "1591280 + 0 read1\n",
      "1591280 + 0 read2\n",
      "3155510 + 0 properly paired (99.15% : N/A)\n",
      "3173626 + 0 with itself and mate mapped\n",
      "3453 + 0 singletons (0.11% : N/A)\n",
      "11772 + 0 with mate mapped to a different chr\n",
      "11528 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Remove sequencing reads that did not match to the assembly and convert the SAM to a BAM.\n",
    "samtools view -@ 24 -Sb  raw_mapped.sam  | samtools sort -@ 24 - -o sorted_mapped.bam\n",
    "\n",
    "# Examine how many reads mapped with samtools\n",
    "samtools flagstat sorted_mapped.bam\n",
    "# Calculate per base coverage with bedtools\n",
    "\n",
    "# index the new bam file\n",
    "samtools index sorted_mapped.bam\n",
    "\n",
    "#bedtools genomecov -ibam sorted_mapped.bam > coverage.out\n",
    "# Calculate per contig coverage with gen_input_table.py\n",
    "#gen_input_table.py  --isbedfiles $fasta coverage.out >  coverage_table.tsv\n",
    "# This outputs a simple file with two columns, the contig header and the average coverage."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06bd84a5-3d7a-4f1b-b824-e0a1332ed623",
   "metadata": {},
   "source": [
    "## Process 4 - Taxonomic assignment using BLAST and blobtools\n",
    "\n",
    "\n",
    "manual: https://www.ncbi.nlm.nih.gov/books/NBK279690/\n",
    "\n",
    "Using the command line BLAST works essentially the same as NCBI BLAST except we have more control. We can specify more options like output formats and also use our own local databases. It is also a lot more useful for pipelines and workflows since it can be automated, you don't need to open a web page and fill out any forms.\n",
    "\n",
    "As a quick example for how BLAST works we will use the same 16S_sequence and BLAST it against our genome assembly. Before we begin we will make a database out of our contig assembly. This is done to construct a set of files that BLAST can use to speed up its sequence lookup. In the end it means we have to wait less time for our results.\n",
    "\n",
    "#### Make a BLAST db from your contig files\n",
    "\n",
    "The only required input is a FASTA file (our contigs), the database type (nucl or prot), and an output name for the new database.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0d01f350-d0be-489d-b1b2-0b4d179a351c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/03/2025 16:37:27\n",
      "New DB name:   /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/contigs_db\n",
      "New DB title:  assembled-genome/contigs.fasta\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 127 sequences in 0.022356 seconds.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "makeblastdb -in $genome -dbtype nucl -out contigs_db "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696979b8-dd31-42a0-ac7f-92513b6c909c",
   "metadata": {},
   "source": [
    "### BLAST genome assembly against the nt database\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/nucleutide-blast-cover.png\" width=\"30%\"/>\n",
    "</p>\n",
    "\n",
    "\n",
    "We store a local copy of the complete nucleotide database on our server. We will be using this to provide a rough taxonomy to every sequence in our assembly and to ultimately identify non-target contaminates (like human and other bacteria) and to confirm our species identification from the 16S BLAST. Later we will be using the output file as in input to blobtools and to visualize this information. blobtools requires a specifically formatted BLAST file, I therefore provide a script that will run the BLAST to the programs specification. We will simply provide the script with our contigs file and it will complete the task. This is a simple script that is not much different than the example we ran above. It will automatically format a meaningfull output name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cd77df83-5089-4a97-bdb1-0117e686ded5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [blastn] Number of threads was reduced to 2 to match the number of available CPUs\n",
      "BLAST Database error: No alias or index file found for nucleotide database [data/ncbi_nt/nt] in search path [/home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis::]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "genome=assembled-genome/contigs.fasta\n",
    "database=data/ncbi_nt/refseq\n",
    "outname=blast_nt\n",
    "\n",
    "blastn \\\n",
    "    -task megablast \\\n",
    "    -query $genome \\\n",
    "    -db $database\\\n",
    "    -outfmt '6 qseqid staxids bitscore std sscinames sskingdoms stitle' \\\n",
    "    -culling_limit 5 \\\n",
    "    -max_target_seqs 10 \\\n",
    "    -num_threads 24 \\\n",
    "    -evalue 1e-5 \\\n",
    "    -out $outname.vs.nt.cul5.maxt10.1e5.megablast.out &\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06b74686-e191-49df-a773-4df1fe8e5da2",
   "metadata": {},
   "source": [
    "## Combine datasets into a blobtools database\n",
    "\n",
    "Program: **BlobTools**\n",
    "Citation: *Laetsch, D. R., & Blaxter, M. L. (2017). BlobTools: Interrogation of genome assemblies. F1000Research, 6, 1287*  \n",
    "Manual: https://blobtools.readme.io/docs  \n",
    "\n",
    "Blobtools is a tool to visualize our genome assembly. It is also useful for filtering read and assembly data sets. There are three main inputs to the program: 1.) Genome assembly **FASTA** file (the one we used for BLAST and BWA), 2.) a 'hits' file generated from **BLAST**, 3.) A SAM or **BAM** file. The main output of the program are blobplots which plot the GC, coverage, taxonomy, and contigs lengths on a single graph.  \n",
    "\n",
    "The first step (blobtools create) in this short pipeline takes all of our input files and creates a lookup table that is used for plotting and constructing tables. This step does the brunt of the working, parsing the BLAST file to assign taxonomy to each of our sequences, and parsing the SAM file to calculate coverage information.  \n",
    "\n",
    "After that is complete we will use 'blobtools view' to output all the data into a human readable table. Finally we will use 'blobtools plot' to construct the blobplot visuals.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5cfb06-247c-41a9-9b97-7ca4f7c1d790",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Create lookup table\n",
    "blobtools create --help\n",
    "blobtools create -i contigs.fasta -b sorted_mapped.bam -t contigs.fasta.vs.nt.cul5.1e5.megablast.out -o blob_out\n",
    "\n",
    "# Create output table\n",
    "blobtools view --help\n",
    "blobtools view -i blob_out.blobDB.json -r all -o blob_taxonomy\n",
    "\n",
    "# view the table, I remove headers with grep -v and view with tabview\n",
    "grep -v '##' blob_taxonomy.blob_out.blobDB.table.txt\n",
    "\n",
    "# Plot the data\n",
    "blobtools plot --help\n",
    "blobtools plot -i blob_out.blobDB.json -r genus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd578fc1-6777-4d18-af07-9763c8b40627",
   "metadata": {},
   "source": [
    "## Filter non-target sequences from *de novo* assembly\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/example_blobplot.png\" width=\"50%\"/>\n",
    "</p>\n",
    "\n",
    "The x-axis on these plots is GC content, the y-axis is the coverage (log transformed). The size of the 'blobs' are the length of the contigs. Colors represent taxonomic assignment (the -r option lets you choose which rank to view). The concept of these plots and ultimately for assembly filtering is that each organism has a unique GC content. For example Streptomyces has an average GC content of about 0.72 while other bacteria can go as low as 0.2. In addition, contamination is most likely has much lower coverage compared to the rest of your assembly. Combine that with the taxonomic assignments and you have multiple lines of evidence to identify your non-target contigs. In the plot above you can fairly easily see what contigs we plan to remove."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4d2f08-d2b1-4dc3-8649-8eb1ac77765f",
   "metadata": {},
   "source": [
    "## Genome annotation using Bakta\n",
    "\n",
    "Program: **Bakta** - Rapid & standardized annotation of bacterial genomes, MAGs & plasmids\n",
    "Citation: *Schwengers, O., Jelonek, L., Dieckmann, M. A., Beyvers, S., Blom, J., & Goesmann, A. (2021). Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial genomics, 7(11), 000685.*  \n",
    "Manual: https://github.com/oschwengers/bakta\n",
    "Annotation GFF3 file format: https://useast.ensembl.org/info/website/upload/gff3.html\n",
    "\n",
    "\n",
    "Bakta is a  tool for fast, taxon-independent, annotation of bacterial genomes. Annotation results are exported in **GFF3** and International Nucleotide Sequence Database Collaboration (INSDC)-compliant flat files that can be submitted with your genome to Genbank. Bakta is an alternative tool to the computationally demanding NCBI PGAP and highly customizable Prokka. We use it here for both its lightweight design (i.e. easy installation) and speed (~5 mins).\n",
    "\n",
    "Bakta annotates ncRNA cis-regulatory regions, oriC/oriV/oriT and assembly gaps as well as standard feature types: tRNA, tmRNA, rRNA, ncRNA genes, CRISPR, CDS and pseudogenes. Just like its inspiration (Prokka), bakta is a workflow that relies on other, more specific, annotation tools. We will focus the lesson on annotation to tRNAs, rRNAs, and protein-coding genes (PCGs). Here is a full list of the tools used by Brakta. \n",
    "\n",
    "- tRNAscan-SE (2.0.8) https://doi.org/10.1101/614032 http://lowelab.ucsc.edu/tRNAscan-SE\n",
    "- Aragorn (1.2.38) http://dx.doi.org/10.1093/nar/gkh152 http://130.235.244.92/ARAGORN\n",
    "- INFERNAL (1.1.4) https://dx.doi.org/10.1093%2Fbioinformatics%2Fbtt509 http://eddylab.org/infernal\n",
    "- PILER-CR (1.06) https://doi.org/10.1186/1471-2105-8-18 http://www.drive5.com/pilercr\n",
    "- Pyrodigal (2.1.0) https://doi.org/10.21105/joss.04296 https://github.com/althonos/pyrodigal\n",
    "- PyHMMER (0.10.0) https://doi.org/10.21105/joss.04296 https://github.com/althonos/pyhmmer\n",
    "- Diamond (2.0.14) https://doi.org/10.1038/nmeth.3176 https://github.com/bbuchfink/diamond\n",
    "- Blast+ (2.12.0) https://www.ncbi.nlm.nih.gov/pubmed/2231712 https://blast.ncbi.nlm.nih.gov\n",
    "- AMRFinderPlus (3.10.23) https://github.com/ncbi/amr\n",
    "- DeepSig (1.2.5) https://doi.org/10.1093/bioinformatics/btx818\n",
    "\n",
    "The program is simple to run, but does a lot. After you start the analysis in the next block of code, I suggest watching the following video about genome annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5240d4-28f6-4481-9727-f58c8b910b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "genome=assembled-genome/contigs.fasta\n",
    "\n",
    "# Run bakta with default options.\n",
    "#bakta --help\n",
    "bakta $genome -o output-bakta/ --db /home/sagemaker-user/Genome-Sequencing-and-Comparative-Genomic-Analysis/databases/db-light --threads 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b7ec592-0435-4037-b2a1-5a262d89d58b",
   "metadata": {},
   "source": [
    "## Visualize Dataset with Circos\n",
    "\n",
    "Program: **Circos** - pyCirclize is a circular visualization python package implemented based on matplotlib. This package is developed for the purpose of easily and beautifully plotting circular figure such as Circos Plot and Chord Diagram in Python.\n",
    "Manual: https://github.com/moshi4/pyCirclize\n",
    "\n",
    "We can use the Circos class in the pycirclize module to visualize our circular bacterial genome. Using some python code, we can visualize the genome contiguity, coverage, and content in one plot. This code uses the GFF3 and GBFF files from BAKTA to add annotations to the contig assembly. Using the per base coverage, we can see the coverage across contigs and look for any abnormally covered regions.\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/example_genome.png\" width=\"40%\"/>\n",
    "</p>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ddf8c1-92ae-4ada-a531-2f31aee7b430",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# get per base coverage\n",
    "bedtools genomecov -ibam sorted_mapped.bam -d > per_base_coverage.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "785c176e-b782-4ef2-b84e-621654a1547d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# We can call the script with the --help flag to learn how to run the program\n",
    "scripts/circos.py --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b2c59e0-811b-4fb8-8d83-60f4b3b040e7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: Most programs will have a built in help menu. Call the program with the flag <code>--help</code> or <code>-h</code> to learn how to run unfamiliar programs.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44e48c8e-1fe7-4821-9a31-8fd5a7878624",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "scripts/circos.py --species \"Enter your species here\" --gff output-bakta/contigs.gff3 --gbk output-bakta/contigs.gbff --bed per_base_coverage.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d312ee2-ffdb-43bc-ac2a-2e37e2e38db0",
   "metadata": {},
   "source": [
    "<p align=\"center\">\n",
    "  <img src=\"circos_plot.png\" width=\"80%\"/>\n",
    "</p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
