{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3054864-6f9a-44d5-b3a3-ff29ccde5e7c",
   "metadata": {},
   "source": [
    "# Submodule 1: Introduction to Genome Sequencing and Assembly\n",
    "--------\n",
    "## Overview\n",
    "In this submodule you will learn how DNA and RNA sequencing data is produced and how to reconstruct the genome of an organism from short DNA sequences generated by sequencing technologies. The process will be completed through an interactive markdown document and will use command-line bioinformatic tools. We will explore the key concepts and methodologies in genomics, including how to assemble a genome, how to assess its quality, how to annotate the genome, and how to perform comparative analyses.\n",
    "\n",
    "<p align=\"center\">\n",
    "  <img src=\"images/diagram-WGS.png\" width=\"50%\"/>\n",
    "</p>\n",
    "\n",
    "Figure 1. Overview of the bioinformatic workflow for the four submodules in this tutorial.\n",
    "### Learning Objectives:\n",
    "\n",
    "- **Understand how sequencing data is generated**: Learn the techniques and technologies used to generate genomic sequencing data, including next-generation sequencing (NGS) and its impact on genome assembly.\n",
    "\n",
    "- **Develop an understanding of core bioinformatic input/output formats in genomics**: Gain proficiency in handling common file formats like FASTQ, BAM, and GFF, and understand their role in processing, storing, and analyzing genomic sequencing data.\n",
    "\n",
    "- **Acquire the skills to assemble raw sequencing reads into a draft genome**: Master the process of genome assembly, from quality control of sequencing data to using assembly tools for constructing a draft genome from raw reads."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15fae599-ea0f-4376-8f97-f1409a2b64d6",
   "metadata": {},
   "source": [
    "## Background: How is sequencing data produced?\n",
    "\n",
    "### What is Next-Generation Sequencing (NGS)?\n",
    "\n",
    "Next-Generation Sequencing (NGS) is a high-throughput method that allows for rapid sequencing of DNA. Illumina remains the most widely used sequencing platform. It is accurate, inexpensive, and fast, providing up to 540 Gb of data in about a 1-2 days. Other long-read sequencing technologies (i.e. PacBio and Nanopore) are becoming widely used in the research community, and will be discussed in later bioinformatic modules.\n",
    "\n",
    "Figure 2 outlines the basic workflow for sequencing a microbial genome. It begins with the **isolation of bacterial cultures**, which involves culturing bacteria from a sample on selective media to obtain a pure strain. Once a pure culture is established, **DNA extraction** is performed to isolate the genomic DNA from the bacterial cells.\n",
    "\n",
    "For Illumina-based sequencing, DNA is **fragmented** into smaller pieces, and adapters and barcode indexes are added to each fragment to enable sequencing and sample multiplexing. This step is called **library preparation**. The prepared sequencing libraries are then attached to a specially coated surface on a microscope slide, forming dense clusters of identical DNA fragments ready for sequencing. In a process called **sequencing by synthesis**, fluorescently labeled nucleotides are incorporated into the clusters one at a time, using the complementary base pairing properties of DNA. Each incorporated nucleotide emits a distinct signal, allowing the sequencer to determine the DNA sequence in real time.\n",
    "\n",
    "\n",
    "This technology enables millions of fragments to be sequenced simultaneously, significantly increasing throughput and reducing costs.\n",
    "\n",
    "\n",
    "Isolate Bacteria            |  Extract DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src=\"https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg\" width=\"420\">\n",
    "\n",
    "Prepare Library           |  Sequence DNA\n",
    ":-------------------------:|:-------------------------:\n",
    "<img src=\"https://jef.works//assets/blog/librarystructure.png\" width=\"520\">  |  <img src=\"https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true\" width=\"320\">\n",
    "\n",
    "Figure 2. The process of sequencing an organisms DNA.\n",
    "\n",
    "\n",
    "## Video Lesson: How next-generation sequencing works.\n",
    "\n",
    "The following video describes the process of preparing an Illumina sequenicng library and how genome sequenicng with an Illumina sequencer works. We also compare and contrast this to other sequencing technologies (PacBio and Nanopore). Run the HTML code below to strat the video."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f56a4ae9-5484-4cd6-9bea-37dc5abc51d7",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: To execute the code you can either press <b>ctrl-ENTER</b> when the cell is selected or by pressing the play button at the top of the notebook</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee5c7c9-6c44-4da7-8df2-f7c5a40c634a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/p4vKJJlNTKA\" frameborder=\"0\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa7626f-3af1-4d7c-8601-4dce2708a4d6",
   "metadata": {},
   "source": [
    "## **Install required software**\n",
    "\n",
    "Four main tools are required for for Submodule 1: **fastq-dump, fastqc, fastp, and spades**. As described in the README, all tools are preinstalled into a docker image for this tutorial, but we will demonstrate how to install these tools using __[Conda](https://docs.conda.io/en/latest/)__, an open-source package management system and environment management tool. Conda helps to easily install, update, and manage software dependencies across different platforms. When we install the tools we select specific versions to ensure consistent behavior and reproducibility of results across different environments and systems.\n",
    "\n",
    "### List of software\n",
    "\n",
    "| **Tool**       | **Description**                                                                                                                                                           |\n",
    "|:---------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| **sra-tools**      | fastq-dump is used to retrieve publicly available data from the NCBI Sequence Read Archive (SRA).                                                 |\n",
    "| **fastqc**        | Used to assess the quality of raw sequencing data by generating visual reports on metrics like read length, GC content, and sequence duplication.|\n",
    "| **fastp**   | Used for quality control and pre-processing of FASTQ files, including read trimming, adapter removal, and filtering out low-quality reads.                          |\n",
    "| **spades**      | Used for genome assembly, creating contigs from high-quality, pre-processed sequencing reads.                  |  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f06ed5a-fa8c-464d-b60b-ac9307e4c1ed",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"> <b>Attention:</b> The code below is written in Bash. The <code>%%bash</code> at the beginning of the cell instructs the Jupyter notebook to execute the code as a Bash script when the cell is run. This is the same code you would run from a standard command-line terminal</a>. </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc1aadb-f7c1-4e1e-a7e4-77cc6e60141d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: To execute the code you can either press <b>ctrl-ENTER</b> when the cell is selected or by pressing the play button at the top of the notebook</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2fa5230-d4ee-47c1-90be-1cebea67de9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "echo -e \"\\033[1mNote: All command-line tools are preinstalled in a docker image\\033[0m\"\n",
    "echo -e \"\\033[1mNote: The command below provides an alternative means to aquire the needed tools using conda, but is not run by default\\033[0m\"\n",
    "echo \"------------------------------------------------------------------------------\"\n",
    "# Install tools using mamba (a conda alternative) with specified versions\n",
    "\n",
    "echo \"mamba install --channel bioconda \\\n",
    "    python=3.9 \\\n",
    "    sra-tools=3.1.1 \\\n",
    "    fastqc=0.11.8 \\\n",
    "    fastp=0.23.4 \\\n",
    "    spades=4.0.0 \\\n",
    "    -y > /dev/null 2>&1\"\n",
    "\n",
    "echo \"------------------------------------------------------------------------------\"\n",
    "echo -e \"\\033[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\\nDisplaying versions:\\033[0m\"\n",
    "\n",
    "# Confirm installation by checking versions\n",
    "fastq-dump --version\n",
    "fastqc --version\n",
    "fastp --version\n",
    "spades.py --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b09b30f-d22a-415f-8dd2-b3338fd07a9f",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>\\</code> to break a long command into multiple lines</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75252e5c-4e8d-4d97-987b-3111db2008d2",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: use <code>/dev/null 2>&1</code> to write standard error and standard output to <i>/dev/null</i>, effectively discarding them so your screen doesn't get flooded with text.</div></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e6bd27d-90a9-4fac-98f5-ce003a5c097c",
   "metadata": {},
   "source": [
    "## **Starting data**\n",
    "The data used for this module is described in a manuscript comparing phenotypic and whole-genome sequencing derived AMR profiles ([Painset et al. 2020](https://pubmed.ncbi.nlm.nih.gov/31943013/)) and is available under the SRA accession [SRR10056829](https://www.ncbi.nlm.nih.gov/sra/?term=PRJNA505131). Genomic DNA was extracted, fragmented and prepared for multiplexing using a **Nextera XT DNA Sample Preperation Kits**. Paired-end sequencing data was produced on an **Illumina HiSeq 2500** platform.\n",
    "\n",
    "The manuscript includes sequencing read datasets for 528 isolates of *Campylobacter* spp. (452 *C. jejuni* and 76 *C. coli*) from human (494), food (21) and environmental (2) sources. We will start by walking through the process on just a single dataset. We will later expand this to include a larger set of this data in submodule 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e1df077-ccb2-47db-8769-55530eb209f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Capture the SRA accession in a variable \n",
    "accession=SRR10056829\n",
    "\n",
    "# Download data from the SRA using the variable from above \n",
    "\n",
    "# prefetching downloads metadata of the SRA records (5.629s)\n",
    "echo \"Prefetching metadata from the NCBI SRA\"\n",
    "prefetch -v $accession > /dev/null 2>&1\n",
    "\n",
    "# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)\n",
    "echo \"Downloading the reads and compressing them using standard gzip compression\"\n",
    "fastq-dump --outdir raw-reads --gzip --split-files \"$accession\"/\"$accession\".sra > /dev/null 2>&1\n",
    "\n",
    "# Remove the prefetch directory\n",
    "rm -r $accession\n",
    "\n",
    "# rename files so this tutorial works with any data\n",
    "mv raw-reads/SRR10056829_1.fastq.gz raw-reads/reads_1.fastq.gz\n",
    "mv raw-reads/SRR10056829_2.fastq.gz raw-reads/reads_2.fastq.gz\n",
    "echo \"Process Complete\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2be5535-b923-4ab5-91d8-3e8dca3c833b",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: Assigning certain values to variables (like the accession above) allows you to easily adjust input to a set of commands and reduces the amount of typing for long or complicated file names</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1b1056-ae52-4290-a2a0-615d06512281",
   "metadata": {},
   "source": [
    "## FASTQ file format\n",
    "The data we just downloaded are in **FASTQ** file format. Let's examine this data and file format.\n",
    "\n",
    "Let's run three sections of code to learn about the FASTQ file format. Please also see a detailed description of FASTQ files in the NIGMS module [Fundementals of Bionformatics](https://github.com/NIGMS/Fundamentals-of-Bioinformatics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea85ebdd-d656-468a-95e0-89426db38912",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# If the above code worked you should have two FASTQ files in a directory called raw-reads\n",
    "ls raw-reads/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e89bc5-b10d-4375-b05a-4fe6390528fe",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"> <b>Attention:</b> Notice the the files above have the .gz extension. FASTQ data is commonly gzip compressed to save sapce. Most bioinformatic tools accept or require gzipped FASTQs. </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ccb32ad-c885-4644-9d7f-fe9d2f34417d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# define variable names for reads\n",
    "forward=raw-reads/reads_1.fastq.gz\n",
    "reverse=raw-reads/reads_2.fastq.gz\n",
    "\n",
    "# view the first four lines\n",
    "echo \"-------- FASTQ READ ---------\"\n",
    "zcat $forward | head -n 4\n",
    "echo \"-----------------------------\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2ae8ac-0fb2-4123-afec-0f37f76058fc",
   "metadata": {},
   "source": [
    "### Explanation:\n",
    "The above code should have displayed the first four lines of the FASTQ file. **Each sequencing read in a FASTQ formatted file is made up of four lines**. \n",
    "\n",
    "The length of the sequencing reads can vary depending on the type of sequencing instrument and flow cell used, but they should be consistent across the raw data. In our example dataset the reads are 100 bps in length. Illumina sequencers produce reads with length lengths randing between 50 and 300 bps. Run the following code to print the length of the reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5120eb17-44a0-44a5-86ab-68b2b7140825",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "zcat raw-reads/reads_1.fastq.gz | head -n 2 | tail -n 1 | wc -c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c457802-3e99-4b3a-bb2b-ce19dcf7794c",
   "metadata": {},
   "source": [
    "<div class= \"alert alert-block alert-info\"><b>Tip</b>: The number above is one higher than the actual read length, this is because of the newline charcater at the end of each line that counted with the wc command</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95ba7784-527e-461b-a3de-a72b79fc52c5",
   "metadata": {},
   "source": [
    "## Do we have enough data? \n",
    "\n",
    "Most bacterial genomes consist of a single, circular chromosome, although some bacteria may have linear chromosomes or multiple chromosomes. In addition to the main chromosome, plasmids (small circular DNA molecules) often coexist, carrying additional genes. Compared to eukaryotes, bacterial genomes are typically much smaller, ranging from approximately 0.5 to 10 megabases (Mb) in size. For instance, Escherichia coli has a genome size of about 4.6 Mb. This genome can be represented as a string of 4.6 million nucleotides—comprising \"A\"s, \"T\"s, \"C\"s, and \"G\"s.\n",
    "\n",
    "| Organism                                  | Genome Size (Mb) |\n",
    "|------------------------------------------|------------------|\n",
    "| *Mycoplasma genitalium* (Bacteria)       | 0.58             |\n",
    "| *Escherichia coli* (Bacteria)            | 4.6              |\n",
    "| *Saccharomyces cerevisiae* (Yeast)       | 12.1             |\n",
    "| *Streptomyces coelicolor* (Bacteria)     | 8.7              |\n",
    "| *Caenorhabditis elegans* (Nematode)      | 100              |\n",
    "| *Arabidopsis thaliana* (Thale Cress)     | 135              |\n",
    "| *Drosophila melanogaster* (Fruit Fly)    | 139              |\n",
    "| *Zea mays* (Maize)                        | 2,300            |\n",
    "| *Mus musculus* (Mouse)                   | 2,700            |\n",
    "| *Homo sapiens* (Human)                   | 3,200            |\n",
    "\n",
    "\n",
    "### What is sequencing 'coverage'?\n",
    "**Breadth of coverage** refers to the percentage of the genome that has been sequenced at least once. It tells you how much of the genome has any sequencing data aligned to it, but it doesn’t indicate how many times each base has been sequenced. High breadth of coverage means most or all regions of the genome are represented in the sequencing data.\n",
    "\n",
    "**Depth of coverage**, on the other hand, refers to the average number of times each base in the genome is sequenced, often represented as \"X\" (e.g., 30x depth). Depth is important for accuracy, as higher depth ensures that individual bases are sequenced multiple times, improving confidence in variant calling and reducing sequencing errors.\n",
    "\n",
    "When researchers say *coverage* they usually mean *depth of coverage*.\n",
    "\n",
    "\n",
    "The equation for genome depth of coverage is:\n",
    "\n",
    "$$\n",
    "\\text{Coverage} = \\frac{N \\times L}{G}\n",
    "$$\n",
    "\n",
    "Where:\n",
    "- \\( N \\) = Number of reads\n",
    "- \\( L \\) = Read length\n",
    "- \\( G \\) = Genome size\n",
    "\n",
    "\n",
    "The total number of bases sequenced, the numerator in the above equation, is also called the total **throughput** for the sample.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d45fee-1511-49e5-9621-4ac218901c6b",
   "metadata": {},
   "source": [
    "### How can we determine our total throughput and coverage?\n",
    "\n",
    "Lets do the math based on the data we have so far, we know the read length, and we can use a best guess at the genome size as 5 Mb for now. So lets count our number of reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1874fbae-7b77-49de-802f-f620729479b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.\n",
    "echo \"The total number of lines in the fastq file:\"\n",
    "zcat raw-reads/*_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f8ae91a-af8f-42fd-b26a-7a8385e01d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)\n",
    "num_reads=6365120/4\n",
    "read_length=100\n",
    "# adjust this number to the organisms genome size.\n",
    "genome_size=5000000\n",
    "\n",
    "# calculate our total throughput\n",
    "echo \"Total number of base pairs sequenced:\"\n",
    "#throughput=$(echo $num_reads \" * \" $read_length \" * 2\" | bc)\n",
    "throughput=$(awk \"BEGIN {print $num_reads * $read_length * 2}\")\n",
    "echo $throughput\n",
    "\n",
    "# calculate our estimated coverage\n",
    "echo \"The estimated depth of coverage for this genome dataset:\"\n",
    "#coverage=$($throughput \" / \" $genome_size | bc)\n",
    "coverage=$(awk \"BEGIN {print $throughput / $genome_size}\")\n",
    "echo $coverage\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e94047a-bdd8-4161-8e74-97cd4a44bc09",
   "metadata": {},
   "source": [
    "### Explanation\n",
    "\n",
    "Our estimated coverage, assuming a 5 Mb genome, is 63X. This means we are estimated to have sequenced every base in the genome approximately 63 times.\n",
    "\n",
    "**Coverage and De Novo Assembly:**\n",
    "For de novo genome assembly, coverage is a critical factor. Typically, **30X to 50X coverage** is recommended for most bacterial genomes, as it provides enough data to accurately assemble the genome without over-assembling, which could lead to potential issues. However, the required coverage can vary depending on factors such as the organism’s genome size, complexity, and the sequencing platform used.\n",
    "\n",
    "- **Bacterial genomes**: 30X to 50X is generally sufficient.\n",
    "- **Eukaryotic genomes**: Larger genomes may require higher coverage, often in the range of 50X to 100X or more, depending on the complexity and repetitive regions.\n",
    "\n",
    "Higher coverage helps ensure the assembly is accurate and complete, reducing the chances of gaps or errors. However, after a certain point, increasing coverage does not significantly improve the assembly quality and only increases sequencing costs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414ab82a-3f91-43d3-8c07-025f8a3379b2",
   "metadata": {},
   "source": [
    "## Process 1: Quality Assessment with FASTQC\n",
    "- **Program**: **FASTQC**\n",
    "- **Citation**: *Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data.*\n",
    "- **Manual**: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/\n",
    "- **FASTQC explained**: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/\n",
    "\n",
    "\n",
    "FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and determine some basic statistics about the data. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, it merely reads it and writes a report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "506c05fd-a403-4eb3-bebb-7d0cfcc51246",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# make a directory to store the output\n",
    "mkdir -p output-fastqc\n",
    "\n",
    "# run the program on the two read files (R1 and R2).\n",
    "echo \"RUNNING FASTQC\"\n",
    "fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24 > /dev/null 2>&1\n",
    "echo \"FASTQC analysis complete, output files:\"\n",
    "\n",
    "# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.\n",
    "ls output-fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8457502-3417-4a46-b213-1382c0b3dbbd",
   "metadata": {},
   "source": [
    "### FASTQC output\n",
    "\n",
    "The code below will open one of the interactive HTML output files produced by FASTQC.   \n",
    "\n",
    "I will highlight a few observations about this data. FASTQC is a standard tool described in detail elsewhere, including in other NIGMS modules.  Take a moment to open the link above [FASTQC explained](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/) and look at each figure in your report. The [manual](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) also provides links to examples of good and bad reports for comparison to your data.  \n",
    "\n",
    "- In our dataset we see a warning (**X**) on the *Per base sequence content*. This figure displays the proportion of each nucleotide (A, T, C, and G) at every position across all reads in the FASTQ file. During library preparation, the DNA underwent fragmentation, typically performed either by physical shearing or enzymatic digestion. Ideally, this process should be truly random. However, we observe a bias in our data, with an overrepresentation of 'T' nucleotides at the beginning of the reads. It appears the DNA may have preferentially fragmented at thymine-rich regions.\n",
    "- We also see a warning (**X**) on the *Sequence Duplication Levels*. FASTQC calculates how often each sequence appears in the dataset and generates a plot showing the distribution of duplication across the library. In a complex and diverse library, most sequences are expected to occur only once. While a low level of duplication can sometimes reflect deep coverage of a target region, a high level of duplication more often indicates technical bias, such as PCR over-amplification or limited library complexity. Identifying excessive duplication is important for assessing library quality and determining whether further optimization of the library preparation process is needed.  \n",
    "- I would like to take a moment to look at the *Per sequence GC content*. This figure displays the GC content calculated for each read in the FASTQ file and compares the observed distribution to a modeled normal distribution based on an ideal dataset. Each organism has a unique GC content signature, and we’ll explore this in more detail in Submodule 2. Deviations from the expected distribution in this figure may indicate contamination, sequencing bias, or the presence of sequences from organisms with differing GC content.\n",
    "- Also see the *Adapter Content* module, which identifies whether any standard adapter sequences are present in the reads. Adapters are added to the ends of DNA fragments during library preparation to enable sequencing. However, if a DNA fragment is shorter than the sequencing read length (100 bp in our case), the sequencing instrument may read through the fragment and into the adapter sequence. When this happens, it’s important to detect and remove these artificial sequences before proceeding with downstream analyses, such as read alignment or assembly, to avoid introducing bias or errors.\n",
    "\n",
    "There are warnings in this dataset, but they are not cause for concern. FASTQC does not alter the data in any way; instead, it provides helpful diagnostics and observations to be aware of as you move forward with your analysis.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5bfb2d6a-1a60-4b89-9898-1f5079d95962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1000\"\n",
       "            height=\"550\"\n",
       "            src=\"output-fastqc/reads_1_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fcc70516620>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## view HTML report\n",
    "from IPython.display import IFrame\n",
    "IFrame('output-fastqc/reads_1_fastqc.html', width=1000, height=550)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb6207a-2b47-453f-a4d0-b38bff239dab",
   "metadata": {},
   "source": [
    "## Process 2: Quality and adapter trimming with FASTP\n",
    "\n",
    "- **Program**: **fastp** - An Ultra-Fast All-in-One FASTQ Preprocessor\n",
    "- **Citation**: *Chen, S. (2023). Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*\n",
    "- **Manuscript**: [https://pubmed.ncbi.nlm.nih.gov/30423086/](https://pubmed.ncbi.nlm.nih.gov/30423086/)\n",
    "- **Manual**: [fastp GitHub Repository](https://github.com/OpenGene/fastp)\n",
    "\n",
    "Description:\n",
    "**fastp** is an all-in-one tool designed for preprocessing and quality control of high-throughput sequencing data. It performs several critical tasks, including base correction, adapter trimming, read filtering, and quality score recalibration. Additionally, it provides deduplication to remove PCR duplicates and offers comprehensive quality control through detailed reports and visualizations. Its efficiency allows it to handle large datasets rapidly, making it ideal for processing next-generation sequencing (NGS) data in various formats like FASTQ. Alternative tools for similar purposes include **Trimmomatic**, **Cutadapt**, and **FASTX-Toolkit**, which offer some of the same functionalities but may differ in speed, ease of use, and specific features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18047f7b-4838-4133-bf51-aa1d139caf2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# make a directory to store the output\n",
    "mkdir output-fastp\n",
    "\n",
    "# define input variables\n",
    "forward=raw-reads/reads_1.fastq.gz\n",
    "reverse=raw-reads/reads_2.fastq.gz\n",
    "\n",
    "# run the program\n",
    "fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# the resulting folder should contain new trimmed read files\n",
    "ls output-fastp/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b04f223-8e6b-4e68-95e3-3ba49dc29f79",
   "metadata": {},
   "source": [
    "### Explanation of fastp Results\n",
    "\n",
    "The **fastp** results printed above show the quality of the sequencing data both before and after filtering. These results provide insight into the number of reads, base quality, and the effectiveness of filtering steps like adapter trimming and quality control.\n",
    "\n",
    "- **<u>Before filtering</u>**:\n",
    "  - **Read1** and **Read2** both start with 1,591,280 reads, with high overall quality scores. Over 99% of bases have a quality score (Q20) above 20, and about 96% of bases have a quality score (Q30) above 30, indicating good overall quality in the raw data.\n",
    "\n",
    "- **<u>After filtering</u>**:\n",
    "  - The total number of reads decreases slightly to 1,568,979 for both **Read1** and **Read2**. Despite this reduction, the quality improves, with over 99% of bases achieving Q20 and over 96% of bases achieving Q30 for **Read1**, and over 98% for **Read2**. This shows that filtering removed low-quality reads while maintaining high-quality data.\n",
    "\n",
    "- **<u>Filtering summary</u>**:\n",
    "  - A total of 31,379,958 reads passed the quality filter, while 43,808 reads failed due to low quality. A small number (794) failed due to the presence of too many 'N' bases, and no reads were discarded due to being too short.\n",
    "  - **Adapter trimming** removed 5,320,114 bases across both reads, which improves the accuracy of downstream analysis.\n",
    "\n",
    "- **<u>Duplication rate</u>**: \n",
    "  - The duplication rate is low at 2.70%, meaning most reads are unique and useful for assembly and analysis.\n",
    "\n",
    "- **<u>Insert size</u>**: \n",
    "  - The insert size peak is 168, which reflects the typical distance between paired-end reads, providing valuable information about the sequencing library.\n",
    "\n",
    "These results suggest that **fastp** was effective in cleaning the raw sequencing data by removing low-quality reads and trimming adapter sequences, resulting in improved data quality suitable for downstream analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aa89305-5b2d-4614-9cd2-c144224187bc",
   "metadata": {},
   "source": [
    "## Process 3: Genome Assembly\n",
    "\n",
    "- Program: **SPAdes - St. Petersburg genome assembler**\n",
    "- Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*\n",
    "- Manuscript: [https://pubmed.ncbi.nlm.nih.gov/22506599/](https://pubmed.ncbi.nlm.nih.gov/22506599/)\n",
    "- Manual: https://ablab.github.io/spades/  \n",
    "\n",
    "SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single-end reads.  \n",
    "\n",
    "Adjust the --threads option to match the number of CPU cores available on the machine type you selected during setup. If you followed the recommended configuration, you can set --threads 32. With this setting, the process should take less than 10 minutes to complete."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40624982-3111-483a-8b9d-3a277339e589",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Tip</b>: I recommend running the code below, then watching the following video tutorial or reading more while it runs.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0a21d5-54e6-4cee-97ca-662d2d0c0885",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "## Assemble trimmed reads using the SPAdes assembler.\n",
    "\n",
    "# define input\n",
    "forward=output-fastp/trimmed_1.fastq.gz\n",
    "reverse=output-fastp/trimmed_2.fastq.gz\n",
    "\n",
    "# assemble reads with SPAdes\n",
    "echo \"RUNNING THE SPADES PIPELINE, PLEASE REVIEW MATERIALS BELOW WHILE IT RUNS\"\n",
    "spades.py -1 $forward -2 $reverse -o output-spades --threads 32 > logfile_spades.txt 2>&1\n",
    "echo \"SPADES assembly complete\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eec824dc-de4e-4456-89ac-955074826565",
   "metadata": {},
   "source": [
    "## Video Lesson: How genome assembly of short reads works.\n",
    "\n",
    "The following video describes the process in detail. Run the HTML code below to start the video."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20b517ea-349f-4534-95b8-4721947c705e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/ZmF6QROPlTU\" frameborder=\"0\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa1800eb-3f18-4590-aa17-e834e78b2b03",
   "metadata": {},
   "source": [
    "## Video Recap\n",
    "\n",
    "### De Bruijn Graphs in Genome Assembly\n",
    "\n",
    "A **de Bruijn graph** is a structure used in genome assembly to represent sequences using **k-mers**. Here’s how it works:\n",
    "\n",
    "1. **K-mers**: These are short sequences of length *k* derived from a longer DNA sequence. For example, from the sequence \"AGCT\", the 2-mers are \"AG\", \"GC\", and \"CT\".\n",
    "\n",
    "2. **Graph Construction**: Each k-mer becomes a vertex in the graph. An edge is drawn from k-mer **A** to k-mer **B** if the last *k-1* nucleotides of **A** match the first *k-1* nucleotides of **B**. This creates a directed graph where paths through the vertices represent potential sequences in the genome.\n",
    "\n",
    "3. **Contigs and Scaffolds**:\n",
    "   - **Contigs**: These are contiguous sequences formed by traversing paths in the graph.\n",
    "   - **Scaffolds**: Longer constructs that may combine multiple contigs, providing additional structure to the genome assembly.\n",
    "\n",
    "De Bruijn graphs enable efficient handling of sequencing data, especially for complex genomes with repetitive sequences.\n",
    "  \n",
    "  \n",
    "### Key challenges for genome assembly:\n",
    "**<u>Intrinsic Challenges:</u>**\n",
    "1. **Heterozygosity:** The alleles of a gene are not he same, yet we force them into a single consensus sequence.\n",
    "2. **Paralogy vs. Alleleism**: Genes come from other genes by a process of duplication. This results in two or more similar genes in an organism. There are two alleles in a diploid organism that are very similiar. How do you tell a duplicated gene from alleles of a gene?\n",
    "3. **Sequence complexity**: Simple sequence repeats (SSR), large scale repeats like transposable elements (TEs).\n",
    "\n",
    "**<u>Extrinsic Challenges:</u>**\n",
    "1. **Quality of DNA sequences (sequencing errors):** Each sequencing technology has specific patterns of error.\n",
    "2. **Length of DNA sequence reads**: Shorter reads are less likely to be unique or to include unique K-mers (see below).\n",
    "3. **Coverage**: Depth of coverage is a random process at best. Consequently some regions of the genome will have low levels of coverage.\n",
    "4. **Memory intensive**: Inherently requires large amounts of RAM for assebly and storage of input and output.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a255e7-e1f3-48d2-83f7-4ad40ff8defd",
   "metadata": {},
   "source": [
    "## Conclusion and wrap up\n",
    "\n",
    "End of submodule 1.  After learning about how genome sequencing works (and where it can go wrong) we went through standard processing steps (FASTQC and FASTP) and ultimetly ran a genome assembly tool. The final output from spades represents our draft genome sequence.\n",
    "\n",
    "Be sure to shut down your instance. We will run one more bit of code to copy our genome file for use in submodule 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab4f439-9a33-48bb-87de-7911ebebaf4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# did the assembler finish?\n",
    "ls output-spades/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fdc4e76-c58e-4e84-a9a4-acea99822fe1",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\"><b>Status Check</b>: If the assembly finished successfully you will see the file called contigs.fasta above. This is our genome assembly. Be sure to run the next set of code to prepare for submodule 02.</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c03cce-0795-4135-bb0a-45910788588e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# make a copy of the genome sequence in a new directory. This data will be used for Submodule 2.\n",
    "mkdir -p assembled-genome\n",
    "cp output-spades/contigs.fasta assembled-genome/genome.fasta\n",
    "\n",
    "\n",
    "# cleanup other files\n",
    "mkdir -p submodule01_data\n",
    "mv output-spades output-fastqc output-fastp submodule01_data\n",
    "mv fastp.json fastp.html logfile_spades.txt submodule01_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9c5af9-3f0e-4111-854d-740b9fdf87cb",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\"> <b>Attention Before You Proceed:</b>  If you're not moving on immediately to the next submodule, be sure to shutdown your instance. </div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
